// chain-registry/mainnet/atomone/chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const info = {
    $schema: '../chain.schema.json',
    chain_name: 'atomone',
    chain_type: 'cosmos',
    chain_id: 'atomone-1',
    website: 'https://atom.one',
    pretty_name: 'AtomOne',
    status: 'live',
    network_type: 'mainnet',
    bech32_prefix: 'atone',
    daemon_name: 'atomoned',
    node_home: '$HOME/.atomone',
    key_algos: ['secp256k1'],
    slip44: 118,
    fees: {
        fee_tokens: [{
                denom: 'uatone',
                fixed_min_gas_price: 0.025,
                low_gas_price: 0.025,
                average_gas_price: 0.04,
                high_gas_price: 0.08
            }]
    },
    staking: {
        staking_tokens: [{
                denom: 'uatone'
            }]
    },
    codebase: {
        git_repo: 'https://github.com/atomone-hub/atomone',
        recommended_version: 'v1.1.2',
        compatible_versions: ['v1.1.2'],
        binaries: {
            "linux/amd64": 'https://github.com/atomone-hub/atomone/releases/download/v1.1.2/atomoned-v1.1.2-linux-amd64',
            "linux/arm64": 'https://github.com/atomone-hub/atomone/releases/download/v1.1.2/atomoned-v1.1.2-linux-arm64',
            "darwin/amd64": 'https://github.com/atomone-hub/atomone/releases/download/v1.1.2/atomoned-v1.1.2-darwin-amd64',
            "darwin/arm64": 'https://github.com/atomone-hub/atomone/releases/download/v1.1.2/atomoned-v1.1.2-darwin-arm64',
            "windows/amd64": 'https://github.com/atomone-hub/atomone/releases/download/v1.1.2/atomoned-v1.1.2-windows-amd64.exe',
            "windows/arm64": 'https://github.com/atomone-hub/atomone/releases/download/v1.1.2/atomoned-v1.1.2-windows-arm64.exe'
        },
        genesis: {
            genesis_url: 'https://atomone.fra1.digitaloceanspaces.com/atomone-1/genesis.json'
        },
        consensus: {
            type: 'cometbft',
            version: 'v0.37.15'
        },
        sdk: {
            type: 'cosmos',
            version: 'v0.47.13'
        }
    },
    logo_URIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/atomone/images/atomone.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/atomone/images/atomone.svg'
    },
    description: 'The Cosmos community, at a crossroads, confronts divergent views on key aspects such as mission, tokenomics, and security philosophy. AtomOne emerges as a beacon, offering an alternative fork to navigate these waters, equipped to handle contingencies and embodying a bastion for diverse political thought.',
    apis: {
        rpc: [
            {
                address: 'https://atomone-rpc.allinbits.com:443',
                provider: 'AllInBits'
            },
            {
                address: 'https://atomone-rpc.cogwheel.zone',
                provider: 'Cogwheel ⚙️'
            },
            {
                address: 'https://atomone.rpc.nodeshub.online:443',
                provider: 'Nodes Hub 🛡️ 100% Slash Protected 🛡️'
            },
            {
                address: 'https://rpc-atomone.nodeist.net',
                provider: 'Nodeist'
            },
            {
                address: 'https://rpc-atomone-1.cros-nest.com:443',
                provider: 'crosnest'
            },
            {
                address: 'https://community.nuxian-node.ch:6797/atomone/trpc',
                provider: 'PRO Delegators'
            },
            {
                address: 'https://atomone-rpc.publicnode.com:443',
                provider: 'Allnodes ⚡️ Nodes & Staking'
            },
            {
                address: 'https://atomone_mainnet_rpc.chain.whenmoonwhenlambo.money',
                provider: '🚀 WHEN MOON 🌕 WHEN LAMBO 🔥'
            },
            {
                address: 'https://atomone-rpc.ibs.team:443',
                provider: 'Inter Blockchain Services'
            },
            {
                address: 'https://atomone-mainnet-rpc.autostake.com:443',
                provider: 'AutoStake 🛡️ Slash Protected'
            },
            {
                address: 'https://rpc.atomone-1.atomone.aviaone.com:443',
                provider: 'AVIAONE.com 🟢'
            },
            {
                address: 'https://atomone.rpc.m.stavr.tech:443',
                provider: '🔥STAVR🔥'
            },
            {
                address: 'https://rpc.atomone.citizenweb3.com:443',
                provider: 'Citizen Web3'
            },
            {
                address: 'https://atomone-mainnet-rpc.itrocket.net:443',
                provider: 'ITRocket'
            },
            {
                address: 'https://rpc-atone.vinjan.xyz',
                provider: 'Vinjan.Inc'
            },
            {
                address: 'https://rpc-atomone.ecostake.com',
                provider: 'ecostake'
            },
            {
                address: 'https://m-atomone.rpc.utsa.tech/',
                provider: 'UTSA'
            },
            {
                address: 'https://rpc.atomone.roomit.xyz',
                provider: 'Roomit'
            },
            {
                address: 'https://rpc.atomone.chaintools.tech',
                provider: 'ChainTools'
            },
            {
                address: 'https://atomone-mainnet-rpc.bonynode.online',
                provider: 'BonyNode💚'
            },
            {
                address: 'https://atomone-mainnet-rpc.shazoe.xyz',
                provider: 'Shazoe'
            }
        ],
        rest: [
            {
                address: 'https://atomone-api.allinbits.com',
                provider: 'AllInBits'
            },
            {
                address: 'https://atomone-api.cogwheel.zone',
                provider: 'Cogwheel ⚙️'
            },
            {
                address: 'https://atomone.api.nodeshub.online',
                provider: 'Nodes Hub 🛡️ 100% Slash Protected 🛡️'
            },
            {
                address: 'https://api-atomone.nodeist.net',
                provider: 'Nodeist'
            },
            {
                address: 'https://rest-atomone-1.cros-nest.com:443',
                provider: 'crosnest'
            },
            {
                address: 'https://community.nuxian-node.ch:6797/atomone/crpc',
                provider: 'PRO Delegators'
            },
            {
                address: 'https://atomone-rest.publicnode.com',
                provider: 'Allnodes ⚡️ Nodes & Staking'
            },
            {
                address: 'https://atomone_mainnet_api.chain.whenmoonwhenlambo.money',
                provider: '🚀 WHEN MOON 🌕 WHEN LAMBO 🔥'
            },
            {
                address: 'https://atomone-api.ibs.team:443',
                provider: 'Inter Blockchain Services'
            },
            {
                address: 'https://atomone-mainnet-lcd.autostake.com:443',
                provider: 'AutoStake 🛡️ Slash Protected'
            },
            {
                address: 'https://api.atomone-1.atomone.aviaone.com',
                provider: 'AVIAONE.com 🟢'
            },
            {
                address: 'https://atomone.api.m.stavr.tech',
                provider: '🔥STAVR🔥'
            },
            {
                address: 'https://api.atomone.citizenweb3.com',
                provider: 'Citizen Web3'
            },
            {
                address: 'https://atomone-mainnet-api.itrocket.net',
                provider: 'ITRocket'
            },
            {
                address: 'https://api-atone.vinjan.xyz',
                provider: 'Vinjan.Inc'
            },
            {
                address: 'https://rest-atomone.ecostake.com',
                provider: 'ecostake'
            },
            {
                address: 'https://m-atomone.api.utsa.tech/',
                provider: 'UTSA'
            },
            {
                address: 'https://api.atomone.roomit.xyz',
                provider: 'Roomit'
            },
            {
                address: 'https://api.atomone.chaintools.tech',
                provider: 'ChainTools'
            },
            {
                address: 'https://atomone-mainnet-api.bonynode.online',
                provider: 'BonyNode💚'
            },
            {
                address: 'https://atomone-mainnet-api.shazoe.xyz',
                provider: 'Shazoe'
            }
        ],
        grpc: [
            {
                address: 'atomone-grpc.allinbits.com:9090',
                provider: 'AllInBits'
            },
            {
                address: 'atomone-grpc.cogwheel.zone:443',
                provider: 'Cogwheel ⚙️'
            },
            {
                address: 'atomone.grpc.nodeshub.online',
                provider: 'Nodes Hub'
            },
            {
                address: 'https://grpc-atomone.nodeist.net',
                provider: 'Nodeist'
            },
            {
                address: 'atomone-grpc.publicnode.com:443',
                provider: 'Allnodes ⚡️ Nodes & Staking'
            },
            {
                address: 'atomone-mainnet-grpc.autostake.com:443',
                provider: 'AutoStake 🛡️ Slash Protected'
            },
            {
                address: 'http://grpc.atomone-1.atomone.aviaone.com:9102',
                provider: 'AVIAONE.com 🟢'
            },
            {
                address: 'atomone.grpc.m.stavr.tech:7201',
                provider: '🔥STAVR🔥'
            },
            {
                address: ' grpc.atomone.citizenweb3.com',
                provider: 'Citizen Web3'
            },
            {
                address: 'atomone-mainnet-grpc.itrocket.net:443',
                provider: 'ITRocket'
            },
            {
                address: 'grpc.atomone.roomit.xyz:8443',
                provider: 'Roomit'
            },
            {
                address: 'grpc.atomone.chaintools.tech:443',
                provider: 'ChainTools'
            },
            {
                address: 'https://atomone-mainnet-grpc.bonynode.online:443',
                provider: 'BonyNode💚'
            },
            {
                address: 'atomone-mainnet-grpc.shazoe.xyz:12090',
                provider: 'Shazoe'
            }
        ]
    },
    explorers: [
        {
            kind: 'ping.pub',
            url: 'https://explorer.allinbits.com/atomone',
            tx_page: 'https://explorer.allinbits.com/atomone/tx/${txHash}',
            account_page: 'https://explorer.allinbits.com/atomone/account/${accountAddress}'
        },
        {
            kind: 'Nodes Hub 🛡️ 100% Slash Protected 🛡️',
            url: 'https://explorer.nodeshub.online/atomone/',
            tx_page: 'https://explorer.nodeshub.online/atomone/tx/${txHash}',
            account_page: 'https://explorer.nodeshub.online/atomone/accounts/${accountAddress}'
        },
        {
            kind: 'Nodeist Explorer',
            url: 'https://explorer.ist/atomone',
            tx_page: 'https://explorer.ist/atomone/tx/${txHash}',
            account_page: 'https://explorer.ist/atomone/account/${accountAddress}'
        },
        {
            kind: '🔥STAVR🔥 Explorer',
            url: 'https://explorer.stavr.tech/Atomone-Mainnet',
            tx_page: 'https://explorer.stavr.tech/Atomone-Mainnet/tx/${txHash}',
            account_page: 'https://explorer.stavr.tech/Atomone-Mainnet/account/${accountAddress}'
        },
        {
            kind: '🚀 WHEN MOON 🌕 WHEN LAMBO 🔥',
            url: 'https://explorer.whenmoonwhenlambo.money/atomone',
            tx_page: 'https://explorer.whenmoonwhenlambo.money/atomone/tx/${txHash}',
            account_page: 'https://explorer.whenmoonwhenlambo.money/atomone/account/${accountAddress}'
        },
        {
            kind: 'AVIAONE.com 🟢',
            url: 'https://mainnet.explorer.aviaone.com/atomone',
            tx_page: 'https://mainnet.explorer.aviaone.com/atomone/tx/${txHash}',
            account_page: 'https://mainnet.explorer.aviaone.com/atomone/account/${accountAddress}'
        },
        {
            kind: 'ITRocket',
            url: 'https://mainnet.itrocket.net/atomone',
            tx_page: 'https://mainnet.itrocket.net/atomone/tx/${txHash}',
            account_page: 'https://mainnet.itrocket.net/atomone/account/${accountAddress}'
        },
        {
            kind: 'Roomit',
            url: 'https://explorer.tendermint.roomit.xyz/atomone-mainnet',
            tx_page: 'https://explorer.tendermint.roomit.xyz/atomone-mainnet/tx/${txHash}',
            account_page: 'https://explorer.tendermint.roomit.xyz/atomone-mainnet/account/${accountAddress}'
        },
        {
            kind: 'ChainTools',
            url: 'https://explorer.chaintools.tech/atomone',
            tx_page: 'https://explorer.chaintools.tech/atomone/tx/${txHash}',
            account_page: 'https://explorer.chaintools.tech/atomone/account/${accountAddress}'
        },
        {
            kind: 'Valopers',
            url: 'https://atomone.valopers.com/',
            tx_page: 'https://atomone.valopers.com/transactions/${txHash}',
            account_page: 'https://atomone.valopers.com/account/${accountAddress}'
        },
        {
            kind: 'BonyNode💚',
            url: 'https://explorer.bonynode.online/atomone',
            tx_page: 'https://explorer.bonynode.online/atomone/transactions/${txHash}',
            account_page: 'https://explorer.bonynode.online/atomone/account/${accountAddress}'
        }
    ],
    images: [{
            png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/atomone/images/atomone.png',
            svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/atomone/images/atomone.svg'
        }]
};
exports.default = info;
