// @injectivelabs/sdk-ts/dist/cjs/core/modules/wasm/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecPrivilegedArgOffChainVaultSubscribe = exports.ExecPrivilegedArgOffChainVaultRedeem = exports.ExecPrivilegedArgVaultSubscribe = exports.ExecPrivilegedArgVaultRedeem = exports.MsgPrivilegedExecuteContract = exports.MsgExecuteContractCompat = exports.MsgInstantiateContract = exports.MsgMigrateContract = exports.MsgExecuteContract = exports.MsgUpdateAdmin = exports.MsgStoreCode = exports.ExecArgCW20AdapterRedeemAndTransfer = exports.ExecArgCreatePerpGridStrategy = exports.ExecArgCreateSpotGridStrategy = exports.ExecArgRemoveGridStrategy = exports.ExecArgIncreaseAllowance = exports.ExecArgUpdateGridConfig = exports.ExecArgInitiateTransfer = exports.ExecArgNeptuneWithdraw = exports.ExecArgSwapExactOutput = exports.ExecArgNeptuneDeposit = exports.ExecArgCreateCampaign = exports.ExecArgDepositTokens = exports.ExecArgSwapMinOutput = exports.ExecArgFundCampaign = exports.ExecArgCW20Transfer = exports.ExecArgCreateRound = exports.ExecArgSubmitVaa = exports.ExecArgCW20Send = void 0;
const ExecArgCW20Send_js_1 = __importDefault(require("./exec-args/ExecArgCW20Send.js"));
exports.ExecArgCW20Send = ExecArgCW20Send_js_1.default;
const ExecArgSubmitVaa_js_1 = __importDefault(require("./exec-args/ExecArgSubmitVaa.js"));
exports.ExecArgSubmitVaa = ExecArgSubmitVaa_js_1.default;
const ExecArgCreateRound_js_1 = __importDefault(require("./exec-args/ExecArgCreateRound.js"));
exports.ExecArgCreateRound = ExecArgCreateRound_js_1.default;
const ExecArgFundCampaign_js_1 = __importDefault(require("./exec-args/ExecArgFundCampaign.js"));
exports.ExecArgFundCampaign = ExecArgFundCampaign_js_1.default;
const ExecArgCW20Transfer_js_1 = __importDefault(require("./exec-args/ExecArgCW20Transfer.js"));
exports.ExecArgCW20Transfer = ExecArgCW20Transfer_js_1.default;
const ExecArgSwapMinOutput_js_1 = __importDefault(require("./exec-args/ExecArgSwapMinOutput.js"));
exports.ExecArgSwapMinOutput = ExecArgSwapMinOutput_js_1.default;
const ExecArgDepositTokens_js_1 = __importDefault(require("./exec-args/ExecArgDepositTokens.js"));
exports.ExecArgDepositTokens = ExecArgDepositTokens_js_1.default;
const ExecArgCreateCampaign_js_1 = __importDefault(require("./exec-args/ExecArgCreateCampaign.js"));
exports.ExecArgCreateCampaign = ExecArgCreateCampaign_js_1.default;
const ExecArgNeptuneDeposit_js_1 = __importDefault(require("./exec-args/ExecArgNeptuneDeposit.js"));
exports.ExecArgNeptuneDeposit = ExecArgNeptuneDeposit_js_1.default;
const ExecArgSwapExactOutput_js_1 = __importDefault(require("./exec-args/ExecArgSwapExactOutput.js"));
exports.ExecArgSwapExactOutput = ExecArgSwapExactOutput_js_1.default;
const ExecArgNeptuneWithdraw_js_1 = __importDefault(require("./exec-args/ExecArgNeptuneWithdraw.js"));
exports.ExecArgNeptuneWithdraw = ExecArgNeptuneWithdraw_js_1.default;
const ExecArgInitiateTransfer_js_1 = __importDefault(require("./exec-args/ExecArgInitiateTransfer.js"));
exports.ExecArgInitiateTransfer = ExecArgInitiateTransfer_js_1.default;
const ExecArgUpdateGridConfig_js_1 = __importDefault(require("./exec-args/ExecArgUpdateGridConfig.js"));
exports.ExecArgUpdateGridConfig = ExecArgUpdateGridConfig_js_1.default;
const ExecArgIncreaseAllowance_js_1 = __importDefault(require("./exec-args/ExecArgIncreaseAllowance.js"));
exports.ExecArgIncreaseAllowance = ExecArgIncreaseAllowance_js_1.default;
const ExecArgRemoveGridStrategy_js_1 = __importDefault(require("./exec-args/ExecArgRemoveGridStrategy.js"));
exports.ExecArgRemoveGridStrategy = ExecArgRemoveGridStrategy_js_1.default;
const ExecArgCreateSpotGridStrategy_js_1 = __importDefault(require("./exec-args/ExecArgCreateSpotGridStrategy.js"));
exports.ExecArgCreateSpotGridStrategy = ExecArgCreateSpotGridStrategy_js_1.default;
const ExecArgCreatePerpGridStrategy_js_1 = __importDefault(require("./exec-args/ExecArgCreatePerpGridStrategy.js"));
exports.ExecArgCreatePerpGridStrategy = ExecArgCreatePerpGridStrategy_js_1.default;
const ExecArgCW20AdapterRedeemAndTransfer_js_1 = __importDefault(require("./exec-args/ExecArgCW20AdapterRedeemAndTransfer.js"));
exports.ExecArgCW20AdapterRedeemAndTransfer = ExecArgCW20AdapterRedeemAndTransfer_js_1.default;
const MsgStoreCode_js_1 = __importDefault(require("./msgs/MsgStoreCode.js"));
exports.MsgStoreCode = MsgStoreCode_js_1.default;
const MsgUpdateAdmin_js_1 = __importDefault(require("./msgs/MsgUpdateAdmin.js"));
exports.MsgUpdateAdmin = MsgUpdateAdmin_js_1.default;
const MsgExecuteContract_js_1 = __importDefault(require("./msgs/MsgExecuteContract.js"));
exports.MsgExecuteContract = MsgExecuteContract_js_1.default;
const MsgMigrateContract_js_1 = __importDefault(require("./msgs/MsgMigrateContract.js"));
exports.MsgMigrateContract = MsgMigrateContract_js_1.default;
const MsgInstantiateContract_js_1 = __importDefault(require("./msgs/MsgInstantiateContract.js"));
exports.MsgInstantiateContract = MsgInstantiateContract_js_1.default;
const MsgExecuteContractCompat_js_1 = __importDefault(require("./msgs/MsgExecuteContractCompat.js"));
exports.MsgExecuteContractCompat = MsgExecuteContractCompat_js_1.default;
const MsgPrivilegedExecuteContract_js_1 = __importDefault(require("./msgs/MsgPrivilegedExecuteContract.js"));
exports.MsgPrivilegedExecuteContract = MsgPrivilegedExecuteContract_js_1.default;
const ExecPrivilegedArgVaultRedeem_js_1 = __importDefault(require("./exec-priv-args/ExecPrivilegedArgVaultRedeem.js"));
exports.ExecPrivilegedArgVaultRedeem = ExecPrivilegedArgVaultRedeem_js_1.default;
const ExecPrivilegedArgVaultSubscribe_js_1 = __importDefault(require("./exec-priv-args/ExecPrivilegedArgVaultSubscribe.js"));
exports.ExecPrivilegedArgVaultSubscribe = ExecPrivilegedArgVaultSubscribe_js_1.default;
const ExecPrivilegedArgOffChainVaultRedeem_js_1 = __importDefault(require("./exec-priv-args/ExecPrivilegedArgOffChainVaultRedeem.js"));
exports.ExecPrivilegedArgOffChainVaultRedeem = ExecPrivilegedArgOffChainVaultRedeem_js_1.default;
const ExecPrivilegedArgOffChainVaultSubscribe_js_1 = __importDefault(require("./exec-priv-args/ExecPrivilegedArgOffChainVaultSubscribe.js"));
exports.ExecPrivilegedArgOffChainVaultSubscribe = ExecPrivilegedArgOffChainVaultSubscribe_js_1.default;
__exportStar(require("./types.js"), exports);
__exportStar(require("./exec-args.js"), exports);
