// @injectivelabs/utils/dist/cjs/classes/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamManager = exports.HttpRestClient = exports.HttpClient = exports.BigNumber = exports.BigNumberInWei = exports.BigNumberInBase = exports.LocalStorage = exports.StatusType = exports.Status = void 0;
const LocalStorage_js_1 = __importDefault(require("./LocalStorage.js"));
exports.LocalStorage = LocalStorage_js_1.default;
const BigNumberInBase_js_1 = __importDefault(require("./BigNumber/BigNumberInBase.js"));
exports.BigNumberInBase = BigNumberInBase_js_1.default;
const BigNumberInWei_js_1 = __importDefault(require("./BigNumber/BigNumberInWei.js"));
exports.BigNumberInWei = BigNumberInWei_js_1.default;
const BigNumber_js_1 = __importDefault(require("./BigNumber/BigNumber.js"));
exports.BigNumber = BigNumber_js_1.default;
const HttpClient_js_1 = __importDefault(require("./HttpClient.js"));
exports.HttpClient = HttpClient_js_1.default;
const HttpRestClient_js_1 = __importDefault(require("./HttpRestClient.js"));
exports.HttpRestClient = HttpRestClient_js_1.default;
const StreamManager_js_1 = __importDefault(require("./StreamManager.js"));
exports.StreamManager = StreamManager_js_1.default;
const Status_js_1 = __importStar(require("./Status.js"));
exports.Status = Status_js_1.default;
Object.defineProperty(exports, "StatusType", { enumerable: true, get: function () { return Status_js_1.StatusType; } });
