// interchain/dist/codegen/ibc/core/channel/v1/query.rpc.Query.js
import { BinaryReader } from "../../../../binary";
import { createProtobufRpcClient } from "@cosmjs/stargate";
import { QueryChannelRequest, QueryChannelResponse, QueryChannelsRequest, QueryChannelsResponse, QueryConnectionChannelsRequest, QueryConnectionChannelsResponse, QueryChannelClientStateRequest, QueryChannelClientStateResponse, QueryChannelConsensusStateRequest, QueryChannelConsensusStateResponse, QueryPacketCommitmentRequest, QueryPacketCommitmentResponse, QueryPacketCommitmentsRequest, QueryPacketCommitmentsResponse, QueryPacketReceiptRequest, QueryPacketReceiptResponse, QueryPacketAcknowledgementRequest, QueryPacketAcknowledgementResponse, QueryPacketAcknowledgementsRequest, QueryPacketAcknowledgementsResponse, QueryUnreceivedPacketsRequest, QueryUnreceivedPacketsResponse, QueryUnreceivedAcksRequest, QueryUnreceivedAcksResponse, QueryNextSequenceReceiveRequest, QueryNextSequenceReceiveResponse, QueryNextSequenceSendRequest, QueryNextSequenceSendResponse } from "./query";
export class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* Channel queries an IBC Channel. */
    channel = async (request) => {
        const data = QueryChannelRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "Channel", data);
        return promise.then(data => QueryChannelResponse.decode(new BinaryReader(data)));
    };
    /* Channels queries all the IBC channels of a chain. */
    channels = async (request = {
        pagination: undefined
    }) => {
        const data = QueryChannelsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "Channels", data);
        return promise.then(data => QueryChannelsResponse.decode(new BinaryReader(data)));
    };
    /* ConnectionChannels queries all the channels associated with a connection
     end. */
    connectionChannels = async (request) => {
        const data = QueryConnectionChannelsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "ConnectionChannels", data);
        return promise.then(data => QueryConnectionChannelsResponse.decode(new BinaryReader(data)));
    };
    /* ChannelClientState queries for the client state for the channel associated
     with the provided channel identifiers. */
    channelClientState = async (request) => {
        const data = QueryChannelClientStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "ChannelClientState", data);
        return promise.then(data => QueryChannelClientStateResponse.decode(new BinaryReader(data)));
    };
    /* ChannelConsensusState queries for the consensus state for the channel
     associated with the provided channel identifiers. */
    channelConsensusState = async (request) => {
        const data = QueryChannelConsensusStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "ChannelConsensusState", data);
        return promise.then(data => QueryChannelConsensusStateResponse.decode(new BinaryReader(data)));
    };
    /* PacketCommitment queries a stored packet commitment hash. */
    packetCommitment = async (request) => {
        const data = QueryPacketCommitmentRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketCommitment", data);
        return promise.then(data => QueryPacketCommitmentResponse.decode(new BinaryReader(data)));
    };
    /* PacketCommitments returns all the packet commitments hashes associated
     with a channel. */
    packetCommitments = async (request) => {
        const data = QueryPacketCommitmentsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketCommitments", data);
        return promise.then(data => QueryPacketCommitmentsResponse.decode(new BinaryReader(data)));
    };
    /* PacketReceipt queries if a given packet sequence has been received on the
     queried chain */
    packetReceipt = async (request) => {
        const data = QueryPacketReceiptRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketReceipt", data);
        return promise.then(data => QueryPacketReceiptResponse.decode(new BinaryReader(data)));
    };
    /* PacketAcknowledgement queries a stored packet acknowledgement hash. */
    packetAcknowledgement = async (request) => {
        const data = QueryPacketAcknowledgementRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketAcknowledgement", data);
        return promise.then(data => QueryPacketAcknowledgementResponse.decode(new BinaryReader(data)));
    };
    /* PacketAcknowledgements returns all the packet acknowledgements associated
     with a channel. */
    packetAcknowledgements = async (request) => {
        const data = QueryPacketAcknowledgementsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketAcknowledgements", data);
        return promise.then(data => QueryPacketAcknowledgementsResponse.decode(new BinaryReader(data)));
    };
    /* UnreceivedPackets returns all the unreceived IBC packets associated with a
     channel and sequences. */
    unreceivedPackets = async (request) => {
        const data = QueryUnreceivedPacketsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "UnreceivedPackets", data);
        return promise.then(data => QueryUnreceivedPacketsResponse.decode(new BinaryReader(data)));
    };
    /* UnreceivedAcks returns all the unreceived IBC acknowledgements associated
     with a channel and sequences. */
    unreceivedAcks = async (request) => {
        const data = QueryUnreceivedAcksRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "UnreceivedAcks", data);
        return promise.then(data => QueryUnreceivedAcksResponse.decode(new BinaryReader(data)));
    };
    /* NextSequenceReceive returns the next receive sequence for a given channel. */
    nextSequenceReceive = async (request) => {
        const data = QueryNextSequenceReceiveRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "NextSequenceReceive", data);
        return promise.then(data => QueryNextSequenceReceiveResponse.decode(new BinaryReader(data)));
    };
    /* NextSequenceSend returns the next send sequence for a given channel. */
    nextSequenceSend = async (request) => {
        const data = QueryNextSequenceSendRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "NextSequenceSend", data);
        return promise.then(data => QueryNextSequenceSendResponse.decode(new BinaryReader(data)));
    };
}
export const createRpcQueryExtension = (base) => {
    const rpc = createProtobufRpcClient(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        channel(request) {
            return queryService.channel(request);
        },
        channels(request) {
            return queryService.channels(request);
        },
        connectionChannels(request) {
            return queryService.connectionChannels(request);
        },
        channelClientState(request) {
            return queryService.channelClientState(request);
        },
        channelConsensusState(request) {
            return queryService.channelConsensusState(request);
        },
        packetCommitment(request) {
            return queryService.packetCommitment(request);
        },
        packetCommitments(request) {
            return queryService.packetCommitments(request);
        },
        packetReceipt(request) {
            return queryService.packetReceipt(request);
        },
        packetAcknowledgement(request) {
            return queryService.packetAcknowledgement(request);
        },
        packetAcknowledgements(request) {
            return queryService.packetAcknowledgements(request);
        },
        unreceivedPackets(request) {
            return queryService.unreceivedPackets(request);
        },
        unreceivedAcks(request) {
            return queryService.unreceivedAcks(request);
        },
        nextSequenceReceive(request) {
            return queryService.nextSequenceReceive(request);
        },
        nextSequenceSend(request) {
            return queryService.nextSequenceSend(request);
        }
    };
};
