// interchain/dist/cjs/codegen/ibc/core/connection/v1/query.rpc.Query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
const binary_1 = require("../../../../binary");
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("./query");
class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* Connection queries an IBC connection end. */
    connection = async (request) => {
        const data = query_1.QueryConnectionRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "Connection", data);
        return promise.then(data => query_1.QueryConnectionResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* Connections queries all the IBC connections of a chain. */
    connections = async (request = {
        pagination: undefined
    }) => {
        const data = query_1.QueryConnectionsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "Connections", data);
        return promise.then(data => query_1.QueryConnectionsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ClientConnections queries the connection paths associated with a client
     state. */
    clientConnections = async (request) => {
        const data = query_1.QueryClientConnectionsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ClientConnections", data);
        return promise.then(data => query_1.QueryClientConnectionsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConnectionClientState queries the client state associated with the
     connection. */
    connectionClientState = async (request) => {
        const data = query_1.QueryConnectionClientStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ConnectionClientState", data);
        return promise.then(data => query_1.QueryConnectionClientStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConnectionConsensusState queries the consensus state associated with the
     connection. */
    connectionConsensusState = async (request) => {
        const data = query_1.QueryConnectionConsensusStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ConnectionConsensusState", data);
        return promise.then(data => query_1.QueryConnectionConsensusStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConnectionParams queries all parameters of the ibc connection submodule. */
    connectionParams = async (request = {}) => {
        const data = query_1.QueryConnectionParamsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ConnectionParams", data);
        return promise.then(data => query_1.QueryConnectionParamsResponse.decode(new binary_1.BinaryReader(data)));
    };
}
exports.QueryClientImpl = QueryClientImpl;
const createRpcQueryExtension = (base) => {
    const rpc = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        connection(request) {
            return queryService.connection(request);
        },
        connections(request) {
            return queryService.connections(request);
        },
        clientConnections(request) {
            return queryService.clientConnections(request);
        },
        connectionClientState(request) {
            return queryService.connectionClientState(request);
        },
        connectionConsensusState(request) {
            return queryService.connectionConsensusState(request);
        },
        connectionParams(request) {
            return queryService.connectionParams(request);
        }
    };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
