// @lifi/sdk/src/_cjs/core/BaseStepExecutor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseStepExecutor = void 0;
const StatusManager_js_1 = require("./StatusManager.js");
const defaultInteractionSettings = {
    allowInteraction: true,
    allowUpdates: true,
    allowExecution: true,
};
class BaseStepExecutor {
    constructor(options) {
        Object.defineProperty(this, "executionOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusManager", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "allowUserInteraction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "allowExecution", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "setInteraction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (settings) => {
                const interactionSettings = {
                    ...defaultInteractionSettings,
                    ...settings,
                };
                this.allowUserInteraction = interactionSettings.allowInteraction;
                this.statusManager.allowUpdates(interactionSettings.allowUpdates);
                this.allowExecution = interactionSettings.allowExecution;
            }
        });
        this.statusManager = new StatusManager_js_1.StatusManager(options.routeId);
        this.executionOptions = options.executionOptions;
    }
}
exports.BaseStepExecutor = BaseStepExecutor;
