// solana-agent-kit/dist/langchain/adrena/close_trade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaPerpCloseTradeTool = void 0;
const web3_js_1 = require("@solana/web3.js");
const tools_1 = require("langchain/tools");
class SolanaPerpCloseTradeTool extends tools_1.Tool {
    constructor(solanaKit) {
        super();
        this.solanaKit = solanaKit;
        this.name = "solana_close_perp_trade";
        this.description = `This tool can be used to close perpetuals trade ( It uses Adrena Protocol ).

  Inputs ( input is a JSON string ):
  tradeMint: string, eg "J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn", "DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263" etc. (optional)
  price?: number, eg 100 (optional)
  side: string, eg: "long" or "short"`;
    }
    async _call(input) {
        try {
            const parsedInput = JSON.parse(input);
            const tx = parsedInput.side === "long"
                ? await this.solanaKit.closePerpTradeLong({
                    price: parsedInput.price,
                    tradeMint: new web3_js_1.PublicKey(parsedInput.tradeMint),
                })
                : await this.solanaKit.closePerpTradeShort({
                    price: parsedInput.price,
                    tradeMint: new web3_js_1.PublicKey(parsedInput.tradeMint),
                });
            return JSON.stringify({
                status: "success",
                message: "Perpetual trade closed successfully",
                transaction: tx,
                price: parsedInput.price,
                tradeMint: new web3_js_1.PublicKey(parsedInput.tradeMint),
                side: parsedInput.side,
            });
        }
        catch (error) {
            return JSON.stringify({
                status: "error",
                message: error.message,
                code: error.code || "UNKNOWN_ERROR",
            });
        }
    }
}
exports.SolanaPerpCloseTradeTool = SolanaPerpCloseTradeTool;
