// @tensor-hq/tensor-common/dist/cjs/metaplex/pdas/auction_house.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findPurchaseReceiptPda = exports.findBidReceiptPda = exports.findListingReceiptPda = exports.findAuctionHouseTradeStatePda = exports.findAuctionHouseBuyerEscrowPda = exports.findAuctionHouseTreasuryPda = exports.findAuctionHouseFeePda = exports.findAuctionHouseProgramAsSignerPda = exports.findAuctioneerPda = exports.findAuctionHousePda = void 0;
const mpl_auction_house_1 = require("@metaplex-foundation/mpl-auction-house");
const web3_js_1 = require("@solana/web3.js");
const buffer_1 = require("buffer");
const findAuctionHousePda = (creator, treasuryMint, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('auction_house', 'utf8'),
        creator.toBuffer(),
        treasuryMint.toBuffer(),
    ], programId);
};
exports.findAuctionHousePda = findAuctionHousePda;
const findAuctioneerPda = (auctionHouse, auctioneerAuthority, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('auctioneer', 'utf8'),
        auctionHouse.toBuffer(),
        auctioneerAuthority.toBuffer(),
    ], programId);
};
exports.findAuctioneerPda = findAuctioneerPda;
const findAuctionHouseProgramAsSignerPda = (programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([buffer_1.Buffer.from('auction_house', 'utf8'), buffer_1.Buffer.from('signer', 'utf8')], programId);
};
exports.findAuctionHouseProgramAsSignerPda = findAuctionHouseProgramAsSignerPda;
const findAuctionHouseFeePda = (auctionHouse, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('auction_house', 'utf8'),
        auctionHouse.toBuffer(),
        buffer_1.Buffer.from('fee_payer', 'utf8'),
    ], programId);
};
exports.findAuctionHouseFeePda = findAuctionHouseFeePda;
const findAuctionHouseTreasuryPda = (auctionHouse, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('auction_house', 'utf8'),
        auctionHouse.toBuffer(),
        buffer_1.Buffer.from('treasury', 'utf8'),
    ], programId);
};
exports.findAuctionHouseTreasuryPda = findAuctionHouseTreasuryPda;
const findAuctionHouseBuyerEscrowPda = (auctionHouse, buyer, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('auction_house', 'utf8'),
        auctionHouse.toBuffer(),
        buyer.toBuffer(),
    ], programId);
};
exports.findAuctionHouseBuyerEscrowPda = findAuctionHouseBuyerEscrowPda;
const findAuctionHouseTradeStatePda = (auctionHouse, wallet, treasuryMint, tokenMint, buyPrice, tokenSize, tokenAccount, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('auction_house', 'utf8'),
        wallet.toBuffer(),
        auctionHouse.toBuffer(),
        ...(tokenAccount ? [tokenAccount.toBuffer()] : []),
        treasuryMint.toBuffer(),
        tokenMint.toBuffer(),
        buyPrice.toArrayLike(buffer_1.Buffer, 'le', 8),
        tokenSize.toArrayLike(buffer_1.Buffer, 'le', 8),
    ], programId);
};
exports.findAuctionHouseTradeStatePda = findAuctionHouseTradeStatePda;
const findListingReceiptPda = (tradeState, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([buffer_1.Buffer.from('listing_receipt', 'utf8'), tradeState.toBuffer()], programId);
};
exports.findListingReceiptPda = findListingReceiptPda;
const findBidReceiptPda = (tradeState, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([buffer_1.Buffer.from('bid_receipt', 'utf8'), tradeState.toBuffer()], programId);
};
exports.findBidReceiptPda = findBidReceiptPda;
const findPurchaseReceiptPda = (sellerTradeState, buyerTradeState, programId = mpl_auction_house_1.PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([
        buffer_1.Buffer.from('purchase_receipt', 'utf8'),
        sellerTradeState.toBuffer(),
        buyerTradeState.toBuffer(),
    ], programId);
};
exports.findPurchaseReceiptPda = findPurchaseReceiptPda;
