// @tensor-hq/tensor-common/dist/cjs/metaplex/pdas/bubblegum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLeafAssetId = exports.findBubblegumMintAuthorityPda = exports.findBubblegumTreeAuthorityPda = exports.findBubblegumSignerPda = exports.computeMetadataArgsHash = exports.BUBBLEGUM_PROGRAM_ID = void 0;
const mpl_bubblegum_1 = require("@metaplex-foundation/mpl-bubblegum");
const web3_js_1 = require("@solana/web3.js");
const js_sha3_1 = require("js-sha3");
exports.BUBBLEGUM_PROGRAM_ID = new web3_js_1.PublicKey('BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
function computeMetadataArgsHash(metadata) {
    const [serializedMetadata] = mpl_bubblegum_1.metadataArgsBeet.serialize(metadata);
    return Buffer.from(js_sha3_1.keccak_256.digest(serializedMetadata));
}
exports.computeMetadataArgsHash = computeMetadataArgsHash;
const findBubblegumSignerPda = (programId = exports.BUBBLEGUM_PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([Buffer.from('collection_cpi')], programId);
};
exports.findBubblegumSignerPda = findBubblegumSignerPda;
const findBubblegumTreeAuthorityPda = (merkleTree, programId = exports.BUBBLEGUM_PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([merkleTree.toBytes()], programId);
};
exports.findBubblegumTreeAuthorityPda = findBubblegumTreeAuthorityPda;
const findBubblegumMintAuthorityPda = (mint, programId = exports.BUBBLEGUM_PROGRAM_ID) => {
    return web3_js_1.PublicKey.findProgramAddressSync([mint.toBytes()], programId);
};
exports.findBubblegumMintAuthorityPda = findBubblegumMintAuthorityPda;
function getLeafAssetId(tree, leafIndex, programId = exports.BUBBLEGUM_PROGRAM_ID) {
    const [assetId] = web3_js_1.PublicKey.findProgramAddressSync([
        Buffer.from('asset', 'utf8'),
        tree.toBuffer(),
        Uint8Array.from(leafIndex.toArray('le', 8)),
    ], programId);
    return assetId;
}
exports.getLeafAssetId = getLeafAssetId;
