// @massalabs/massa-web3/dist/cmd/contracts-wrappers/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WMASBuildnet = exports.WETHbt = exports.USDTbt = exports.USDCs = exports.WETHs = exports.DAIs = exports.WMAS = exports.PUR = exports.WETHb = exports.WETHe = exports.DAIe = exports.USDTb = exports.USDCe = exports.checkNetwork = exports.BUILDNET_TOKENS = exports.MAINNET_TOKENS = void 0;
const utils_1 = require("../utils");
const token_1 = require("./token");
exports.MAINNET_TOKENS = {
    USDCe: 'AS1hCJXjndR4c9vekLWsXGnrdigp4AaZ7uYG3UKFzzKnWVsrNLPJ',
    USDTb: 'AS12LKs9txoSSy8JgFJgV96m8k5z9pgzjYMYSshwN67mFVuj3bdUV',
    DAIe: 'AS1ZGF1upwp9kPRvDKLxFAKRebgg7b3RWDnhgV7VvdZkZsUL7Nuv',
    WETHe: 'AS124vf3YfAJCSCQVYKczzuWWpXrximFpbTmX4rheLs5uNSftiiRY',
    WETHb: 'AS125oPLYRTtfVjpWisPZVTLjBhCFfQ1jDsi75XNtRm1NZux54eCj',
    PUR: 'AS133eqPPaPttJ6hJnk3sfoG5cjFFqBDi1VGxdo2wzWkq8AfZnan',
    WMAS: 'AS12U4TZfNK7qoLyEERBBRDMu8nm5MKoRzPXDXans4v9wdATZedz9',
};
exports.BUILDNET_TOKENS = {
    DAIs: 'AS12LpYyAjYRJfYhyu7fkrS224gMdvFHVEeVWoeHZzMdhis7UZ3Eb',
    WETHs: 'AS1gt69gqYD92dqPyE6DBRJ7KjpnQHqFzFs2YCkBcSnuxX5bGhBC',
    USDCs: 'AS12k8viVmqPtRuXzCm6rKXjLgpQWqbuMjc37YHhB452KSUUb9FgL',
    USDTbt: 'AS12ix1Qfpue7BB8q6mWVtjNdNE9UV3x4MaUo7WhdUubov8sJ3CuP',
    WETHbt: 'AS12RmCXTA9NZaTBUBnRJuH66AGNmtEfEoqXKxLdmrTybS6GFJPFs',
    WMAS: 'AS12FW5Rs5YN2zdpEnqwj4iHUUPt9R4Eqjq2qtpJFNKW3mn33RuLU',
};
function checkNetwork(provider, isMainnet) {
    provider.networkInfos().then((network) => {
        if (isMainnet && network.chainId !== utils_1.CHAIN_ID.Mainnet) {
            console.warn('This contract is only available on mainnet');
        }
        else if (!isMainnet && network.chainId === utils_1.CHAIN_ID.Mainnet) {
            console.warn('This contract is only available on buildnet');
        }
    });
}
exports.checkNetwork = checkNetwork;
///////////////// MAINNET TOKENS //////////////////////
class USDCe extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.USDCe);
        this.provider = provider;
    }
}
exports.USDCe = USDCe;
class USDTb extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.USDTb);
        this.provider = provider;
    }
}
exports.USDTb = USDTb;
class DAIe extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.DAIe);
        this.provider = provider;
    }
}
exports.DAIe = DAIe;
class WETHe extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.WETHe);
        this.provider = provider;
    }
}
exports.WETHe = WETHe;
class WETHb extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.WETHb);
        this.provider = provider;
    }
}
exports.WETHb = WETHb;
class PUR extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.PUR);
        this.provider = provider;
    }
}
exports.PUR = PUR;
class WMAS extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, true);
        super(provider, exports.MAINNET_TOKENS.WMAS);
        this.provider = provider;
    }
}
exports.WMAS = WMAS;
///////////////// BUILDNET TOKENS //////////////////////
class DAIs extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, false);
        super(provider, exports.BUILDNET_TOKENS.DAIs);
        this.provider = provider;
    }
}
exports.DAIs = DAIs;
class WETHs extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, false);
        super(provider, exports.BUILDNET_TOKENS.WETHs);
        this.provider = provider;
    }
}
exports.WETHs = WETHs;
class USDCs extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, false);
        super(provider, exports.BUILDNET_TOKENS.USDCs);
        this.provider = provider;
    }
}
exports.USDCs = USDCs;
class USDTbt extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, false);
        super(provider, exports.BUILDNET_TOKENS.USDTbt);
        this.provider = provider;
    }
}
exports.USDTbt = USDTbt;
class WETHbt extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, false);
        super(provider, exports.BUILDNET_TOKENS.WETHbt);
        this.provider = provider;
    }
}
exports.WETHbt = WETHbt;
class WMASBuildnet extends token_1.MRC20 {
    provider;
    constructor(provider) {
        checkNetwork(provider, false);
        super(provider, exports.BUILDNET_TOKENS.WMAS);
        this.provider = provider;
    }
}
exports.WMASBuildnet = WMASBuildnet;
