// @3land/listings-sdk/dist/types/types/ShortCreator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShortCreator = void 0;
const web3_js_1 = require("@solana/web3.js"); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class ShortCreator {
    constructor(fields) {
        this.address = fields.address;
        this.share = fields.share;
    }
    static layout(property) {
        return borsh.struct([borsh.publicKey("address"), borsh.u8("share")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ShortCreator({
            address: obj.address,
            share: obj.share,
        });
    }
    static toEncodable(fields) {
        return {
            address: fields.address,
            share: fields.share,
        };
    }
    toJSON() {
        return {
            address: this.address.toString(),
            share: this.share,
        };
    }
    static fromJSON(obj) {
        return new ShortCreator({
            address: new web3_js_1.PublicKey(obj.address),
            share: obj.share,
        });
    }
    toEncodable() {
        return ShortCreator.toEncodable(this);
    }
}
exports.ShortCreator = ShortCreator;
