// solana-agent-kit/dist/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FLUXBEAM_SWAP_PROGRAM_ID = exports.FLUXBEAM_BASE_URI = exports.FEE_ACCOUNT = exports.SWITCHBOARD_DEFAULT_CROSSBAR = exports.DEBRIDGE_API = exports.MINIMUM_COMPUTE_PRICE_FOR_COMPLEX_ACTIONS = exports.METEORA_DLMM_PROGRAM_ID = exports.METEORA_DYNAMIC_AMM_PROGRAM_ID = exports.JUP_REFERRAL_ADDRESS = exports.JUP_API = exports.METEORA_DYNAMIC_FEE_DENOMINATOR = exports.DEFAULT_OPTIONS = exports.TOKENS = void 0;
const web3_js_1 = require("@solana/web3.js");
/**
 * Common token addresses used across the toolkit
 */
exports.TOKENS = {
    USDC: new web3_js_1.PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"),
    USDT: new web3_js_1.PublicKey("Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"),
    USDS: new web3_js_1.PublicKey("USDSwr9ApdHk5bvJKMjzff41FfuX8bSxdKcR81vTwcA"),
    SOL: new web3_js_1.PublicKey("So11111111111111111111111111111111111111112"),
    jitoSOL: new web3_js_1.PublicKey("J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn"),
    bSOL: new web3_js_1.PublicKey("bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1"),
    mSOL: new web3_js_1.PublicKey("mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So"),
    BONK: new web3_js_1.PublicKey("DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263"),
};
/**
 * Default configuration options
 * @property {number} SLIPPAGE_BPS - Default slippage tolerance in basis points (300 = 3%)
 * @property {number} TOKEN_DECIMALS - Default number of decimals for new tokens
 * @property {number} LEVERAGE_BPS - Default leverage for trading PERP
 */
exports.DEFAULT_OPTIONS = {
    SLIPPAGE_BPS: 300,
    TOKEN_DECIMALS: 9,
    RERERRAL_FEE: 200,
    LEVERAGE_BPS: 50000, // 10000 = x1, 50000 = x5, 100000 = x10, 1000000 = x100
};
exports.METEORA_DYNAMIC_FEE_DENOMINATOR = 100000;
/**
 * Jupiter API URL
 */
exports.JUP_API = "https://quote-api.jup.ag/v6";
exports.JUP_REFERRAL_ADDRESS = "REFER4ZgmyYx9c6He5XfaTMiGfdLwRnkV4RPp9t9iF3";
exports.METEORA_DYNAMIC_AMM_PROGRAM_ID = new web3_js_1.PublicKey("Eo7WjKq67rjJQSZxS6z3YkapzY3eMj6Xy8X5EQVn5UaB");
exports.METEORA_DLMM_PROGRAM_ID = new web3_js_1.PublicKey("LbVRzDTvBDEcrthxfZ4RL6yiq3uZw8bS6MwtdY6UhFQ");
/**
 * Minimum compute price required to carry out complex transactions on the Drift protocol
 */
exports.MINIMUM_COMPUTE_PRICE_FOR_COMPLEX_ACTIONS = 0.000003 * 1000000 * 1000000;
/**
 * Debridge API URL
 */
exports.DEBRIDGE_API = "https://dln.debridge.finance/v1.0";
/**
 * Switchboard public crossbar instance.
 * https://docs.switchboard.xyz/docs/switchboard/crossbar-and-task-runner
 */
exports.SWITCHBOARD_DEFAULT_CROSSBAR = "https://crossbar.switchboard.xyz";
/**
 * Fluxbeam constants
 */
// fluxbeam fee account
exports.FEE_ACCOUNT = new web3_js_1.PublicKey("FLUXR4McuD2iXyP3wpP4XTjSWmB86ppMiyoA52UA9bKb");
exports.FLUXBEAM_BASE_URI = "https://api.fluxbeam.xyz/v1";
exports.FLUXBEAM_SWAP_PROGRAM_ID = new web3_js_1.PublicKey("FLUXubRmkEi2q6K3Y9kBPg9248ggaZVsoSFhtJHSrm1X");
