// @injectivelabs/utils/dist/cjs/classes/BigNumber/BigNumberInWei.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BigNumber_js_1 = __importDefault(require("./BigNumber.js"));
const utils_js_1 = require("./utils.js");
class BigNumberInWei extends BigNumber_js_1.default {
    static make(number) {
        return new BigNumberInWei(number);
    }
    minus(n, base) {
        return new BigNumberInWei(super.minus(n, base));
    }
    plus(n, base) {
        return new BigNumberInWei(super.plus(n, base));
    }
    dividedBy(n, base) {
        return new BigNumberInWei(super.dividedBy(n, base));
    }
    div(n, base) {
        return new BigNumberInWei(super.div(n, base));
    }
    multipliedBy(n, base) {
        return new BigNumberInWei(super.multipliedBy(n, base));
    }
    times(n, base) {
        return new BigNumberInWei(super.times(n, base));
    }
    pow(n, base) {
        return new BigNumberInWei(super.pow(n, base));
    }
    toBase(decimals = 18) {
        return (0, utils_js_1.bigNumberWeiToBase)(this, decimals);
    }
}
exports.default = BigNumberInWei;
