// quais/lib/esm/transaction/utxo.js
import { validateAddress } from '../address/index.js';
/**
 * List of supported Qi denominations.
 *
 * @category Transaction
 */
export const denominations = [
    BigInt(1),
    BigInt(5),
    BigInt(10),
    BigInt(50),
    BigInt(100),
    BigInt(500),
    BigInt(1000),
    BigInt(5000),
    BigInt(10000),
    BigInt(20000),
    BigInt(100000),
    BigInt(1000000),
    BigInt(10000000),
    BigInt(100000000),
    BigInt(1000000000), // 1,000,000 Qi (1,000,000,000 Qit)
];
/**
 * Checks if the provided denomination index is valid.
 *
 * @category Transaction
 * @param {number} index - The denomination index to check.
 * @returns {boolean} True if the denomination index is valid, false otherwise.
 */
function isValidDenominationIndex(index) {
    return index >= 0 && index < denominations.length;
}
/**
 * Given a value, returns an array of supported denominations that sum to the value.
 *
 * @category Transaction
 * @param {bigint} value - The value to denominate.
 * @returns {bigint[]} An array of denominations that sum to the value.
 * @throws {Error} If the value is less than or equal to 0 or cannot be matched with available denominations.
 */
export function denominate(value, maxDenomination) {
    if (value <= BigInt(0)) {
        throw new Error('Value must be greater than 0');
    }
    const result = [];
    let remainingValue = BigInt(value);
    // Find the index of the maximum allowed denomination
    let maxDenominationIndex;
    if (maxDenomination != null) {
        maxDenominationIndex = denominations.findIndex((d) => d === maxDenomination);
        if (maxDenominationIndex === -1) {
            throw new Error('Invalid maximum denomination');
        }
    }
    else {
        // No maximum denomination set, use the highest denomination
        maxDenominationIndex = denominations.length - 1;
    }
    // Iterate through denominations in descending order, up to the maximum allowed denomination
    for (let i = maxDenominationIndex; i >= 0; i--) {
        const denomination = denominations[i];
        // Add the denomination to the result array as many times as possible
        while (remainingValue >= denomination) {
            result.push(denomination);
            remainingValue -= BigInt(denomination);
        }
    }
    if (remainingValue > 0) {
        throw new Error('Unable to match the value with available denominations');
    }
    return result;
}
/**
 * Represents a UTXO (Unspent Transaction Output).
 *
 * @category Transaction
 * @implements {UTXOLike}
 */
export class UTXO {
    #txhash;
    #index;
    #address;
    #denomination;
    #lock;
    /**
     * Gets the transaction hash.
     *
     * @returns {null | string} The transaction hash.
     */
    get txhash() {
        return this.#txhash;
    }
    /**
     * Sets the transaction hash.
     *
     * @param {null | string} value - The transaction hash.
     */
    set txhash(value) {
        this.#txhash = value;
    }
    /**
     * Gets the index.
     *
     * @returns {null | number} The index.
     */
    get index() {
        return this.#index;
    }
    /**
     * Sets the index.
     *
     * @param {null | number} value - The index.
     */
    set index(value) {
        this.#index = value;
    }
    /**
     * Gets the address.
     *
     * @returns {string} The address.
     */
    get address() {
        return this.#address || '';
    }
    /**
     * Sets the address.
     *
     * @param {string} value - The address.
     * @throws {Error} If the address is invalid.
     */
    set address(value) {
        validateAddress(value);
        this.#address = value;
    }
    /**
     * Gets the denomination.
     *
     * @returns {null | number} The denomination.
     */
    get denomination() {
        return this.#denomination;
    }
    /**
     * Sets the denomination.
     *
     * @param {null | number} value - The denomination.
     * @throws {Error} If the denomination value is invalid.
     */
    set denomination(value) {
        if (value == null) {
            this.#denomination = null;
            return;
        }
        if (!isValidDenominationIndex(value)) {
            throw new Error('Invalid denomination value');
        }
        this.#denomination = value;
    }
    get lock() {
        return this.#lock;
    }
    set lock(value) {
        this.#lock = value;
    }
    /**
     * Constructs a new UTXO instance with null properties.
     */
    constructor() {
        this.#txhash = null;
        this.#index = null;
        this.#address = null;
        this.#denomination = null;
        this.#lock = null;
    }
    /**
     * Converts the UTXO instance to a JSON object.
     *
     * @returns {any} A JSON representation of the UTXO instance.
     */
    toJSON() {
        return {
            txhash: this.txhash,
            index: this.index,
            address: this.address,
            denomination: this.denomination,
        };
    }
    /**
     * Creates a UTXO instance from a UTXOLike object.
     *
     * @param {UTXOLike} utxo - The UTXOLike object to convert.
     * @returns {UTXO} The UTXO instance.
     */
    static from(utxo) {
        if (utxo === null) {
            return new UTXO();
        }
        const result = utxo instanceof UTXO ? utxo : new UTXO();
        if (utxo.txhash != null) {
            result.txhash = utxo.txhash;
        }
        if (utxo.index != null) {
            result.index = utxo.index;
        }
        if (utxo.address != null && utxo.address !== '') {
            result.address = utxo.address;
        }
        if (utxo.denomination != null) {
            result.denomination = utxo.denomination;
        }
        return result;
    }
}
