// @lifi/data-types/src/_cjs/chains/supportedChains.utxo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedUXTOChains = void 0;
const types_1 = require("@lifi/types");
exports.supportedUXTOChains = [
    {
        key: types_1.ChainKey.BTC,
        chainType: types_1.ChainType.UTXO,
        name: 'Bitcoin',
        coin: types_1.CoinKey.BTC,
        id: types_1.ChainId.BTC,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/bitcoin.svg',
        faucetUrls: [],
        metamask: {
            chainId: types_1.ChainId.BTC.toString(),
            blockExplorerUrls: [
                'https://mempool.space/',
                'https://blockchair.com/bitcoin/',
            ],
            chainName: 'Bitcoin',
            nativeCurrency: {
                name: 'BTC',
                symbol: 'BTC',
                decimals: 8,
            },
            rpcUrls: [
                'https://node-router.thorswap.net/bitcoin',
                'https://rpc.ankr.com/btc',
                'https://bitcoin-rpc.publicnode.com',
            ],
        },
    },
    {
        key: types_1.ChainKey.BCH,
        chainType: types_1.ChainType.UTXO,
        name: 'Bitcoin Cash',
        coin: types_1.CoinKey.BCH,
        id: types_1.ChainId.BCH,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/bitcoincash.svg',
        faucetUrls: [],
        metamask: {
            chainId: types_1.ChainId.BCH.toString(),
            blockExplorerUrls: ['https://blockchair.com/bitcoin-cash/'],
            chainName: 'Bitcoin Cash',
            nativeCurrency: {
                name: 'BCH',
                symbol: 'BCH',
                decimals: 8,
            },
            rpcUrls: ['https://node-router.thorswap.net/bitcoin-cash'],
        },
    },
    {
        key: types_1.ChainKey.LTC,
        chainType: types_1.ChainType.UTXO,
        name: 'Litecoin',
        coin: types_1.CoinKey.LTC,
        id: types_1.ChainId.LTC,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/litecoin.svg',
        faucetUrls: [],
        metamask: {
            chainId: types_1.ChainId.LTC.toString(),
            blockExplorerUrls: ['https://blockchair.com/litecoin/'],
            chainName: 'Litecoin',
            nativeCurrency: {
                name: 'LTC',
                symbol: 'LTC',
                decimals: 8,
            },
            rpcUrls: ['https://node-router.thorswap.net/litecoin'],
        },
    },
    {
        key: types_1.ChainKey.DGE,
        chainType: types_1.ChainType.UTXO,
        name: 'Dogecoin',
        coin: types_1.CoinKey.DOGE,
        id: types_1.ChainId.DGE,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/dogecoin.svg',
        faucetUrls: [],
        metamask: {
            chainId: types_1.ChainId.DGE.toString(),
            blockExplorerUrls: ['https://blockchair.com/dogecoin/'],
            chainName: 'Dogecoin',
            nativeCurrency: {
                name: 'DODGE',
                symbol: 'DOGE',
                decimals: 8,
            },
            rpcUrls: ['https://node-router.thorswap.net/dogecoin'],
        },
    },
];
