// @lifi/data-types/src/_cjs/chains/supportedChains.evm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedEVMChains = void 0;
const types_1 = require("@lifi/types");
const multicall_js_1 = require("../multicall.js");
const utils_js_1 = require("./utils.js");
exports.supportedEVMChains = [
    {
        key: types_1.ChainKey.ETH,
        chainType: types_1.ChainType.EVM,
        name: 'Ethereum',
        coin: types_1.CoinKey.ETH,
        id: 1,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/ethereum.svg',
        tokenlistUrl: 'https://gateway.ipfs.io/ipns/tokens.uniswap.org',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.ETH],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1),
            blockExplorerUrls: ['https://etherscan.io/'],
            chainName: 'Ethereum Mainnet',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://ethereum-rpc.publicnode.com',
                'https://eth.drpc.org',
                'https://eth.public-rpc.com',
                'https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161',
                'https://rpc.ankr.com/eth',
            ],
        },
    },
    {
        key: types_1.ChainKey.POL,
        chainType: types_1.ChainType.EVM,
        name: 'Polygon',
        coin: types_1.CoinKey.POL,
        id: types_1.ChainId.POL,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/polygon.svg',
        tokenlistUrl: 'https://unpkg.com/quickswap-default-token-list@1.0.71/build/quickswap-default.tokenlist.json',
        faucetUrls: ['https://stakely.io/faucet/polygon-matic'],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.POL],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(137),
            blockExplorerUrls: ['https://polygonscan.com/'],
            chainName: 'Polygon Mainnet',
            nativeCurrency: {
                name: 'Polygon Ecosystem Token',
                symbol: 'POL',
                decimals: 18,
            },
            rpcUrls: [
                'https://polygon-bor-rpc.publicnode.com',
                'https://polygon.drpc.org',
                'https://polygon-rpc.com',
                'https://rpc.ankr.com/polygon',
            ],
        },
    },
    {
        key: types_1.ChainKey.BSC,
        chainType: types_1.ChainType.EVM,
        name: 'BSC',
        coin: types_1.CoinKey.BNB,
        id: 56,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/bsc.svg',
        tokenlistUrl: 'https://tokens.pancakeswap.finance/pancakeswap-extended.json',
        faucetUrls: ['https://stakely.io/faucet/bsc-chain-bnb'],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.BSC],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(56),
            blockExplorerUrls: ['https://bscscan.com/'],
            chainName: 'BNB Smart Chain Mainnet',
            nativeCurrency: {
                name: 'BNB',
                symbol: 'BNB',
                decimals: 18,
            },
            rpcUrls: [
                'https://bsc-dataseed.binance.org',
                'https://bsc-dataseed.bnbchain.org',
                'https://bsc-rpc.publicnode.com',
                'https://bsc-dataseed1.defibit.io',
                'https://bsc-dataseed1.ninicoin.io',
            ],
        },
    },
    {
        key: types_1.ChainKey.DAI,
        chainType: types_1.ChainType.EVM,
        name: 'Gnosis',
        coin: types_1.CoinKey.DAI,
        id: 100,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/gnosis.svg',
        tokenlistUrl: 'https://tokens.honeyswap.org/',
        faucetUrls: ['https://stakely.io/faucet/xdai-chain'],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.DAI],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(100),
            blockExplorerUrls: ['https://gnosis.blockscout.com/'],
            chainName: 'Gnosis',
            nativeCurrency: {
                name: 'XDAI',
                symbol: 'XDAI',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.gnosischain.com',
                'https://gnosis-rpc.publicnode.com',
                'https://gnosis.drpc.org',
                'https://rpc.ankr.com/gnosis',
            ],
        },
    },
    {
        key: types_1.ChainKey.FTM,
        chainType: types_1.ChainType.EVM,
        name: 'Fantom',
        coin: types_1.CoinKey.FTM,
        id: 250,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/fantom.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/SpookySwap/spooky-info/master/src/constants/token/spookyswap.json',
        faucetUrls: [
            'https://stakely.io/faucet/fantom-blockchain-ftm',
            'https://docs.spookyswap.finance/getting-started/how-to-get-fantom-gas',
        ],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.FTM],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(250),
            blockExplorerUrls: ['https://ftmscan.com/'],
            chainName: 'Fantom Opera',
            nativeCurrency: {
                name: 'FTM',
                symbol: 'FTM',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpcapi.fantom.network',
                'https://rpc.fantom.network',
                'https://fantom-rpc.publicnode.com',
                'https://fantom.drpc.org',
                'https://rpc.ftm.tools',
            ],
        },
    },
    {
        key: types_1.ChainKey.AVA,
        chainType: types_1.ChainType.EVM,
        name: 'Avalanche',
        coin: types_1.CoinKey.AVAX,
        id: 43114,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/avalanche.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/sushiswap/default-token-list/master/tokens/avalanche.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.AVA],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(43114),
            blockExplorerUrls: ['https://snowtrace.io/'],
            chainName: 'Avalanche C-Chain',
            nativeCurrency: {
                name: 'AVAX',
                symbol: 'AVAX',
                decimals: 18,
            },
            rpcUrls: [
                'https://api.avax.network/ext/bc/C/rpc',
                'https://avalanche-c-chain-rpc.publicnode.com',
                'https://avalanche.drpc.org',
                'https://avalanche.public-rpc.com',
                'https://rpc.ankr.com/avalanche',
            ],
        },
    },
    {
        key: types_1.ChainKey.ARB,
        chainType: types_1.ChainType.EVM,
        name: 'Arbitrum',
        coin: types_1.CoinKey.ETH,
        id: 42161,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/arbitrum.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/sushiswap/default-token-list/master/tokens/arbitrum.json',
        faucetUrls: ['https://bridge.arbitrum.io/'],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.ARB],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(42161),
            blockExplorerUrls: ['https://arbiscan.io/'],
            chainName: 'Arbitrum One',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://arb1.arbitrum.io/rpc',
                'https://arbitrum-one-rpc.publicnode.com',
                'https://arbitrum.drpc.org',
                'https://rpc.ankr.com/arbitrum',
            ],
        },
    },
    {
        key: types_1.ChainKey.OPT,
        chainType: types_1.ChainType.EVM,
        name: 'Optimism',
        coin: types_1.CoinKey.ETH,
        id: 10,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/optimism.svg',
        tokenlistUrl: 'https://static.optimism.io/optimism.tokenlist.json',
        faucetUrls: ['https://gateway.optimism.io/'],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.OPT],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(10),
            blockExplorerUrls: ['https://optimistic.etherscan.io/'],
            chainName: 'OP Mainnet',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://mainnet.optimism.io/',
                'https://optimism-rpc.publicnode.com',
                'https://optimism.drpc.org',
                'https://rpc.ankr.com/optimism',
            ],
        },
    },
    {
        key: types_1.ChainKey.FSN,
        chainType: types_1.ChainType.EVM,
        name: 'Fusion Mainnet',
        coin: types_1.CoinKey.FSN,
        id: 32659,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/fusion.png',
        mainnet: true,
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.FSN],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(32659),
            blockExplorerUrls: ['https://fsnscan.com/'],
            chainName: 'Fusion Mainnet',
            nativeCurrency: {
                name: 'FSN',
                symbol: 'FSN',
                decimals: 18,
            },
            rpcUrls: ['https://mainnet.fusionnetwork.io'],
        },
    },
    {
        key: types_1.ChainKey.ONE,
        chainType: types_1.ChainType.EVM,
        name: 'Harmony',
        coin: types_1.CoinKey.ONE,
        id: 1666600000,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/harmony.svg',
        tokenlistUrl: 'https://d1xrz6ki9z98vb.cloudfront.net/venomswap/lists/venomswap-default.tokenlist.json',
        faucetUrls: ['https://stakely.io/faucet/harmony-one'],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.ONE],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1666600000),
            blockExplorerUrls: ['https://explorer.harmony.one/'],
            chainName: 'Harmony Mainnet Shard 0',
            nativeCurrency: {
                name: 'ONE',
                symbol: 'ONE',
                decimals: 18,
            },
            rpcUrls: ['https://api.harmony.one'],
        },
    },
    {
        key: types_1.ChainKey.MOR,
        chainType: types_1.ChainType.EVM,
        name: 'Moonriver',
        coin: types_1.CoinKey.MOVR,
        id: 1285,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/moonriver.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/sushiswap/default-token-list/master/tokens/moonriver.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.MOR],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1285),
            blockExplorerUrls: ['https://moonriver.moonscan.io/'],
            chainName: 'Moonriver',
            nativeCurrency: {
                name: 'Moonriver',
                symbol: 'MOVR',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.api.moonriver.moonbeam.network',
                'https://moonriver-rpc.publicnode.com',
                'https://moonriver.drpc.org',
            ],
        },
    },
    {
        key: types_1.ChainKey.MOO,
        chainType: types_1.ChainType.EVM,
        name: 'Moonbeam',
        coin: types_1.CoinKey.GLMR,
        id: 1284,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/moonbeam.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/BeamSwap/beamswap-tokenlist/main/tokenlist.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.MOO],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1284),
            blockExplorerUrls: ['https://moonbeam.moonscan.io/'],
            chainName: 'Moonbeam',
            nativeCurrency: {
                name: 'GLMR',
                symbol: 'GLMR',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.api.moonbeam.network',
                'https://moonbeam-rpc.publicnode.com',
                'https://moonbeam.drpc.org',
            ],
        },
    },
    {
        key: types_1.ChainKey.CEL,
        chainType: types_1.ChainType.EVM,
        name: 'Celo',
        coin: types_1.CoinKey.CELO,
        id: 42220,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/celo.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/sushiswap/default-token-list/master/tokens/celo.json',
        faucetUrls: [
            'https://stakely.io/faucet/celo-platform',
            'https://free-online-app.com/faucet-for-eth-evm-chains/',
        ],
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.CEL],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(42220),
            blockExplorerUrls: ['https://celoscan.io/'],
            chainName: 'Celo Mainnet',
            nativeCurrency: {
                name: 'CELO',
                symbol: 'CELO',
                decimals: 18,
            },
            rpcUrls: [
                'https://forno.celo.org',
                'https://celo.drpc.org',
                'https://rpc.ankr.com/celo',
            ],
        },
    },
    {
        key: types_1.ChainKey.FUS,
        chainType: types_1.ChainType.EVM,
        name: 'FUSE',
        coin: types_1.CoinKey.FUSE,
        id: 122,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/fuse.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/sushiswap/default-token-list/master/tokens/fuse.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.FUS],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(122),
            blockExplorerUrls: ['https://explorer.fuse.io/'],
            chainName: 'Fuse Mainnet',
            nativeCurrency: {
                name: 'FUSE',
                symbol: 'FUSE',
                decimals: 18,
            },
            rpcUrls: ['https://rpc.fuse.io', 'https://fuse.drpc.org'],
        },
    },
    {
        key: types_1.ChainKey.TLO,
        chainType: types_1.ChainType.EVM,
        name: 'Telos',
        coin: types_1.CoinKey.TLOS,
        id: 40,
        mainnet: true,
        tokenlistUrl: 'https://raw.githubusercontent.com/sushiswap/default-token-list/master/tokens/telos.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.TLO],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(40),
            blockExplorerUrls: ['https://www.telos.net/'],
            chainName: 'Telos EVM Mainnet',
            nativeCurrency: {
                name: 'Telos',
                symbol: 'TLOS',
                decimals: 18,
            },
            rpcUrls: ['https://mainnet.telos.net/evm'],
        },
    },
    {
        key: types_1.ChainKey.CRO,
        chainType: types_1.ChainType.EVM,
        name: 'Cronos',
        coin: types_1.CoinKey.CRO,
        id: 25,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/cronos.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/cronaswap/default-token-list/refs/heads/main/assets/tokens/cronos.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.CRO],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(25),
            blockExplorerUrls: [
                'https://explorer.cronos.org/',
                'https://cronoscan.com/',
            ],
            chainName: 'Cronos Mainnet',
            nativeCurrency: {
                name: 'Cronos',
                symbol: 'CRO',
                decimals: 18,
            },
            rpcUrls: ['https://evm.cronos.org', 'https://cronos.drpc.org'],
        },
    },
    {
        key: types_1.ChainKey.BOB,
        chainType: types_1.ChainType.EVM,
        name: 'Boba',
        coin: types_1.CoinKey.ETH,
        id: 288,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/boba.png',
        tokenlistUrl: 'https://raw.githubusercontent.com/OolongSwap/boba-community-token-list/main/build/boba.tokenlist.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.BOB],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(288),
            blockExplorerUrls: ['https://bobascan.com/'],
            chainName: 'Boba Network',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: ['https://mainnet.boba.network', 'https://replica.boba.network'],
        },
    },
    {
        key: types_1.ChainKey.VEL,
        chainType: types_1.ChainType.EVM,
        name: 'Velas',
        coin: types_1.CoinKey.VLX,
        id: 106,
        mainnet: true,
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.VEL],
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/velas.png',
        faucetUrls: ['https://stakely.io/faucet/velas-vlx'],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(106),
            blockExplorerUrls: ['https://evmexplorer.velas.com/'],
            chainName: 'Velas EVM Mainnet',
            nativeCurrency: {
                name: 'VLX',
                symbol: 'VLX',
                decimals: 18,
            },
            rpcUrls: ['https://evmexplorer.velas.com/rpc'],
        },
    },
    {
        key: types_1.ChainKey.MAM,
        chainType: types_1.ChainType.EVM,
        name: 'Metis',
        coin: types_1.CoinKey.METIS,
        id: 1088,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/metis.svg',
        tokenlistUrl: 'https://raw.githubusercontent.com/digitalnativeinc/default-token-list/dnf/tokens/metis.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.MAM],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1088),
            blockExplorerUrls: ['https://andromeda-explorer.metis.io/'],
            chainName: 'Metis Andromeda Mainnet',
            nativeCurrency: {
                name: 'METIS',
                symbol: 'METIS',
                decimals: 18,
            },
            rpcUrls: ['https://andromeda.metis.io', 'https://metis.drpc.org'],
        },
    },
    {
        key: types_1.ChainKey.AUR,
        chainType: types_1.ChainType.EVM,
        name: 'Aurora',
        coin: types_1.CoinKey.ETH,
        id: 1313161554,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/aurora.png',
        tokenlistUrl: 'https://aurora.dev/tokens.json',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.AUR],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1313161554),
            blockExplorerUrls: ['https://explorer.aurora.dev/'],
            chainName: 'Aurora Mainnet',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: ['https://mainnet.aurora.dev'],
        },
    },
    {
        key: types_1.ChainKey.EVM,
        chainType: types_1.ChainType.EVM,
        name: 'Evmos',
        coin: types_1.CoinKey.EVM,
        id: 9001,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/evmos.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.EVM],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(9001),
            blockExplorerUrls: ['https://www.mintscan.io/evmos/'],
            chainName: 'Evmos',
            nativeCurrency: {
                name: 'EVMOS',
                symbol: 'EVMOS',
                decimals: 18,
            },
            rpcUrls: [
                'https://evmos.lava.build',
                'https://evmos-evm-rpc.publicnode.com',
                'https://evmos.drpc.org',
            ],
        },
    },
    {
        key: types_1.ChainKey.ERA,
        chainType: types_1.ChainType.EVM,
        name: 'zkSync',
        coin: types_1.CoinKey.ETH,
        id: 324,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/zksync.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.ERA],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(324),
            blockExplorerUrls: ['https://explorer.zksync.io/'],
            chainName: 'zkSync Mainnet',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: ['https://mainnet.era.zksync.io'],
        },
    },
    {
        key: types_1.ChainKey.PZE,
        chainType: types_1.ChainType.EVM,
        name: 'Polygon zkEVM',
        coin: types_1.CoinKey.ETH,
        id: 1101,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/zkevm.png',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.PZE],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1101),
            blockExplorerUrls: ['https://zkevm.polygonscan.com/'],
            chainName: 'Polygon zkEVM',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: ['https://zkevm-rpc.com', 'https://polygon-zkevm.drpc.org'],
        },
    },
    {
        key: types_1.ChainKey.BAS,
        chainType: types_1.ChainType.EVM,
        name: 'Base',
        coin: types_1.CoinKey.ETH,
        id: types_1.ChainId.BAS,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/base.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.BAS],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(types_1.ChainId.BAS),
            blockExplorerUrls: ['https://basescan.org/'],
            chainName: 'Base',
            nativeCurrency: {
                name: 'Ethereum',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://mainnet.base.org',
                'https://base-rpc.publicnode.com',
                'https://base.drpc.org',
                'https://rpc.ankr.com/base',
            ],
        },
    },
    {
        key: types_1.ChainKey.LNA,
        chainType: types_1.ChainType.EVM,
        name: 'Linea',
        coin: types_1.CoinKey.ETH,
        id: 59144,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/linea.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.LNA],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(59144),
            blockExplorerUrls: ['https://lineascan.build/'],
            chainName: 'Linea',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: ['https://rpc.linea.build'],
        },
    },
    {
        key: types_1.ChainKey.SCL,
        chainType: types_1.ChainType.EVM,
        name: 'Scroll',
        coin: types_1.CoinKey.ETH,
        id: 534352,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/scroll.png',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.SCL],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(534352),
            blockExplorerUrls: ['https://scrollscan.com/'],
            chainName: 'Scroll',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.scroll.io',
                'https://scroll.drpc.org',
                'https://scroll-mainnet.public.blastapi.io',
                'https://1rpc.io/scroll',
            ],
        },
    },
    {
        key: types_1.ChainKey.MOD,
        chainType: types_1.ChainType.EVM,
        name: 'Mode',
        coin: types_1.CoinKey.ETH,
        id: 34443,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/mode.png',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.MOD],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(34443),
            blockExplorerUrls: [
                'https://explorer.mode.network/',
                'https://modescan.io/',
            ],
            chainName: 'Mode',
            nativeCurrency: {
                name: 'ETH',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: ['https://mainnet.mode.network', 'https://mode.drpc.org'],
        },
    },
    {
        key: types_1.ChainKey.MNT,
        chainType: types_1.ChainType.EVM,
        name: 'Mantle',
        coin: types_1.CoinKey.MNT,
        id: 5000,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/mantle.png',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.MNT],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(5000),
            blockExplorerUrls: [
                'https://explorer.mantle.xyz/',
                'https://mantlescan.xyz/',
            ],
            chainName: 'Mantle',
            nativeCurrency: {
                name: 'Mantle',
                symbol: 'MNT',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.mantle.xyz',
                'https://mantle-rpc.publicnode.com',
                'https://mantle.drpc.org',
                'https://mantle.public-rpc.com',
                'https://rpc.ankr.com/mantle',
            ],
        },
    },
    {
        key: types_1.ChainKey.BLS,
        chainType: types_1.ChainType.EVM,
        name: 'Blast',
        coin: types_1.CoinKey.ETH,
        id: 81457,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/blast.png',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.BLS],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(81457),
            blockExplorerUrls: ['https://blastscan.io/', 'https://blastexplorer.io/'],
            chainName: 'Blast',
            nativeCurrency: {
                name: 'Ethereum',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.blast.io',
                'https://blast-rpc.publicnode.com',
                'https://blast.drpc.org',
                'https://rpc.ankr.com/blast',
            ],
        },
    },
    {
        key: types_1.ChainKey.RSK,
        chainType: types_1.ChainType.EVM,
        name: 'Rootstock',
        coin: types_1.CoinKey.RBTC,
        id: 30,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/rootstock.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.RSK],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(30),
            blockExplorerUrls: ['https://explorer.rootstock.io/'],
            chainName: 'Rootstock Mainnet',
            nativeCurrency: {
                name: 'Rootstock Smart Bitcoin',
                symbol: 'RBTC',
                decimals: 18,
            },
            rpcUrls: ['https://public-node.rsk.co', 'https://mycrypto.rsk.co'],
        },
    },
    {
        key: types_1.ChainKey.SEI,
        chainType: types_1.ChainType.EVM,
        name: 'Sei',
        coin: types_1.CoinKey.SEI,
        id: 1329,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/sei.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.SEI],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1329),
            blockExplorerUrls: ['https://seitrace.com/', 'https://seistream.app/'],
            chainName: 'Sei Network',
            nativeCurrency: {
                name: 'SEI',
                symbol: 'SEI',
                decimals: 18,
            },
            rpcUrls: ['https://evm-rpc.sei-apis.com'],
        },
    },
    {
        key: types_1.ChainKey.FRA,
        chainType: types_1.ChainType.EVM,
        name: 'Fraxtal',
        coin: types_1.CoinKey.frxETH,
        id: 252,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/fraxtal.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.FRA],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(252),
            blockExplorerUrls: ['https://fraxscan.com/'],
            chainName: 'Fraxtal',
            nativeCurrency: {
                name: 'Frax Ethereum',
                symbol: 'frxETH',
                decimals: 18,
            },
            rpcUrls: ['https://rpc.frax.com'],
        },
    },
    {
        key: types_1.ChainKey.TAI,
        chainType: types_1.ChainType.EVM,
        name: 'Taiko',
        coin: types_1.CoinKey.ETH,
        id: 167000,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/taiko.png',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.TAI],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(167000),
            blockExplorerUrls: ['https://taikoscan.io/'],
            chainName: 'Taiko Mainnet',
            nativeCurrency: {
                name: 'Ethereum',
                symbol: 'ETH',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.mainnet.taiko.xyz',
                'https://rpc.taiko.xyz',
                'https://rpc.taiko.tools',
            ],
        },
    },
    {
        key: types_1.ChainKey.GRA,
        chainType: types_1.ChainType.EVM,
        name: 'Gravity',
        coin: types_1.CoinKey.G,
        id: 1625,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/gravity.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.GRA],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(1625),
            blockExplorerUrls: ['https://explorer.gravity.xyz/'],
            chainName: 'Gravity Alpha Mainnet',
            nativeCurrency: {
                name: 'G',
                symbol: 'G',
                decimals: 18,
            },
            rpcUrls: ['https://rpc.gravity.xyz/'],
        },
    },
    {
        key: types_1.ChainKey.IMX,
        chainType: types_1.ChainType.EVM,
        name: 'Immutable zkEVM',
        coin: types_1.CoinKey.IMX,
        id: 13371,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/imx.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.IMX],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(13371),
            blockExplorerUrls: [
                'https://immutable-mainnet.blockscout.com/',
                'https://explorer.immutable.com/',
            ],
            chainName: 'Immutable',
            nativeCurrency: {
                name: 'ImmutableX',
                symbol: 'IMX',
                decimals: 18,
            },
            rpcUrls: [
                'https://rpc.immutable.com/',
                'https://immutable-zkevm.drpc.org',
            ],
        },
    },
    {
        key: types_1.ChainKey.KAI,
        chainType: types_1.ChainType.EVM,
        name: 'Kaia',
        coin: types_1.CoinKey.KLAY,
        id: 8217,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/kaia.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.KAI],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(8217),
            blockExplorerUrls: ['https://kaiascan.io/', 'https://kaiascope.com/'],
            chainName: 'Kaia Mainnet',
            nativeCurrency: {
                name: 'KAIA',
                symbol: 'KAIA',
                decimals: 18,
            },
            rpcUrls: [
                'https://public-en.node.kaia.io',
                'https://klaytn-pokt.nodies.app',
            ],
        },
    },
    {
        key: types_1.ChainKey.XLY,
        chainType: types_1.ChainType.EVM,
        name: 'XLayer',
        coin: types_1.CoinKey.OKB,
        id: 196,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/xlayer.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.XLY],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(196),
            blockExplorerUrls: [
                'https://www.oklink.com/xlayer/',
                'https://www.okx.com/web3/explorer/xlayer/',
            ],
            chainName: 'X Layer Mainnet',
            nativeCurrency: {
                name: 'OKB',
                symbol: 'OKB',
                decimals: 18,
            },
            rpcUrls: ['https://rpc.xlayer.tech/', 'https://xlayerrpc.okx.com'],
        },
    },
    {
        key: types_1.ChainKey.OPB,
        chainType: types_1.ChainType.EVM,
        name: 'opBNB',
        coin: types_1.CoinKey.BNB,
        id: 204,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/opbnb.svg',
        multicallAddress: multicall_js_1.multicallAddresses[types_1.ChainId.OPB],
        metamask: {
            chainId: (0, utils_js_1.prefixChainId)(204),
            blockExplorerUrls: [
                'https://opbnb.bscscan.com/',
                'https://opbnbscan.com/',
            ],
            chainName: 'opBNB Mainnet',
            nativeCurrency: {
                name: 'BNB',
                symbol: 'BNB',
                decimals: 18,
            },
            rpcUrls: [
                'https://opbnb-mainnet-rpc.bnbchain.org',
                'https://1rpc.io/opbnb',
            ],
        },
    },
];
