// @lifi/data-types/src/_cjs/chains/supportedChains.svm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportedSolanaChains = void 0;
const types_1 = require("@lifi/types");
exports.supportedSolanaChains = [
    {
        key: types_1.ChainKey.SOL,
        chainType: types_1.ChainType.SVM,
        name: 'Solana',
        coin: types_1.CoinKey.SOL,
        id: types_1.ChainId.SOL,
        mainnet: true,
        logoURI: 'https://raw.githubusercontent.com/lifinance/types/main/src/assets/icons/chains/solana.svg',
        faucetUrls: ['https://stakely.io/faucet/solana-sol'],
        metamask: {
            chainId: types_1.ChainId.SOL.toString(),
            blockExplorerUrls: [
                'https://explorer.solana.com/',
                'https://solscan.io/',
                'https://solana.fm/',
            ],
            chainName: 'Solana',
            nativeCurrency: {
                name: 'SOL',
                symbol: 'SOL',
                decimals: 9,
            },
            rpcUrls: [
                'https://api.mainnet-beta.solana.com',
                'https://solana-rpc.publicnode.com',
            ],
        },
    },
];
