// quais/lib/commonjs/address/index.js
"use strict";
/**
 * Addresses are a fundamental part of interacting with Ethereum. They represent the gloabal identity of Externally
 * Owned Accounts (accounts backed by a private key) and contracts.
 *
 * These functions help convert between various formats and validate addresses
 *
 * @category Address
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isQiAddress = exports.isQuaiAddress = exports.validateAddress = exports.resolveAddress = exports.isAddress = exports.isAddressable = exports.getCreate2Address = exports.getCreateAddress = exports.getContractAddress = exports.formatMixedCaseChecksumAddress = exports.recoverAddress = exports.computeAddress = exports.getAddress = void 0;
null;
var address_js_1 = require("./address.js");
Object.defineProperty(exports, "getAddress", { enumerable: true, get: function () { return address_js_1.getAddress; } });
Object.defineProperty(exports, "computeAddress", { enumerable: true, get: function () { return address_js_1.computeAddress; } });
Object.defineProperty(exports, "recoverAddress", { enumerable: true, get: function () { return address_js_1.recoverAddress; } });
Object.defineProperty(exports, "formatMixedCaseChecksumAddress", { enumerable: true, get: function () { return address_js_1.formatMixedCaseChecksumAddress; } });
Object.defineProperty(exports, "getContractAddress", { enumerable: true, get: function () { return address_js_1.getContractAddress; } });
var contract_address_js_1 = require("./contract-address.js");
Object.defineProperty(exports, "getCreateAddress", { enumerable: true, get: function () { return contract_address_js_1.getCreateAddress; } });
Object.defineProperty(exports, "getCreate2Address", { enumerable: true, get: function () { return contract_address_js_1.getCreate2Address; } });
var checks_js_1 = require("./checks.js");
Object.defineProperty(exports, "isAddressable", { enumerable: true, get: function () { return checks_js_1.isAddressable; } });
Object.defineProperty(exports, "isAddress", { enumerable: true, get: function () { return checks_js_1.isAddress; } });
Object.defineProperty(exports, "resolveAddress", { enumerable: true, get: function () { return checks_js_1.resolveAddress; } });
Object.defineProperty(exports, "validateAddress", { enumerable: true, get: function () { return checks_js_1.validateAddress; } });
Object.defineProperty(exports, "isQuaiAddress", { enumerable: true, get: function () { return checks_js_1.isQuaiAddress; } });
Object.defineProperty(exports, "isQiAddress", { enumerable: true, get: function () { return checks_js_1.isQiAddress; } });
