// @injectivelabs/sdk-ts/dist/cjs/core/modules/exchange/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgInstantBinaryOptionsMarketLaunch = exports.MsgBatchCancelBinaryOptionsOrders = exports.MsgCreateBinaryOptionsMarketOrder = exports.MsgAdminUpdateBinaryOptionsMarket = exports.MsgCreateBinaryOptionsLimitOrder = exports.MsgBatchCancelDerivativeOrders = exports.MsgCreateDerivativeMarketOrder = exports.MsgCreateDerivativeLimitOrder = exports.MsgCancelBinaryOptionsOrder = exports.MsgInstantSpotMarketLaunch = exports.MsgIncreasePositionMargin = exports.MsgCreateSpotMarketOrder = exports.MsgBatchCancelSpotOrders = exports.MsgCancelDerivativeOrder = exports.MsgCreateSpotLimitOrder = exports.MsgAuthorizeStakeGrants = exports.MsgReclaimLockedFunds = exports.MsgBatchUpdateOrders = exports.MsgLiquidatePosition = exports.MsgExternalTransfer = exports.MsgCancelSpotOrder = exports.MsgRewardsOptOut = exports.MsgSignData = exports.MsgWithdraw = exports.MsgDeposit = void 0;
const MsgDeposit_js_1 = __importDefault(require("./msgs/MsgDeposit.js"));
exports.MsgDeposit = MsgDeposit_js_1.default;
const MsgSignData_js_1 = __importDefault(require("./msgs/MsgSignData.js"));
exports.MsgSignData = MsgSignData_js_1.default;
const MsgWithdraw_js_1 = __importDefault(require("./msgs/MsgWithdraw.js"));
exports.MsgWithdraw = MsgWithdraw_js_1.default;
const MsgRewardsOptOut_js_1 = __importDefault(require("./msgs/MsgRewardsOptOut.js"));
exports.MsgRewardsOptOut = MsgRewardsOptOut_js_1.default;
const MsgCancelSpotOrder_js_1 = __importDefault(require("./msgs/MsgCancelSpotOrder.js"));
exports.MsgCancelSpotOrder = MsgCancelSpotOrder_js_1.default;
const MsgExternalTransfer_js_1 = __importDefault(require("./msgs/MsgExternalTransfer.js"));
exports.MsgExternalTransfer = MsgExternalTransfer_js_1.default;
const MsgLiquidatePosition_js_1 = __importDefault(require("./msgs/MsgLiquidatePosition.js"));
exports.MsgLiquidatePosition = MsgLiquidatePosition_js_1.default;
const MsgBatchUpdateOrders_js_1 = __importDefault(require("./msgs/MsgBatchUpdateOrders.js"));
exports.MsgBatchUpdateOrders = MsgBatchUpdateOrders_js_1.default;
const MsgReclaimLockedFunds_js_1 = __importDefault(require("./msgs/MsgReclaimLockedFunds.js"));
exports.MsgReclaimLockedFunds = MsgReclaimLockedFunds_js_1.default;
const MsgCreateSpotLimitOrder_js_1 = __importDefault(require("./msgs/MsgCreateSpotLimitOrder.js"));
exports.MsgCreateSpotLimitOrder = MsgCreateSpotLimitOrder_js_1.default;
const MsgBatchCancelSpotOrders_js_1 = __importDefault(require("./msgs/MsgBatchCancelSpotOrders.js"));
exports.MsgBatchCancelSpotOrders = MsgBatchCancelSpotOrders_js_1.default;
const MsgCancelDerivativeOrder_js_1 = __importDefault(require("./msgs/MsgCancelDerivativeOrder.js"));
exports.MsgCancelDerivativeOrder = MsgCancelDerivativeOrder_js_1.default;
const MsgCreateSpotMarketOrder_js_1 = __importDefault(require("./msgs/MsgCreateSpotMarketOrder.js"));
exports.MsgCreateSpotMarketOrder = MsgCreateSpotMarketOrder_js_1.default;
const MsgIncreasePositionMargin_js_1 = __importDefault(require("./msgs/MsgIncreasePositionMargin.js"));
exports.MsgIncreasePositionMargin = MsgIncreasePositionMargin_js_1.default;
const MsgInstantSpotMarketLaunch_js_1 = __importDefault(require("./msgs/MsgInstantSpotMarketLaunch.js"));
exports.MsgInstantSpotMarketLaunch = MsgInstantSpotMarketLaunch_js_1.default;
const MsgCancelBinaryOptionsOrder_js_1 = __importDefault(require("./msgs/MsgCancelBinaryOptionsOrder.js"));
exports.MsgCancelBinaryOptionsOrder = MsgCancelBinaryOptionsOrder_js_1.default;
const MsgCreateDerivativeLimitOrder_js_1 = __importDefault(require("./msgs/MsgCreateDerivativeLimitOrder.js"));
exports.MsgCreateDerivativeLimitOrder = MsgCreateDerivativeLimitOrder_js_1.default;
const MsgCreateDerivativeMarketOrder_js_1 = __importDefault(require("./msgs/MsgCreateDerivativeMarketOrder.js"));
exports.MsgCreateDerivativeMarketOrder = MsgCreateDerivativeMarketOrder_js_1.default;
const MsgBatchCancelDerivativeOrders_js_1 = __importDefault(require("./msgs/MsgBatchCancelDerivativeOrders.js"));
exports.MsgBatchCancelDerivativeOrders = MsgBatchCancelDerivativeOrders_js_1.default;
const MsgCreateBinaryOptionsLimitOrder_js_1 = __importDefault(require("./msgs/MsgCreateBinaryOptionsLimitOrder.js"));
exports.MsgCreateBinaryOptionsLimitOrder = MsgCreateBinaryOptionsLimitOrder_js_1.default;
const MsgCreateBinaryOptionsMarketOrder_js_1 = __importDefault(require("./msgs/MsgCreateBinaryOptionsMarketOrder.js"));
exports.MsgCreateBinaryOptionsMarketOrder = MsgCreateBinaryOptionsMarketOrder_js_1.default;
const MsgInstantBinaryOptionsMarketLaunch_js_1 = __importDefault(require("./msgs/MsgInstantBinaryOptionsMarketLaunch.js"));
exports.MsgInstantBinaryOptionsMarketLaunch = MsgInstantBinaryOptionsMarketLaunch_js_1.default;
const MsgBatchCancelBinaryOptionsOrders_js_1 = __importDefault(require("./msgs/MsgBatchCancelBinaryOptionsOrders.js"));
exports.MsgBatchCancelBinaryOptionsOrders = MsgBatchCancelBinaryOptionsOrders_js_1.default;
const MsgAdminUpdateBinaryOptionsMarket_js_1 = __importDefault(require("./msgs/MsgAdminUpdateBinaryOptionsMarket.js"));
exports.MsgAdminUpdateBinaryOptionsMarket = MsgAdminUpdateBinaryOptionsMarket_js_1.default;
const MsgAuthorizeStakeGrants_js_1 = __importDefault(require("./msgs/MsgAuthorizeStakeGrants.js"));
exports.MsgAuthorizeStakeGrants = MsgAuthorizeStakeGrants_js_1.default;
__exportStar(require("./utils/index.js"), exports);
