// @chain-registry/utils/esm/chains.js
import { customFind } from './utils';
export const getGasPriceRangesFromChain = (chain) => {
    const feeToken = chain.fees?.fee_tokens?.[0];
    return {
        low: feeToken?.low_gas_price ?? 0.01,
        average: feeToken?.average_gas_price ?? 0.025,
        high: feeToken?.high_gas_price ?? 0.04
    };
};
export const getChainByChainName = (chains, chainName) => {
    return customFind(chains, (chain) => chain.chain_name === chainName, `chainName:${chainName}`);
};
export const getChainByChainId = (chains, chainId) => {
    return customFind(chains, (chain) => chain.chain_id === chainId, `chainId:${chainId}`);
};
export const getChainNameByChainId = (chains, chainId) => {
    return getChainByChainId(chains, chainId)?.chain_name;
};
export const getChainIdByChainName = (chains, chainName) => {
    return getChainByChainName(chains, chainName)?.chain_id;
};
export const getChainGasPriceRanges = (chains, chainName) => {
    const chain = getChainByChainName(chains, chainName);
    return chain ? getGasPriceRangesFromChain(chain) : undefined;
};
export const getChainPrettyName = (chains, chainName) => {
    return getChainByChainName(chains, chainName)?.pretty_name;
};
export const getChainBech32Prefix = (chains, chainName) => {
    return getChainByChainName(chains, chainName)?.bech32_prefix;
};
