// @tensor-hq/tensor-common/dist/cjs/programs/auction_house/buy.js
"use strict";
/*
buy
https://explorer.solana.com/tx/5na6GnUhy1hMX4Q9mRREJkkmwbpQRiApgnpMwnQQ6ku2hP4KCzWyG3qeJiXrodu9xLrkLyF9N4dVt8APur5cGDcd
deposit + buy + transfer + withdraw from fee + transfer + transfer + exec sale
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAHBuyTx = void 0;
const mpl_auction_house_1 = require("@metaplex-foundation/mpl-auction-house");
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const bn_js_1 = __importDefault(require("bn.js"));
const shared_1 = require("./shared");
const solana_contrib_1 = require("../../solana_contrib");
const metaplex_1 = require("../../metaplex");
const makeAHBuyTx = async (connections, tokenMint, auctionHouse, buyer, priceLamports, tokenSize = 1, ahProgramId = mpl_auction_house_1.PROGRAM_ID) => {
    const connection = connections[0];
    const instructions = [];
    const additionalSigners = [];
    const auctionHouseKey = new web3_js_1.PublicKey(auctionHouse);
    const mintKey = new web3_js_1.PublicKey(tokenMint);
    const buyerKey = new web3_js_1.PublicKey(buyer);
    const auctionHouseObj = await mpl_auction_house_1.AuctionHouse.fromAccountAddress(connection, auctionHouseKey);
    const tokenSizeAdjusted = await (0, shared_1.getQuantityWithMantissa)(connection, tokenSize, mintKey);
    const buyerTokenAccountKey = await (0, spl_token_1.getAssociatedTokenAddress)(mintKey, buyerKey);
    //this is supposed to be the account holding the NFT
    const largestTokenHolders = await connection.getTokenLargestAccounts(mintKey);
    const sellerTokenAccountKey = largestTokenHolders.value[0].address;
    const sellerTokenAcc = await (0, spl_token_1.getAccount)(connection, sellerTokenAccountKey);
    const sellerKey = new web3_js_1.PublicKey(sellerTokenAcc.owner);
    const [programAsSigner, programAsSignerBump] = (0, metaplex_1.findAuctionHouseProgramAsSignerPda)(ahProgramId);
    const [sellerTradeState] = (0, metaplex_1.findAuctionHouseTradeStatePda)(auctionHouseKey, sellerKey, auctionHouseObj.treasuryMint, mintKey, priceLamports, tokenSizeAdjusted, sellerTokenAccountKey);
    const [buyerTradeState, buyerTradeBump] = (0, metaplex_1.findAuctionHouseTradeStatePda)(auctionHouseKey, buyerKey, auctionHouseObj.treasuryMint, mintKey, priceLamports, tokenSizeAdjusted, sellerTokenAccountKey);
    const [freeTradeState, freeTradeBump] = (0, metaplex_1.findAuctionHouseTradeStatePda)(auctionHouseKey, sellerKey, auctionHouseObj.treasuryMint, mintKey, new bn_js_1.default(0), tokenSizeAdjusted, sellerTokenAccountKey, ahProgramId);
    const [escrowPaymentAccount, escrowPaymentBump] = (0, metaplex_1.findAuctionHouseBuyerEscrowPda)(auctionHouseKey, buyerKey, ahProgramId);
    const [metadata] = (0, metaplex_1.findMetadataPda)(mintKey);
    const depositIx = (0, mpl_auction_house_1.createDepositInstruction)({
        auctionHouse: auctionHouseKey,
        auctionHouseFeeAccount: auctionHouseObj.auctionHouseFeeAccount,
        authority: auctionHouseObj.authority,
        escrowPaymentAccount,
        paymentAccount: buyerKey,
        transferAuthority: auctionHouseObj.authority,
        treasuryMint: auctionHouseObj.treasuryMint,
        wallet: buyerKey,
    }, {
        amount: priceLamports,
        escrowPaymentBump,
    });
    const buyIx = (0, mpl_auction_house_1.createBuyInstruction)({
        auctionHouse: auctionHouseKey,
        auctionHouseFeeAccount: auctionHouseObj.auctionHouseFeeAccount,
        authority: auctionHouseObj.authority,
        buyerTradeState,
        escrowPaymentAccount,
        metadata,
        paymentAccount: buyerKey,
        tokenAccount: sellerTokenAccountKey,
        transferAuthority: web3_js_1.SystemProgram.programId,
        treasuryMint: auctionHouseObj.treasuryMint,
        wallet: buyerKey,
    }, {
        buyerPrice: priceLamports,
        escrowPaymentBump,
        tokenSize: tokenSizeAdjusted,
        tradeStateBump: buyerTradeBump,
    });
    const createAtaIx = (0, spl_token_1.createAssociatedTokenAccountInstruction)(buyerKey, buyerTokenAccountKey, buyerKey, mintKey);
    const execSaleIx = (0, mpl_auction_house_1.createExecuteSaleInstruction)({
        auctionHouse: auctionHouseKey,
        auctionHouseFeeAccount: auctionHouseObj.auctionHouseFeeAccount,
        auctionHouseTreasury: auctionHouseObj.auctionHouseTreasury,
        authority: auctionHouseObj.authority,
        buyer: buyerKey,
        buyerReceiptTokenAccount: buyerTokenAccountKey,
        buyerTradeState,
        escrowPaymentAccount,
        freeTradeState,
        metadata,
        programAsSigner,
        seller: sellerKey,
        sellerPaymentReceiptAccount: sellerKey,
        sellerTradeState,
        tokenAccount: sellerTokenAccountKey,
        tokenMint: mintKey,
        treasuryMint: auctionHouseObj.treasuryMint,
    }, {
        buyerPrice: priceLamports,
        escrowPaymentBump,
        freeTradeStateBump: freeTradeBump,
        programAsSignerBump,
        tokenSize: tokenSizeAdjusted,
    });
    //add creators for royalty payments
    const metadataDecoded = await mpl_token_metadata_1.Metadata.fromAccountAddress(connection, metadata);
    for (let i = 0; i < metadataDecoded.data.creators.length; i++) {
        execSaleIx.keys.push({
            pubkey: new web3_js_1.PublicKey(metadataDecoded.data.creators[i].address),
            isWritable: true,
            isSigner: false,
        });
    }
    instructions.push(depositIx, buyIx);
    if (auctionHouseObj.requiresSignOff) {
        execSaleIx.keys[9].isSigner = true;
    }
    //optionally create ata for buyer
    const buyerAtaInfo = await connection.getAccountInfo(buyerTokenAccountKey);
    if (!buyerAtaInfo?.lamports || !buyerAtaInfo.data?.length) {
        instructions.push(createAtaIx);
    }
    instructions.push(execSaleIx);
    return {
        ...(await (0, solana_contrib_1.buildTx)({
            maybeBlockhash: {
                type: 'blockhashArgs',
                args: {
                    connections,
                },
            },
            instructions,
            additionalSigners,
            feePayer: buyerKey,
        })),
        auctionHouseObj,
        sellerTradeState,
    };
};
exports.makeAHBuyTx = makeAHBuyTx;
