// @tensor-hq/tensor-common/dist/cjs/programs/auction_house/shared.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getQuantityWithMantissa = exports.AUCTION_HOUSE_KEYS = exports.AUCTION_HOUSE_PROGRAM_ID = exports.AUCTION_HOUSE = void 0;
const web3_js_1 = require("@solana/web3.js");
const bn_js_1 = __importDefault(require("bn.js"));
const spl_token_1 = require("@solana/spl-token");
exports.AUCTION_HOUSE = 'auction_house';
exports.AUCTION_HOUSE_PROGRAM_ID = new web3_js_1.PublicKey('hausS13jsjafwWwGqZTUQRmWyvyxn9EQpqMwV1PBBmk');
exports.AUCTION_HOUSE_KEYS = {
    OpenSea: '3o9d13qUvEuuauhFrVom1vuCzgNsJifeaBYDPquaT73Y',
    Solanart: 'GWErq8nJf5JQtohg5k7RTkiZmoCxvGBJqbMSfkrxYFFy',
    CoralCube: '29xtkHHFLUHXiLoxTzbC7U8kekTwN3mVQSkfXnB1sQ6e',
    Fractal: 'BAmKB58MgkeYF2VueVBfASL5q8Qf6VKp4nA4cRuVUVft',
};
const getQuantityWithMantissa = async (conn, quantity, mint) => {
    const mintInfo = await (0, spl_token_1.getMint)(conn, new web3_js_1.PublicKey(mint));
    const mantissa = new bn_js_1.default(10).pow(new bn_js_1.default(mintInfo.decimals));
    return mantissa.mul(new bn_js_1.default(quantity));
};
exports.getQuantityWithMantissa = getQuantityWithMantissa;
