// @injectivelabs/sdk-ts/dist/cjs/core/modules/gov/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgGovDeposit = exports.MsgVote = exports.MsgSubmitProposalExpiryFuturesMarketLaunch = exports.MsgSubmitProposalSpotMarketParamUpdate = exports.MsgSubmitProposalPerpetualMarketLaunch = exports.MsgSubmitProposalSpotMarketLaunch = exports.MsgSubmitGenericProposal = exports.MsgSubmitTextProposal = void 0;
const MsgVote_js_1 = __importDefault(require("./msgs/MsgVote.js"));
exports.MsgVote = MsgVote_js_1.default;
const MsgDeposit_js_1 = __importDefault(require("./msgs/MsgDeposit.js"));
exports.MsgGovDeposit = MsgDeposit_js_1.default;
const MsgSubmitTextProposal_js_1 = __importDefault(require("./msgs/MsgSubmitTextProposal.js"));
exports.MsgSubmitTextProposal = MsgSubmitTextProposal_js_1.default;
const MsgSubmitGenericProposal_js_1 = __importDefault(require("./msgs/MsgSubmitGenericProposal.js"));
exports.MsgSubmitGenericProposal = MsgSubmitGenericProposal_js_1.default;
const MsgSubmitProposalSpotMarketLaunch_js_1 = __importDefault(require("./msgs/MsgSubmitProposalSpotMarketLaunch.js"));
exports.MsgSubmitProposalSpotMarketLaunch = MsgSubmitProposalSpotMarketLaunch_js_1.default;
const MsgSubmitProposalPerpetualMarketLaunch_js_1 = __importDefault(require("./msgs/MsgSubmitProposalPerpetualMarketLaunch.js"));
exports.MsgSubmitProposalPerpetualMarketLaunch = MsgSubmitProposalPerpetualMarketLaunch_js_1.default;
const MsgSubmitProposalSpotMarketParamUpdate_js_1 = __importDefault(require("./msgs/MsgSubmitProposalSpotMarketParamUpdate.js"));
exports.MsgSubmitProposalSpotMarketParamUpdate = MsgSubmitProposalSpotMarketParamUpdate_js_1.default;
const MsgSubmitProposalExpiryFuturesMarketLaunch_js_1 = __importDefault(require("./msgs/MsgSubmitProposalExpiryFuturesMarketLaunch.js"));
exports.MsgSubmitProposalExpiryFuturesMarketLaunch = MsgSubmitProposalExpiryFuturesMarketLaunch_js_1.default;
__exportStar(require("./ProposalContentDecomposer.js"), exports);
