// @chain-registry/utils/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChainBech32Prefix = exports.getChainPrettyName = exports.getChainGasPriceRanges = exports.getChainIdByChainName = exports.getChainNameByChainId = exports.getChainByChainId = exports.getChainByChainName = exports.getGasPriceRangesFromChain = void 0;
const utils_1 = require("./utils");
const getGasPriceRangesFromChain = (chain) => {
    const feeToken = chain.fees?.fee_tokens?.[0];
    return {
        low: feeToken?.low_gas_price ?? 0.01,
        average: feeToken?.average_gas_price ?? 0.025,
        high: feeToken?.high_gas_price ?? 0.04
    };
};
exports.getGasPriceRangesFromChain = getGasPriceRangesFromChain;
const getChainByChainName = (chains, chainName) => {
    return (0, utils_1.customFind)(chains, (chain) => chain.chain_name === chainName, `chainName:${chainName}`);
};
exports.getChainByChainName = getChainByChainName;
const getChainByChainId = (chains, chainId) => {
    return (0, utils_1.customFind)(chains, (chain) => chain.chain_id === chainId, `chainId:${chainId}`);
};
exports.getChainByChainId = getChainByChainId;
const getChainNameByChainId = (chains, chainId) => {
    return (0, exports.getChainByChainId)(chains, chainId)?.chain_name;
};
exports.getChainNameByChainId = getChainNameByChainId;
const getChainIdByChainName = (chains, chainName) => {
    return (0, exports.getChainByChainName)(chains, chainName)?.chain_id;
};
exports.getChainIdByChainName = getChainIdByChainName;
const getChainGasPriceRanges = (chains, chainName) => {
    const chain = (0, exports.getChainByChainName)(chains, chainName);
    return chain ? (0, exports.getGasPriceRangesFromChain)(chain) : undefined;
};
exports.getChainGasPriceRanges = getChainGasPriceRanges;
const getChainPrettyName = (chains, chainName) => {
    return (0, exports.getChainByChainName)(chains, chainName)?.pretty_name;
};
exports.getChainPrettyName = getChainPrettyName;
const getChainBech32Prefix = (chains, chainName) => {
    return (0, exports.getChainByChainName)(chains, chainName)?.bech32_prefix;
};
exports.getChainBech32Prefix = getChainBech32Prefix;
