// chain-registry/mainnet/babylon/ibc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const info = [
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'axelar',
            client_id: '07-tendermint-244',
            connection_id: 'connection-233'
        },
        chain_2: {
            chain_name: 'babylon',
            client_id: '07-tendermint-2',
            connection_id: 'connection-2'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-175',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-2',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live'
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'babylon',
            client_id: '07-tendermint-0',
            connection_id: 'connection-0'
        },
        chain_2: {
            chain_name: 'cosmoshub',
            client_id: '07-tendermint-1381',
            connection_id: 'connection-1104'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-0',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-1341',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'babylon',
            client_id: '07-tendermint-6',
            connection_id: 'connection-6'
        },
        chain_2: {
            chain_name: 'elys',
            client_id: '07-tendermint-51',
            connection_id: 'connection-35'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-6',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-22',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'babylon',
            client_id: '07-tendermint-5',
            connection_id: 'connection-5'
        },
        chain_2: {
            chain_name: 'neutron',
            client_id: '07-tendermint-164',
            connection_id: 'connection-123'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-5',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-6980',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'babylon',
            client_id: '07-tendermint-1',
            connection_id: 'connection-1'
        },
        chain_2: {
            chain_name: 'noble',
            client_id: '07-tendermint-147',
            connection_id: 'connection-140'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-1',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-132',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live'
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'babylon',
            client_id: '07-tendermint-3',
            connection_id: 'connection-3'
        },
        chain_2: {
            chain_name: 'osmosis',
            client_id: '07-tendermint-3470',
            connection_id: 'connection-10694'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-3',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-101635',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }, {
                chain_1: {
                    channel_id: 'channel-7',
                    port_id: 'wasm.bbn16tfcsjcg4trlnj2t4cpdq20q4d7nzusew3ev50scz30mmfzdx8ssgx7eq6'
                },
                chain_2: {
                    channel_id: 'channel-101979',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live'
                }
            }]
    }
];
exports.default = info;
