// @tensor-hq/tensor-common/dist/esm/tensor/types.js
// ======================== Rarities.
// NB: Make sure keys are CamelCased vs CAMELCased.
// This is so it aligns with the generated graphql enum keys to make life easier.
export var RaritySystem;
(function (RaritySystem) {
    RaritySystem["Hrtt"] = "Hrtt";
    RaritySystem["Stat"] = "Stat";
    RaritySystem["Team"] = "Team";
    RaritySystem["Tn"] = "Tn";
})(RaritySystem || (RaritySystem = {}));
// ======================== Sorting.
// NB: Make sure keys are CamelCased vs CAMELCased.
// This is so it aligns with the generated graphql enum keys to make life easier.
export var MintsSortBy;
(function (MintsSortBy) {
    MintsSortBy["PriceAsc"] = "PriceAsc";
    MintsSortBy["PriceDesc"] = "PriceDesc";
    MintsSortBy["LastSaleAsc"] = "LastSaleAsc";
    MintsSortBy["LastSaleDesc"] = "LastSaleDesc";
    MintsSortBy["ListedDesc"] = "ListedDesc";
    MintsSortBy["RankHrttAsc"] = "RankHrttAsc";
    MintsSortBy["RankHrttDesc"] = "RankHrttDesc";
    MintsSortBy["RankStatAsc"] = "RankStatAsc";
    MintsSortBy["RankStatDesc"] = "RankStatDesc";
    MintsSortBy["RankTeamAsc"] = "RankTeamAsc";
    MintsSortBy["RankTeamDesc"] = "RankTeamDesc";
    MintsSortBy["RankTnAsc"] = "RankTnAsc";
    MintsSortBy["RankTnDesc"] = "RankTnDesc";
    MintsSortBy["OrdinalAsc"] = "OrdinalAsc";
    MintsSortBy["OrdinalDesc"] = "OrdinalDesc";
    MintsSortBy["NormalizedPriceAsc"] = "NormalizedPriceAsc";
    MintsSortBy["NormalizedPriceDesc"] = "NormalizedPriceDesc";
    MintsSortBy["HybridAmountAsc"] = "HybridAmountAsc";
    MintsSortBy["HybridAmountDesc"] = "HybridAmountDesc";
})(MintsSortBy || (MintsSortBy = {}));
