// quais/lib/commonjs/wallet/qi-hdwallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QiHDWallet = exports.AddressStatus = void 0;
const tslib_1 = require("tslib");
const hdwallet_js_1 = require("./hdwallet.js");
const hdnodewallet_js_1 = require("./hdnodewallet.js");
const index_js_1 = require("../address/index.js");
const index_js_2 = require("../utils/index.js");
const index_js_3 = require("../transaction/index.js");
const musig_1 = require("@brandonblack/musig");
const secp256k1_1 = require("@noble/curves/secp256k1");
const index_js_4 = require("../crypto/index.js");
const utxo_js_1 = require("../transaction/utxo.js");
const index_js_5 = require("../constants/index.js");
const mnemonic_js_1 = require("./mnemonic.js");
const payment_codes_js_1 = require("./payment-codes.js");
const bip32_js_1 = require("./bip32/bip32.js");
const crypto_js_1 = require("./bip32/crypto.js");
const types_js_1 = require("./bip32/types.js");
const secp256k1_2 = tslib_1.__importDefault(require("@bitcoinerlab/secp256k1"));
/**
 * Enum representing the status of an address in the wallet.
 *
 * @enum {string}
 */
var AddressStatus;
(function (AddressStatus) {
    AddressStatus["USED"] = "USED";
    AddressStatus["UNUSED"] = "UNUSED";
    AddressStatus["ATTEMPTED_USE"] = "ATTEMPTED_USE";
    AddressStatus["UNKNOWN"] = "UNKNOWN";
})(AddressStatus = exports.AddressStatus || (exports.AddressStatus = {}));
/**
 * Current known issues:
 *
 * - When generating send addresses we are not checking if the address has already been used before
 * - When syncing is seems like we are adding way too many change addresses
 * - Bip44 external and change address maps also have gap addresses in them
 * - It is unclear if we have checked if addresses have been used and if they are used
 * - We should always check all addresses that were previously included in a transaction to see if they have been used
 */
/**
 * The Qi HD wallet is a BIP44-compliant hierarchical deterministic wallet used for managing a set of addresses in the
 * Qi ledger. This is wallet implementation is the primary way to interact with the Qi UTXO ledger on the Quai network.
 *
 * The Qi HD wallet supports:
 *
 * - Adding accounts to the wallet heierchy
 * - Generating addresses for a specific account in any {@link Zone}
 * - Signing and sending transactions for any address in the wallet
 * - Serializing the wallet to JSON and deserializing it back to a wallet instance.
 *
 * @category Wallet
 * @example
 *
 * ```ts
 * import { QiHDWallet, Zone } from 'quais';
 *
 * const wallet = new QiHDWallet();
 * const cyrpus1Address = await wallet.getNextAddress(0, Zone.Cyrpus1); // get the first address in the Cyrpus1 zone
 * await wallet.sendTransaction({ txInputs: [...], txOutputs: [...] }); // send a transaction
 * const serializedWallet = wallet.serialize(); // serialize current (account/address) state of the wallet
 * .
 * .
 * .
 * const deserializedWallet = QiHDWallet.deserialize(serializedWallet); // create a new wallet instance from the serialized data
 * ```
 */
class QiHDWallet extends hdwallet_js_1.AbstractHDWallet {
    /**
     * @ignore
     * @type {number}
     */
    static _version = 1;
    /**
     * @ignore
     * @type {number}
     */
    static _GAP_LIMIT = 5;
    /**
     * @ignore
     * @type {AllowedCoinType}
     */
    static _coinType = 969;
    /**
     * @ignore
     * @type {string}
     */
    static PRIVATE_KEYS_PATH = 'privateKeys';
    /**
     * A map containing address information for all addresses known to the wallet. This includes:
     *
     * - BIP44 derived addresses (external)
     * - BIP44 derived change addresses
     * - BIP47 payment code derived addresses for receiving funds
     *
     * The key is the derivation path or payment code, and the value is an array of QiAddressInfo objects.
     *
     * @private
     * @type {Map<DerivationPath, QiAddressInfo[]>}
     */
    _addressesMap = new Map();
    /**
     * Array of outpoint information.
     *
     * @ignore
     * @type {OutpointInfo[]}
     */
    _availableOutpoints = [];
    /**
     * Map of outpoints that are pending confirmation of being spent.
     */
    _pendingOutpoints = [];
    /**
     * @ignore
     * @type {AddressUsageCallback}
     */
    _addressUseChecker;
    /**
     * A map containing address information for sending funds to counterparties using BIP47 payment codes.
     *
     * @remarks
     * The key is the receiver's payment code, and the value is an array of QiAddressInfo objects. These addresses are
     * derived from the receiver's payment code and are used only for sending funds. They are not part of the set of
     * addresses that this wallet can control or spend from. This map is used to keep track of addresses generated for
     * each payment channel to ensure proper address rotation and avoid address reuse when sending funds.
     * @private
     * @type {Map<string, QiAddressInfo[]>}
     */
    _paymentCodeSendAddressMap = new Map();
    /**
     * @ignore
     * @param {HDNodeWallet} root - The root HDNodeWallet.
     * @param {Provider} [provider] - The provider (optional).
     */
    constructor(guard, root, provider) {
        super(guard, root, provider);
        this._addressesMap.set('BIP44:external', []);
        this._addressesMap.set('BIP44:change', []);
        this._addressesMap.set(QiHDWallet.PRIVATE_KEYS_PATH, []);
    }
    /**
     * Gets the payment codes for all open channels.
     *
     * @returns {string[]} The payment codes for all open channels.
     */
    get openChannels() {
        return Array.from(this._addressesMap.keys()).filter((key) => !key.startsWith('BIP44:') && key !== QiHDWallet.PRIVATE_KEYS_PATH);
    }
    /**
     * Sets the address use checker. The provided callback function should accept an address as input and return a
     * boolean indicating whether the address is in use. If the callback returns true, the address is considered used
     * and if it returns false, the address is considered unused.
     *
     * @param {AddressUsageCallback} checker - The address use checker.
     */
    setAddressUseChecker(checker) {
        this._addressUseChecker = checker;
    }
    /**
     * Finds the last used index in an array of QiAddressInfo objects. If no index is found, returns -1.
     *
     * @param {QiAddressInfo[]} addresses - The array of QiAddressInfo objects.
     * @returns {number} The last used index.
     */
    _findLastUsedIndex(addresses, account, zone) {
        const filteredAddresses = addresses?.filter((addressInfo) => addressInfo.account === account && addressInfo.zone === zone);
        return filteredAddresses?.reduce((maxIndex, addressInfo) => Math.max(maxIndex, addressInfo.index), -1) || -1;
    }
    /**
     * Derives the next Qi BIP 44 address for the specified account and zone.
     *
     * @param {number} account - The account number.
     * @param {Zone} zone - The zone.
     * @param {boolean} isChange - Whether to derive a change address.
     * @returns {QiAddressInfo} The next Qi address information.
     */
    _getNextQiAddress(account, zone, isChange) {
        const addresses = this._addressesMap.get(isChange ? 'BIP44:change' : 'BIP44:external') || [];
        const lastIndex = this._findLastUsedIndex(addresses, account, zone);
        const addressNode = this.deriveNextAddressNode(account, lastIndex + 1, zone, isChange);
        const privateKeysArray = this._addressesMap.get(QiHDWallet.PRIVATE_KEYS_PATH) || [];
        const existingPrivateKeyIndex = privateKeysArray.findIndex((info) => info.address === addressNode.address);
        if (existingPrivateKeyIndex !== -1) {
            privateKeysArray.splice(existingPrivateKeyIndex, 1);
            this._addressesMap.set(QiHDWallet.PRIVATE_KEYS_PATH, privateKeysArray);
        }
        const newAddrInfo = {
            pubKey: addressNode.publicKey,
            address: addressNode.address,
            account,
            index: addressNode.index,
            change: isChange,
            zone,
            status: AddressStatus.UNUSED,
            derivationPath: isChange ? 'BIP44:change' : 'BIP44:external',
        };
        this._addressesMap.get(isChange ? 'BIP44:change' : 'BIP44:external')?.push(newAddrInfo);
        return newAddrInfo;
    }
    /**
     * Promise that resolves to the next address for the specified account and zone.
     *
     * @param {number} account - The account number.
     * @param {Zone} zone - The zone.
     * @returns {Promise<QiAddressInfo>} The next Qi address information.
     */
    async getNextAddress(account, zone) {
        return Promise.resolve(this.getNextAddressSync(account, zone));
    }
    /**
     * Synchronously retrieves the next address for the specified account and zone.
     *
     * @param {number} account - The account number.
     * @param {Zone} zone - The zone.
     * @returns {QiAddressInfo} The next Qi address information.
     */
    getNextAddressSync(account, zone) {
        return this._getNextQiAddress(account, zone, false);
    }
    /**
     * Promise that resolves to the next change address for the specified account and zone.
     *
     * @param {number} account - The index of the account for which to retrieve the next change address.
     * @param {Zone} zone - The zone in which to retrieve the next change address.
     * @returns {Promise<NeuteredAddressInfo>} The next change neutered address information.
     */
    async getNextChangeAddress(account, zone) {
        return Promise.resolve(this.getNextChangeAddressSync(account, zone));
    }
    /**
     * Synchronously retrieves the next change address for the specified account and zone.
     *
     * @param {number} account - The index of the account for which to retrieve the next change address.
     * @param {Zone} zone - The zone in which to retrieve the next change address.
     * @returns {NeuteredAddressInfo} The next change neutered address information.
     */
    getNextChangeAddressSync(account, zone) {
        return this._getNextQiAddress(account, zone, true);
    }
    /**
     * Imports an array of outpoints.
     *
     * @param {OutpointInfo[]} outpoints - The outpoints to import.
     */
    importOutpoints(outpoints) {
        this.validateOutpointInfo(outpoints);
        // check if any of the outpoints are already in the availableOutpoints
        const newOutpoints = outpoints.filter((outpoint) => !this._availableOutpoints.some((o) => o.outpoint.txhash === outpoint.outpoint.txhash && o.outpoint.index === outpoint.outpoint.index));
        this._availableOutpoints.push(...newOutpoints);
    }
    /**
     * Gets the outpoints for the specified zone.
     *
     * @param {Zone} zone - The zone.
     * @returns {OutpointInfo[]} The outpoints for the zone.
     */
    getOutpoints(zone) {
        this.validateZone(zone);
        return this._availableOutpoints.filter((outpoint) => outpoint.zone === zone);
    }
    /**
     * Signs a Qi transaction and returns the serialized transaction.
     *
     * @param {QiTransactionRequest} tx - The transaction to sign.
     * @returns {Promise<string>} The serialized transaction.
     * @throws {Error} If the UTXO transaction is invalid.
     */
    async signTransaction(tx) {
        const txobj = index_js_3.QiTransaction.from(tx);
        if (!txobj.txInputs || txobj.txInputs.length == 0 || !txobj.txOutputs)
            throw new Error('Invalid UTXO transaction, missing inputs or outputs');
        const hash = (0, index_js_2.getBytes)((0, index_js_4.keccak256)(txobj.unsignedSerialized));
        let signature;
        if (txobj.txInputs.length === 1) {
            signature = this.createSchnorrSignature(txobj.txInputs[0], hash);
        }
        else {
            signature = this.createMuSigSignature(txobj, hash);
        }
        txobj.signature = signature;
        return txobj.serialized;
    }
    /**
     * Locates the address information for the given address, searching through standard addresses, change addresses,
     * and payment channel addresses.
     *
     * @param {string} address - The address to locate.
     * @returns {QiAddressInfo | null} The address info or null if not found.
     */
    locateAddressInfo(address) {
        for (const [, addressInfos] of this._addressesMap.entries()) {
            const addressInfo = addressInfos.find((info) => info.address === address);
            if (addressInfo) {
                return addressInfo;
            }
        }
        return null;
    }
    /**
     * Gets the **total** balance for the specified zone, including locked UTXOs.
     *
     * @param {Zone} zone - The zone to get the balance for.
     * @returns {bigint} The total balance for the zone.
     */
    getBalanceForZone(zone) {
        this.validateZone(zone);
        return this._availableOutpoints
            .filter((outpoint) => outpoint.zone === zone)
            .reduce((total, outpoint) => {
            const denominationValue = utxo_js_1.denominations[outpoint.outpoint.denomination];
            return total + denominationValue;
        }, BigInt(0));
    }
    /**
     * Gets the locked balance for the specified zone.
     *
     * @param {Zone} zone - The zone to get the locked balance for.
     * @returns {bigint} The locked balance for the zone.
     */
    async getSpendableBalanceForZone(zone, blockNumber) {
        this.validateZone(zone);
        if (!this.provider) {
            throw new Error('Provider is not set');
        }
        if (!blockNumber) {
            blockNumber = await this.provider.getBlockNumber((0, index_js_5.toShard)(zone));
        }
        return this._availableOutpoints
            .filter((utxo) => utxo.outpoint.lock === 0 || utxo.outpoint.lock < blockNumber)
            .reduce((total, utxo) => {
            const denominationValue = utxo_js_1.denominations[utxo.outpoint.denomination];
            return total + denominationValue;
        }, BigInt(0));
    }
    /**
     * Gets the locked balance for the specified zone.
     *
     * @param {Zone} zone - The zone to get the locked balance for.
     * @returns {bigint} The locked balance for the zone.
     */
    async getLockedBalanceForZone(zone, blockNumber) {
        this.validateZone(zone);
        if (!this.provider) {
            throw new Error('Provider is not set');
        }
        if (!blockNumber) {
            blockNumber = await this.provider.getBlockNumber((0, index_js_5.toShard)(zone));
        }
        return this._availableOutpoints
            .filter((utxo) => utxo.outpoint.lock !== 0 && blockNumber < utxo.outpoint.lock)
            .reduce((total, utxo) => {
            const denominationValue = utxo_js_1.denominations[utxo.outpoint.denomination];
            return total + denominationValue;
        }, BigInt(0));
    }
    /**
     * Converts outpoints for a specific zone to UTXO format.
     *
     * @param {Zone} zone - The zone to filter outpoints for.
     * @param {number} [minDenominationToUse] - The minimum denomination to allow for the UTXOs.
     * @returns {UTXO[]} An array of UTXO objects.
     */
    outpointsToUTXOs(zone, minDenominationToUse) {
        this.validateZone(zone);
        let zoneOutpoints = this._availableOutpoints.filter((outpointInfo) => outpointInfo.zone === zone);
        // Filter outpoints by minimum denomination if specified
        // This will likely only be used for converting to Quai
        // as the min denomination for converting is 10 (100 Qi)
        if (minDenominationToUse !== undefined) {
            zoneOutpoints = zoneOutpoints.filter((outpointInfo) => outpointInfo.outpoint.denomination >= minDenominationToUse);
        }
        return zoneOutpoints.map((outpointInfo) => {
            const utxo = new utxo_js_1.UTXO();
            utxo.txhash = outpointInfo.outpoint.txhash;
            utxo.index = outpointInfo.outpoint.index;
            utxo.address = outpointInfo.address;
            utxo.denomination = outpointInfo.outpoint.denomination;
            utxo.lock = outpointInfo.outpoint.lock ?? null;
            return utxo;
        });
    }
    async prepareAndSendTransaction(amount, originZone, getDestinationAddresses, minDenominationToUse) {
        if (!this.provider) {
            throw new Error('Provider is not set');
        }
        // 1. Check the wallet has enough balance in the originating zone to send the transaction
        const currentBlock = await this.provider.getBlockNumber(originZone);
        const balance = await this.getSpendableBalanceForZone(originZone, currentBlock);
        if (balance < amount) {
            throw new Error(`Insufficient balance in the originating zone: want ${Number(amount) / 1000} Qi got ${balance} Qi`);
        }
        // 2. Select the UXTOs from the specified zone to use as inputs, and generate the spend and change outputs
        const zoneUTXOs = this.outpointsToUTXOs(originZone, minDenominationToUse);
        if (zoneUTXOs.length === 0) {
            if (minDenominationToUse === 10) {
                throw new Error('Qi denominations too small to convert.');
            }
            else {
                throw new Error('No Qi available in zone.');
            }
        }
        const unlockedUTXOs = zoneUTXOs.filter((utxo) => utxo.lock === 0 || utxo.lock < currentBlock);
        if (unlockedUTXOs.length === 0) {
            throw new Error('Insufficient spendable balance in zone.');
        }
        const fewestCoinSelector = new index_js_3.FewestCoinSelector(unlockedUTXOs);
        const spendTarget = amount;
        let selection = fewestCoinSelector.performSelection(spendTarget);
        // 3. Generate as many unused addresses as required to populate the spend outputs
        const sendAddresses = await getDestinationAddresses(selection.spendOutputs.length);
        const getChangeAddressesForOutputs = async (count) => {
            const currentChangeAddresses = this._addressesMap.get('BIP44:change') || [];
            const outpusChangeAddresses = [];
            for (let i = 0; i < currentChangeAddresses.length; i++) {
                if (currentChangeAddresses[i].status === AddressStatus.UNUSED) {
                    outpusChangeAddresses.push(currentChangeAddresses[i]);
                }
                if (outpusChangeAddresses.length === count)
                    break;
            }
            // Generate the remaining number of change addresses if needed
            const remainingAddressesNeeded = count - outpusChangeAddresses.length;
            if (remainingAddressesNeeded > 0) {
                outpusChangeAddresses.push(...Array(remainingAddressesNeeded)
                    .fill(0)
                    .map(() => this.getNextChangeAddressSync(0, originZone)));
            }
            // Combine the existing change addresses with the newly generated addresses and ensure they are unique and sorted by index
            const mergedChangeAddresses = [
                ...outpusChangeAddresses.map((address) => ({ ...address, status: AddressStatus.ATTEMPTED_USE })),
                ...currentChangeAddresses,
            ];
            const sortedAndFilteredChangeAddresses = mergedChangeAddresses
                .filter((address, index, self) => self.findIndex((t) => t.address === address.address) === index)
                .sort((a, b) => a.index - b.index);
            // Update the _addressesMap with the modified change addresses and statuses
            this._addressesMap.set('BIP44:change', sortedAndFilteredChangeAddresses);
            return outpusChangeAddresses.map((address) => address.address);
        };
        // 4. Get change addresses
        const changeAddresses = await getChangeAddressesForOutputs(selection.changeOutputs.length);
        // 5. Create the transaction and sign it using the signTransaction method
        let inputPubKeys = selection.inputs.map((input) => this.locateAddressInfo(input.address)?.pubKey);
        if (inputPubKeys.some((pubkey) => !pubkey)) {
            throw new Error('Missing public key for input address');
        }
        let attempts = 0;
        let finalFee = 0n;
        const MAX_FEE_ESTIMATION_ATTEMPTS = 5;
        while (attempts < MAX_FEE_ESTIMATION_ATTEMPTS) {
            const feeEstimationTx = this.prepareFeeEstimationTransaction(selection, inputPubKeys.map((pubkey) => pubkey), sendAddresses, changeAddresses);
            finalFee = await this.provider.estimateFeeForQi(feeEstimationTx);
            // Get new selection with updated fee 2x
            selection = fewestCoinSelector.performSelection(spendTarget, finalFee * 2n);
            // Determine if new addresses are needed for the change outputs
            const changeAddressesNeeded = selection.changeOutputs.length - changeAddresses.length;
            if (changeAddressesNeeded > 0) {
                // Need more change addresses
                const newChangeAddresses = await getChangeAddressesForOutputs(changeAddressesNeeded);
                changeAddresses.push(...newChangeAddresses);
            }
            else if (changeAddressesNeeded < 0) {
                // Have extra change addresses, remove the addresses starting from the end
                // TODO: Set the status of the addresses to UNUSED in _addressesMap. This fine for now as it will be fixed during next sync
                changeAddresses.splice(changeAddressesNeeded);
            }
            // Determine if new addresses are needed for the spend outputs
            const spendAddressesNeeded = selection.spendOutputs.length - sendAddresses.length;
            if (spendAddressesNeeded > 0) {
                // Need more send addresses
                const newSendAddresses = await getDestinationAddresses(spendAddressesNeeded);
                sendAddresses.push(...newSendAddresses);
            }
            else if (spendAddressesNeeded < 0) {
                // Have extra send addresses, remove the excess
                // TODO: Set the status of the addresses to UNUSED in _addressesMap. This fine for now as it will be fixed during next sync
                sendAddresses.splice(spendAddressesNeeded);
            }
            inputPubKeys = selection.inputs.map((input) => this.locateAddressInfo(input.address)?.pubKey);
            // Calculate total new outputs needed (absolute value)
            const totalNewOutputsNeeded = Math.abs(changeAddressesNeeded) + Math.abs(spendAddressesNeeded);
            // If we need 5 or fewer new outputs, we can break the loop
            if ((changeAddressesNeeded <= 0 && spendAddressesNeeded <= 0) || totalNewOutputsNeeded <= 5) {
                break;
            }
            attempts++;
        }
        // Proceed with creating and signing the transaction
        const chainId = (await this.provider.getNetwork()).chainId;
        const tx = await this.prepareTransaction(selection, inputPubKeys.map((pubkey) => pubkey), sendAddresses, changeAddresses, Number(chainId));
        // Move used outpoints to pendingOutpoints
        this.moveOutpointsToPending(tx.txInputs);
        // Sign the transaction
        const signedTx = await this.signTransaction(tx);
        // Broadcast the transaction to the network using the provider
        return this.provider.broadcastTransaction(originZone, signedTx);
    }
    /**
     * Converts an amount of Qi to Quai and sends it to a specified Quai address.
     *
     * @param {string} destinationAddress - The Quai address to send the converted Quai to.
     * @param {bigint} amount - The amount of Qi to convert to Quai.
     * @returns {Promise<TransactionResponse>} A promise that resolves to the transaction response.
     * @throws {Error} If the destination address is invalid, the amount is zero, or the conversion fails.
     */
    async convertToQuai(destinationAddress, amount) {
        const zone = (0, index_js_2.getZoneForAddress)(destinationAddress);
        if (!zone) {
            throw new Error(`Invalid zone for Quai address: ${destinationAddress}`);
        }
        if ((0, index_js_1.isQiAddress)(destinationAddress)) {
            throw new Error(`Invalid Quai address: ${destinationAddress}`);
        }
        if (amount <= 0) {
            throw new Error('Amount must be greater than 0');
        }
        const getDestinationAddresses = async (count) => {
            return Array(count).fill(destinationAddress);
        };
        return this.prepareAndSendTransaction(amount, zone, getDestinationAddresses, 10);
    }
    /**
     * Sends a transaction to a specified recipient payment code in a specified zone.
     *
     * @param {string} recipientPaymentCode - The payment code of the recipient.
     * @param {bigint} amount - The amount of Qi to send.
     * @param {Zone} originZone - The zone where the transaction originates.
     * @param {Zone} destinationZone - The zone where the transaction is sent.
     * @returns {Promise<TransactionResponse>} A promise that resolves to the transaction response.
     * @throws {Error} If the payment code is invalid, the amount is zero, or the zones are invalid.
     */
    async sendTransaction(recipientPaymentCode, amount, originZone, destinationZone) {
        if (!(0, payment_codes_js_1.validatePaymentCode)(recipientPaymentCode)) {
            throw new Error('Invalid payment code');
        }
        if (amount <= 0) {
            throw new Error('Amount must be greater than 0');
        }
        this.validateZone(originZone);
        this.validateZone(destinationZone);
        const getDestinationAddresses = async (count) => {
            const addresses = [];
            while (addresses.length < count) {
                const address = this.getNextSendAddress(recipientPaymentCode, destinationZone).address;
                const { isUsed } = await this.checkAddressUse(address);
                if (!isUsed) {
                    addresses.push(address);
                }
            }
            return addresses;
        };
        return this.prepareAndSendTransaction(amount, originZone, getDestinationAddresses);
    }
    async prepareTransaction(selection, inputPubKeys, sendAddresses, changeAddresses, chainId) {
        const tx = new index_js_3.QiTransaction();
        tx.txInputs = selection.inputs.map((input, index) => ({
            txhash: input.txhash,
            index: input.index,
            pubkey: inputPubKeys[index],
        }));
        // 5.3 Create the "sender" outputs
        const senderOutputs = selection.spendOutputs.map((output, index) => ({
            address: sendAddresses[index],
            denomination: output.denomination,
        }));
        // 5.4 Create the "change" outputs
        const changeOutputs = selection.changeOutputs.map((output, index) => ({
            address: changeAddresses[index],
            denomination: output.denomination,
        }));
        tx.txOutputs = [...senderOutputs, ...changeOutputs].map((output) => ({
            address: output.address,
            denomination: output.denomination,
        }));
        tx.chainId = chainId;
        return tx;
    }
    prepareFeeEstimationTransaction(selection, inputPubKeys, sendAddresses, changeAddresses) {
        const txIn = selection.inputs.map((input, index) => ({
            previousOutpoint: { txHash: input.txhash, index: (0, index_js_2.toQuantity)(input.index) },
            pubkey: inputPubKeys[index],
        }));
        // 5.3 Create the "sender" outputs
        const senderOutputs = selection.spendOutputs.map((output, index) => ({
            address: sendAddresses[index],
            denomination: output.denomination,
        }));
        // 5.4 Create the "change" outputs
        const changeOutputs = selection.changeOutputs.map((output, index) => ({
            address: changeAddresses[index],
            denomination: output.denomination,
        }));
        const txOut = [...senderOutputs, ...changeOutputs].map((output) => ({
            address: output.address,
            denomination: (0, index_js_2.toQuantity)(output.denomination),
        }));
        return {
            txType: 2,
            txIn,
            txOut,
        };
    }
    /**
     * Checks the status of pending outpoints and updates the wallet's UTXO set accordingly.
     *
     * @param zone The zone in which to check the pending outpoints.
     */
    async checkPendingOutpoints(zone) {
        // Create a copy to iterate over, as we'll be modifying the _pendingOutpoints array
        const pendingOutpoints = [...this._pendingOutpoints.filter((info) => info.zone === zone)];
        const uniqueAddresses = new Set(pendingOutpoints.map((info) => info.address));
        let outpointsByAddress = [];
        try {
            outpointsByAddress = (await Promise.all(Array.from(uniqueAddresses).map((address) => this.getOutpointsByAddress(address)))).flat();
        }
        catch (error) {
            console.error('Error getting outpoints by address', error);
        }
        const allOutpointsByAddress = outpointsByAddress.flat();
        for (const outpointInfo of pendingOutpoints) {
            const isSpent = !allOutpointsByAddress.some((outpoint) => outpoint.txhash === outpointInfo.outpoint.txhash && outpoint.index === outpointInfo.outpoint.index);
            if (isSpent) {
                // Outpoint has been spent; remove it from pendingOutpoints
                this.removeOutpointFromPending(outpointInfo.outpoint);
            }
            else {
                // Outpoint is still unspent; move it back to available outpoints
                this.moveOutpointToAvailable(outpointInfo);
            }
        }
    }
    /**
     * Moves specified inputs to pending outpoints.
     *
     * @param inputs List of inputs used in the transaction.
     */
    moveOutpointsToPending(inputs) {
        inputs.forEach((input) => {
            const index = this._availableOutpoints.findIndex((outpointInfo) => outpointInfo.outpoint.txhash === input.txhash && outpointInfo.outpoint.index === input.index);
            if (index !== -1) {
                const [outpointInfo] = this._availableOutpoints.splice(index, 1);
                this._pendingOutpoints.push(outpointInfo);
            }
        });
    }
    /**
     * Removes an outpoint from the pending outpoints.
     *
     * @param outpoint The outpoint to remove.
     */
    removeOutpointFromPending(outpoint) {
        this._pendingOutpoints = this._pendingOutpoints.filter((info) => !(info.outpoint.txhash === outpoint.txhash && info.outpoint.index === outpoint.index));
    }
    /**
     * Moves an outpoint from pending back to available outpoints.
     *
     * @param outpointInfo The outpoint info to move.
     */
    moveOutpointToAvailable(outpointInfo) {
        this.removeOutpointFromPending(outpointInfo.outpoint);
        this._availableOutpoints.push(outpointInfo);
    }
    /**
     * Returns a schnorr signature for the given message and private key.
     *
     * @ignore
     * @param {TxInput} input - The transaction input.
     * @param {Uint8Array} hash - The hash of the message.
     * @returns {string} The schnorr signature.
     */
    createSchnorrSignature(input, hash) {
        const privKey = this.getPrivateKeyForTxInput(input);
        const signature = secp256k1_1.schnorr.sign(hash, (0, index_js_2.getBytes)(privKey));
        return (0, index_js_2.hexlify)(signature);
    }
    /**
     * Returns a MuSig signature for the given message and private keys corresponding to the input addresses.
     *
     * @ignore
     * @param {QiTransaction} tx - The Qi transaction.
     * @param {Uint8Array} hash - The hash of the message.
     * @returns {string} The MuSig signature.
     */
    createMuSigSignature(tx, hash) {
        const musig = (0, musig_1.MuSigFactory)(index_js_4.musigCrypto);
        // Collect private keys corresponding to the pubkeys found on the inputs
        const privKeys = tx.txInputs.map((input) => this.getPrivateKeyForTxInput(input));
        // Create an array of public keys corresponding to the private keys for musig aggregation
        const pubKeys = privKeys
            .map((privKey) => index_js_4.musigCrypto.getPublicKey((0, index_js_2.getBytes)(privKey), true))
            .filter((pubKey) => pubKey !== null);
        // Generate nonces for each public key
        const nonces = pubKeys.map((pk) => musig.nonceGen({ publicKey: (0, index_js_2.getBytes)(pk) }));
        const aggNonce = musig.nonceAgg(nonces);
        const signingSession = musig.startSigningSession(aggNonce, hash, pubKeys);
        // Create partial signatures for each private key
        const partialSignatures = privKeys.map((sk, index) => musig.partialSign({
            secretKey: (0, index_js_2.getBytes)(sk || ''),
            publicNonce: nonces[index],
            sessionKey: signingSession,
            verify: true,
        }));
        // Aggregate the partial signatures into a final aggregated signature
        const finalSignature = musig.signAgg(partialSignatures, signingSession);
        return (0, index_js_2.hexlify)(finalSignature);
    }
    /**
     * Retrieves the private key for a given transaction input.
     *
     * This method derives the private key for a transaction input by locating the address info and then deriving the
     * private key based on where the address info was found:
     *
     * - For BIP44 addresses (standard or change), it uses the HD wallet to derive the private key.
     * - For payment channel addresses (BIP47), it uses PaymentCodePrivate to derive the private key.
     *
     * @param {TxInput} input - The transaction input containing the public key.
     * @returns {string} The private key corresponding to the transaction input.
     * @throws {Error} If the input does not contain a public key or if the address information cannot be found.
     */
    getPrivateKeyForTxInput(input) {
        if (!input.pubkey)
            throw new Error('Missing public key for input');
        const address = (0, index_js_1.computeAddress)(input.pubkey);
        return this.getPrivateKey(address);
    }
    /**
     * Returns the private key for a given address. This method should be used with caution as it exposes the private
     * key to the user.
     *
     * @param {string} address - The address associated with the desired private key.
     * @returns {string} The private key.
     */
    getPrivateKey(address) {
        const addressInfo = this.locateAddressInfo(address);
        if (!addressInfo) {
            throw new Error(`Address not found: ${address}`);
        }
        // Handle imported private keys
        if ((0, index_js_2.isHexString)(addressInfo.derivationPath, 32)) {
            return addressInfo.derivationPath;
        }
        if (addressInfo.derivationPath === 'BIP44:external' || addressInfo.derivationPath === 'BIP44:change') {
            // (BIP44 addresses)
            const changeIndex = addressInfo.change ? 1 : 0;
            const addressNode = this._root
                .deriveChild(addressInfo.account + hdwallet_js_1.HARDENED_OFFSET)
                .deriveChild(changeIndex)
                .deriveChild(addressInfo.index);
            return addressNode.privateKey;
        }
        else {
            // (BIP47 addresses)
            const pcAddressInfo = addressInfo;
            const account = pcAddressInfo.account;
            const index = pcAddressInfo.index;
            const counterpartyPaymentCode = pcAddressInfo.derivationPath;
            if (!counterpartyPaymentCode) {
                throw new Error('Counterparty payment code not found for payment channel address');
            }
            const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_2.default);
            const buf = crypto_js_1.bs58check.decode(counterpartyPaymentCode);
            const version = buf[0];
            if (version !== payment_codes_js_1.PC_VERSION)
                throw new Error('Invalid payment code version');
            const counterpartyPCodePublic = new payment_codes_js_1.PaymentCodePublic(secp256k1_2.default, bip32, buf.slice(1));
            const paymentCodePrivate = this._getPaymentCodePrivate(account);
            const paymentPrivateKey = paymentCodePrivate.derivePaymentPrivateKey(counterpartyPCodePublic, index);
            return (0, index_js_2.hexlify)(paymentPrivateKey);
        }
    }
    /**
     * Scans the specified zone for addresses with unspent outputs. Starting at index 0, it will generate new addresses
     * until the gap limit is reached for external and change BIP44 addresses and payment channel addresses.
     *
     * @param {Zone} zone - The zone in which to scan for addresses.
     * @param {number} [account=0] - The index of the account to scan. Default is `0`
     * @returns {Promise<void>} A promise that resolves when the scan is complete.
     * @throws {Error} If the zone is invalid.
     */
    async scan(zone, account = 0) {
        this.validateZone(zone);
        // set status of all addresses to unknown
        this._addressesMap = new Map(Array.from(this._addressesMap.entries()).map(([key, addresses]) => [
            key,
            addresses.map((addr) => ({ ...addr, status: AddressStatus.UNKNOWN })),
        ]));
        // flush available and pending outpoints
        this._availableOutpoints = [];
        this._pendingOutpoints = [];
        // Reset each map so that all keys have empty array values but keys are preserved
        this._paymentCodeSendAddressMap = new Map(Array.from(this._paymentCodeSendAddressMap.keys()).map((key) => [key, []]));
        await this._scan(zone, account);
    }
    /**
     * Scans the specified zone for addresses with unspent outputs. Starting at the last address index, it will generate
     * new addresses until the gap limit is reached for external and change BIP44 addresses and payment channel
     * addresses.
     *
     * @param {Zone} zone - The zone in which to sync addresses.
     * @param {number} [account=0] - The index of the account to sync. Default is `0`
     * @returns {Promise<void>} A promise that resolves when the sync is complete.
     * @throws {Error} If the zone is invalid.
     */
    async sync(zone, account = 0) {
        this.validateZone(zone);
        await this._scan(zone, account);
        await this.checkPendingOutpoints(zone);
    }
    /**
     * Internal method to scan the specified zone for addresses with unspent outputs. This method handles the actual
     * scanning logic, generating new addresses until the gap limit is reached for both gap and change addresses.
     *
     * @param {Zone} zone - The zone in which to scan for addresses.
     * @param {number} [account=0] - The index of the account to scan. Default is `0`
     * @returns {Promise<void>} A promise that resolves when the scan is complete.
     * @throws {Error} If the provider is not set.
     */
    async _scan(zone, account = 0) {
        if (!this.provider)
            throw new Error('Provider not set');
        const derivationPaths = ['BIP44:external', 'BIP44:change', ...this.openChannels];
        await Promise.all(derivationPaths.map((path) => this._scanDerivationPath(path, zone, account)));
    }
    /**
     * Scans for the next address in the specified zone and account, checking for associated outpoints, and updates the
     * address count and gap addresses accordingly.
     *
     * @param {Zone} zone - The zone in which the address is being scanned.
     * @param {number} account - The index of the account for which the address is being scanned.
     * @param {boolean} isChange - A flag indicating whether the address is a change address.
     * @returns {Promise<void>} A promise that resolves when the scan is complete.
     * @throws {Error} If an error occurs during the address scanning or outpoints retrieval process.
     */
    async _scanDerivationPath(path, zone, account) {
        const addresses = this._addressesMap.get(path) || [];
        let consecutiveUnusedCount = 0;
        const checkStatuses = [AddressStatus.UNKNOWN, AddressStatus.ATTEMPTED_USE, AddressStatus.UNUSED];
        // Check existing addresses
        for (let i = 0; i < addresses.length; i++) {
            const addr = addresses[i];
            if (checkStatuses.includes(addr.status)) {
                const { isUsed, outpoints } = await this.checkAddressUse(addr.address);
                addresses[i].status = isUsed ? AddressStatus.USED : AddressStatus.UNUSED;
                // import outpoints if any are found
                if (outpoints.length > 0) {
                    this.importOutpoints(outpoints.map((outpoint) => ({
                        outpoint,
                        address: addr.address,
                        zone: addr.zone,
                        account: addr.account,
                    })));
                }
            }
            if (addr.status === AddressStatus.USED) {
                consecutiveUnusedCount = 0;
            }
            else {
                consecutiveUnusedCount++;
            }
            // If the consecutive unused count has reached the gap limit, break
            if (consecutiveUnusedCount >= QiHDWallet._GAP_LIMIT)
                break;
        }
        // Generate new addresses if needed
        while (consecutiveUnusedCount < QiHDWallet._GAP_LIMIT) {
            const isChange = path.endsWith(':change');
            const newAddrInfo = path.includes('BIP44')
                ? this._getNextQiAddress(account, zone, isChange)
                : this.getNextReceiveAddress(path, zone, account);
            const { isUsed, outpoints } = await this.checkAddressUse(newAddrInfo.address);
            newAddrInfo.status = isUsed ? AddressStatus.USED : AddressStatus.UNUSED;
            // import outpoints if any are found
            if (outpoints.length > 0) {
                this.importOutpoints(outpoints.map((outpoint) => ({
                    outpoint,
                    address: newAddrInfo.address,
                    zone: newAddrInfo.zone,
                    account: newAddrInfo.account,
                })));
            }
            if (newAddrInfo.status === AddressStatus.USED) {
                consecutiveUnusedCount = 0;
            }
            else {
                consecutiveUnusedCount++;
            }
        }
    }
    /**
     * Queries the network node for the outpoints of the specified address.
     *
     * @ignore
     * @param {string} address - The address to query.
     * @returns {Promise<Outpoint[]>} The outpoints for the address.
     * @throws {Error} If the query fails.
     */
    async getOutpointsByAddress(address) {
        try {
            return await this.provider.getOutpointsByAddress(address);
        }
        catch (error) {
            throw new Error(`Failed to get outpoints for address: ${address} - error: ${error}`);
        }
    }
    /**
     * Checks if the specified address is used by querying the network node for the outpoints of the address. If the
     * address is used, the outpoints are imported into the wallet.
     *
     * @param {string} address - The address to check.
     * @returns {Promise<{ isUsed: boolean; outpoints: Outpoint[] }>} A promise that resolves to an object containing a
     *   boolean indicating whether the address is used and an array of outpoints.
     * @throws {Error} If the query fails.
     */
    async checkAddressUse(address) {
        let isUsed = false;
        let outpoints = [];
        try {
            outpoints = await this.getOutpointsByAddress(address);
            if (outpoints.length > 0) {
                isUsed = true;
            }
            else if (this._addressUseChecker !== undefined && (await this._addressUseChecker(address))) {
                // address checker returned true, so the address is used
                isUsed = true;
            }
        }
        catch (error) {
            throw new Error(`Failed to get outpoints for address: ${address} - error: ${error}`);
        }
        return { isUsed, outpoints };
    }
    /**
     * Gets the addresses for the specified zone.
     *
     * @param {Zone} zone - The zone.
     * @returns {QiAddressInfo[]} The addresses for the zone.
     */
    getAddressesForZone(zone) {
        this.validateZone(zone);
        return this._addressesMap.get('BIP44:external')?.filter((addressInfo) => addressInfo.zone === zone) || [];
    }
    /**
     * Gets the change addresses for the specified zone.
     *
     * @param {Zone} zone - The zone.
     * @returns {QiAddressInfo[]} The change addresses for the zone.
     */
    getChangeAddressesForZone(zone) {
        this.validateZone(zone);
        return this._addressesMap.get('BIP44:change')?.filter((addressInfo) => addressInfo.zone === zone) || [];
    }
    /**
     * Gets the gap addresses for the specified zone.
     *
     * @param {Zone} zone - The zone.
     * @returns {QiAddressInfo[]} The gap addresses for the zone.
     */
    getGapAddressesForZone(zone) {
        this.validateZone(zone);
        const gapAddresses = this._addressesMap.get('BIP44:external') || [];
        return gapAddresses.filter((addressInfo) => addressInfo.zone === zone && addressInfo.status === AddressStatus.UNUSED);
    }
    /**
     * Gets the gap change addresses for the specified zone.
     *
     * @param {Zone} zone - The zone.
     * @returns {QiAddressInfo[]} The gap change addresses for the zone.
     */
    getGapChangeAddressesForZone(zone) {
        this.validateZone(zone);
        const gapChangeAddresses = this._addressesMap.get('BIP44:change') || [];
        return gapChangeAddresses.filter((addressInfo) => addressInfo.zone === zone && addressInfo.status === AddressStatus.UNUSED);
    }
    /**
     * Gets the payment channel addresses for the specified zone.
     *
     * @param {string} paymentCode - The payment code.
     * @param {Zone} zone - The zone.
     * @returns {QiAddressInfo[]} The payment channel addresses for the zone.
     */
    getPaymentChannelAddressesForZone(paymentCode, zone) {
        return this._addressesMap.get(paymentCode)?.filter((addressInfo) => addressInfo.zone === zone) || [];
    }
    /**
     * Gets the gap payment channel addresses for the specified payment code.
     *
     * @param {string} paymentCode - The payment code.
     * @returns {QiAddressInfo[]} The gap payment channel addresses for the payment code.
     */
    getGapPaymentChannelAddressesForZone(paymentCode, zone) {
        return (this._addressesMap
            .get(paymentCode)
            ?.filter((addressInfo) => addressInfo.status === AddressStatus.UNUSED && addressInfo.zone === zone) ||
            []);
    }
    /**
     * Signs a message using the private key associated with the given address.
     *
     * @param {string} address - The address for which the message is to be signed.
     * @param {string | Uint8Array} message - The message to be signed, either as a string or Uint8Array.
     * @returns {Promise<string>} A promise that resolves to the signature of the message in hexadecimal string format.
     * @throws {Error} If the address does not correspond to a valid HD node or if signing fails.
     */
    async signMessage(address, message) {
        const privKey = this.getPrivateKey(address);
        const digest = (0, index_js_4.keccak256)(message);
        const signature = secp256k1_1.schnorr.sign(digest, (0, index_js_2.getBytes)(privKey));
        return (0, index_js_2.hexlify)(signature);
    }
    /**
     * Serializes the HD wallet state into a format suitable for storage or transmission.
     *
     * @returns {SerializedQiHDWallet} An object representing the serialized state of the HD wallet, including
     *   outpoints, change addresses, gap addresses, and other inherited properties.
     */
    serialize() {
        const hdwalletSerialized = super.serialize();
        return {
            ...hdwalletSerialized,
            outpoints: this._availableOutpoints,
            pendingOutpoints: this._pendingOutpoints,
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            addresses: Array.from(this._addressesMap.entries()).flatMap(([_, addresses]) => addresses),
            senderPaymentCodeInfo: Object.fromEntries(Array.from(this._paymentCodeSendAddressMap.entries()).map(([key, value]) => [key, Array.from(value)])),
        };
    }
    /**
     * Deserializes a serialized QiHDWallet object and reconstructs the wallet instance.
     *
     * @param {SerializedQiHDWallet} serialized - The serialized object representing the state of a QiHDWallet.
     * @returns {Promise<QiHDWallet>} A promise that resolves to a reconstructed QiHDWallet instance.
     * @throws {Error} If the serialized data is invalid or if any addresses in the gap addresses or gap change
     *   addresses do not exist in the wallet.
     */
    static async deserialize(serialized) {
        super.validateSerializedWallet(serialized);
        // create the wallet instance
        const mnemonic = mnemonic_js_1.Mnemonic.fromPhrase(serialized.phrase);
        const path = this.parentPath(serialized.coinType);
        const root = hdnodewallet_js_1.HDNodeWallet.fromMnemonic(mnemonic, path);
        const wallet = new this(hdwallet_js_1._guard, root);
        const validateQiAddressInfo = (addressInfo) => {
            wallet.validateNeuteredAddressInfo(addressInfo);
            if (!Object.values(AddressStatus).includes(addressInfo.status)) {
                throw new Error(`Invalid QiAddressInfo: status '${addressInfo.status}' is not a valid AddressStatus`);
            }
            if (addressInfo.derivationPath !== 'BIP44:external' &&
                addressInfo.derivationPath !== 'BIP44:change' &&
                !(0, payment_codes_js_1.validatePaymentCode)(addressInfo.derivationPath)) {
                throw new Error(`Invalid QiAddressInfo: derivationPath '${addressInfo.derivationPath}' is not valid. It should be 'BIP44:external', 'BIP44:change', or a valid BIP47 payment code`);
            }
        };
        // First, group addresses by derivation path
        const addressesByPath = new Map();
        for (const addressInfo of serialized.addresses) {
            validateQiAddressInfo(addressInfo);
            let key = addressInfo.derivationPath;
            if ((0, index_js_2.isHexString)(key, 32)) {
                key = QiHDWallet.PRIVATE_KEYS_PATH;
            }
            if (!addressesByPath.has(key)) {
                addressesByPath.set(key, []);
            }
            addressesByPath.get(key).push(addressInfo);
        }
        // Then, set all paths in the wallet's address map
        for (const [key, addresses] of addressesByPath) {
            wallet._addressesMap.set(key, addresses);
        }
        // validate and import the counter party payment code info
        for (const [paymentCode, paymentCodeInfoArray] of Object.entries(serialized.senderPaymentCodeInfo)) {
            if (!(0, payment_codes_js_1.validatePaymentCode)(paymentCode)) {
                throw new Error(`Invalid payment code: ${paymentCode}`);
            }
            for (const pcInfo of paymentCodeInfoArray) {
                validateQiAddressInfo(pcInfo);
            }
            wallet._paymentCodeSendAddressMap.set(paymentCode, paymentCodeInfoArray);
        }
        // validate the available outpoints and import them
        wallet.validateOutpointInfo(serialized.outpoints);
        wallet._availableOutpoints.push(...serialized.outpoints);
        // validate the pending outpoints and import them
        wallet.validateOutpointInfo(serialized.pendingOutpoints);
        wallet._pendingOutpoints.push(...serialized.pendingOutpoints);
        return wallet;
    }
    /**
     * Validates an array of OutpointInfo objects. This method checks the validity of each OutpointInfo object by
     * performing the following validations:
     *
     * - Validates the zone using the `validateZone` method.
     * - Checks if the address exists in the wallet.
     * - Checks if the account (if provided) exists in the wallet.
     * - Validates the Outpoint by ensuring that `Txhash`, `Index`, and `Denomination` are not null.
     *
     * @ignore
     * @param {OutpointInfo[]} outpointInfo - An array of OutpointInfo objects to be validated.
     * @throws {Error} If any of the validations fail, an error is thrown with a descriptive message.
     */
    validateOutpointInfo(outpointInfo) {
        outpointInfo.forEach((info) => {
            // validate zone
            this.validateZone(info.zone);
            // validate address and account
            this.validateAddressAndAccount(info.address, info.account);
            // validate Outpoint
            if (info.outpoint.txhash == null || info.outpoint.index == null || info.outpoint.denomination == null) {
                throw new Error(`Invalid Outpoint: ${JSON.stringify(info)} `);
            }
        });
    }
    validateAddressAndAccount(address, account) {
        const addressInfo = this.locateAddressInfo(address);
        if (!addressInfo) {
            throw new Error(`Address ${address} not found in wallet`);
        }
        if (account && account !== addressInfo.account) {
            throw new Error(`Address ${address} does not match account ${account}`);
        }
    }
    /**
     * Creates a new BIP47 payment code for the specified account. The payment code is derived from the account's BIP32
     * root key.
     *
     * @param {number} account - The account index to derive the payment code from.
     * @returns {Promise<string>} A promise that resolves to the Base58-encoded BIP47 payment code.
     */
    getPaymentCode(account = 0) {
        const privatePcode = this._getPaymentCodePrivate(account);
        return privatePcode.toBase58();
    }
    /**
     * Generates a BIP47 private payment code for the specified account. The payment code is created by combining the
     * account's public key and chain code.
     *
     * @private
     * @param {number} account - The account index for which to generate the private payment code.
     * @returns {Promise<PaymentCodePrivate>} A promise that resolves to the PaymentCodePrivate instance.
     */
    _getPaymentCodePrivate(account) {
        const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_2.default);
        const accountNode = this._root.deriveChild(account + hdwallet_js_1.HARDENED_OFFSET);
        // payment code array
        const pc = new Uint8Array(80);
        // set version + options
        pc.set([1, 0]);
        // set the public key
        const pubKey = accountNode.publicKey;
        pc.set((0, index_js_2.getBytes)(pubKey), 2);
        // set the chain code
        const chainCode = accountNode.chainCode;
        pc.set((0, index_js_2.getBytes)(chainCode), 35);
        const adapter = new types_js_1.HDNodeBIP32Adapter(accountNode);
        return new payment_codes_js_1.PaymentCodePrivate(adapter, secp256k1_2.default, bip32, pc);
    }
    /**
     * Generates a payment address for sending funds to the specified receiver's BIP47 payment code. Uses Diffie-Hellman
     * key exchange to derive the address from the receiver's public key and sender's private key.
     *
     * @param {string} receiverPaymentCode - The Base58-encoded BIP47 payment code of the receiver.
     * @returns {Promise<string>} A promise that resolves to the payment address for sending funds.
     * @throws {Error} Throws an error if the payment code version is invalid.
     */
    getNextSendAddress(receiverPaymentCode, zone, account = 0) {
        const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_2.default);
        const buf = crypto_js_1.bs58check.decode(receiverPaymentCode);
        const version = buf[0];
        if (version !== payment_codes_js_1.PC_VERSION)
            throw new Error('Invalid payment code version');
        const walletPCodePrivate = this._getPaymentCodePrivate(account);
        const receiverPCodePublic = new payment_codes_js_1.PaymentCodePublic(secp256k1_2.default, bip32, buf.slice(1));
        const paymentCodeInfoArray = this._paymentCodeSendAddressMap.get(receiverPaymentCode);
        const lastIndex = this._findLastUsedIndex(paymentCodeInfoArray, account, zone);
        let addrIndex = lastIndex + 1;
        for (let attempts = 0; attempts < hdwallet_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS; attempts++) {
            const address = receiverPCodePublic.getPaymentAddress(walletPCodePrivate, addrIndex);
            if (this.isValidAddressForZone(address, zone)) {
                const pubkey = receiverPCodePublic.derivePaymentPublicKey(walletPCodePrivate, addrIndex);
                const pcInfo = {
                    address,
                    pubKey: (0, index_js_2.hexlify)(pubkey),
                    index: addrIndex,
                    account,
                    zone,
                    change: false,
                    status: AddressStatus.UNUSED,
                    derivationPath: receiverPaymentCode,
                };
                if (paymentCodeInfoArray) {
                    paymentCodeInfoArray.push(pcInfo);
                }
                else {
                    this._paymentCodeSendAddressMap.set(receiverPaymentCode, [pcInfo]);
                }
                return pcInfo;
            }
            addrIndex++;
        }
        throw new Error(`Failed to derive a valid address for the zone ${zone} after ${hdwallet_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS} attempts.`);
    }
    /**
     * Generates a payment address for receiving funds from the specified sender's BIP47 payment code. Uses
     * Diffie-Hellman key exchange to derive the address from the sender's public key and receiver's private key.
     *
     * @param {string} senderPaymentCode - The Base58-encoded BIP47 payment code of the sender.
     * @returns {Promise<string>} A promise that resolves to the payment address for receiving funds.
     * @throws {Error} Throws an error if the payment code version is invalid.
     */
    getNextReceiveAddress(senderPaymentCode, zone, account = 0) {
        const bip32 = (0, bip32_js_1.BIP32Factory)(secp256k1_2.default);
        const buf = crypto_js_1.bs58check.decode(senderPaymentCode);
        const version = buf[0];
        if (version !== payment_codes_js_1.PC_VERSION)
            throw new Error('Invalid payment code version');
        const senderPCodePublic = new payment_codes_js_1.PaymentCodePublic(secp256k1_2.default, bip32, buf.slice(1));
        const walletPCodePrivate = this._getPaymentCodePrivate(account);
        const paymentCodeInfoArray = this._addressesMap.get(senderPaymentCode);
        const lastIndex = this._findLastUsedIndex(paymentCodeInfoArray, account, zone);
        let addrIndex = lastIndex + 1;
        for (let attempts = 0; attempts < hdwallet_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS; attempts++) {
            const address = walletPCodePrivate.getPaymentAddress(senderPCodePublic, addrIndex);
            if (this.isValidAddressForZone(address, zone)) {
                const pubkey = walletPCodePrivate.derivePaymentPublicKey(senderPCodePublic, addrIndex);
                const pcInfo = {
                    address,
                    pubKey: (0, index_js_2.hexlify)(pubkey),
                    index: addrIndex,
                    account,
                    zone,
                    change: false,
                    status: AddressStatus.UNUSED,
                    derivationPath: senderPaymentCode,
                };
                if (paymentCodeInfoArray) {
                    paymentCodeInfoArray.push(pcInfo);
                }
                else {
                    this._addressesMap.set(senderPaymentCode, [pcInfo]);
                }
                return pcInfo;
            }
            addrIndex++;
        }
        throw new Error(`Failed to derive a valid address for the zone ${zone} after ${hdwallet_js_1.MAX_ADDRESS_DERIVATION_ATTEMPTS} attempts.`);
    }
    /**
     * Receives a payment code and stores it in the wallet for future use. If the payment code is already in the wallet,
     * it will be ignored.
     *
     * @param {string} paymentCode - The payment code to store.
     */
    openChannel(paymentCode) {
        if (!(0, payment_codes_js_1.validatePaymentCode)(paymentCode)) {
            throw new Error(`Invalid payment code: ${paymentCode}`);
        }
        if (!this._addressesMap.has(paymentCode)) {
            this._addressesMap.set(paymentCode, []);
        }
        if (!this._paymentCodeSendAddressMap.has(paymentCode)) {
            this._paymentCodeSendAddressMap.set(paymentCode, []);
        }
    }
    channelIsOpen(paymentCode) {
        return this._addressesMap.has(paymentCode) && this._paymentCodeSendAddressMap.has(paymentCode);
    }
    /**
     * Gets the address info for a given address.
     *
     * @param {string} address - The address.
     * @returns {QiAddressInfo | null} The address info or null if not found.
     */
    getAddressInfo(address) {
        const externalAddressInfo = this._addressesMap.get('BIP44:external')?.find((addr) => addr.address === address);
        if (!externalAddressInfo) {
            return null;
        }
        return externalAddressInfo;
    }
    /**
     * Gets the address info for a given address.
     *
     * @param {string} address - The address.
     * @returns {QiAddressInfo | null} The address info or null if not found.
     */
    getChangeAddressInfo(address) {
        const changeAddressInfo = this._addressesMap.get('BIP44:change')?.find((addr) => addr.address === address);
        if (!changeAddressInfo) {
            return null;
        }
        return changeAddressInfo;
    }
    /**
     * Imports a private key and adds it to the wallet.
     *
     * @param {string} privateKey - The private key to import (hex string)
     * @returns {Promise<QiAddressInfo>} The address information for the imported key
     * @throws {Error} If the private key is invalid or the address is already in use
     */
    async importPrivateKey(privateKey) {
        if (!(0, index_js_2.isHexString)(privateKey, 32)) {
            throw new Error(`Invalid private key format: must be 32-byte hex string (got ${privateKey})`);
        }
        const pubKey = index_js_4.SigningKey.computePublicKey(privateKey, true);
        const address = (0, index_js_1.computeAddress)(pubKey);
        // Validate address is for correct zone and ledger
        const addressZone = (0, index_js_2.getZoneForAddress)(address);
        if (!addressZone) {
            throw new Error(`Private key does not correspond to a valid address for any zone (got ${address})`);
        }
        if (!(0, index_js_1.isQiAddress)(address)) {
            throw new Error(`Private key does not correspond to a valid Qi address (got ${address})`);
        }
        for (const [path, addresses] of this._addressesMap.entries()) {
            if (addresses.some((info) => info.address === address)) {
                throw new Error(`Address ${address} already exists in wallet under path ${path}`);
            }
        }
        const addressInfo = {
            pubKey,
            address,
            account: 0,
            index: -1,
            change: false,
            zone: addressZone,
            status: AddressStatus.UNUSED,
            derivationPath: privateKey, // Store private key in derivationPath
        };
        this._addressesMap.get(QiHDWallet.PRIVATE_KEYS_PATH).push(addressInfo);
        return addressInfo;
    }
    /**
     * Gets all addresses that were imported via private keys.
     *
     * @param {Zone} [zone] - Optional zone to filter addresses by
     * @returns {QiAddressInfo[]} Array of address info objects for imported addresses
     */
    getImportedAddresses(zone) {
        const importedAddresses = this._addressesMap.get(QiHDWallet.PRIVATE_KEYS_PATH) || [];
        if (zone !== undefined) {
            this.validateZone(zone);
            return importedAddresses.filter((info) => info.zone === zone);
        }
        return [...importedAddresses];
    }
    /**
     * Adds a new address to the wallet.
     *
     * @param {number} account - The account number.
     * @param {number} addressIndex - The address index.
     * @returns {QiAddressInfo} The address info for the new address.
     */
    addAddress(account, addressIndex) {
        return this._addAddress(account, addressIndex, false);
    }
    /**
     * Adds a new change address to the wallet.
     *
     * @param {number} account - The account number.
     * @param {number} addressIndex - The address index.
     * @returns {QiAddressInfo} The address info for the new address.
     */
    addChangeAddress(account, addressIndex) {
        return this._addAddress(account, addressIndex, true);
    }
    _addAddress(account, addressIndex, isChange) {
        const derivationPath = isChange ? 'BIP44:change' : 'BIP44:external';
        const existingAddresses = this._addressesMap.get(derivationPath) || [];
        if (existingAddresses.some((info) => info.index === addressIndex)) {
            throw new Error(`Address index ${addressIndex} already exists in wallet under path ${derivationPath}`);
        }
        const addressNode = this._root
            .deriveChild(account + hdwallet_js_1.HARDENED_OFFSET)
            .deriveChild(isChange ? 1 : 0)
            .deriveChild(addressIndex);
        const zone = (0, index_js_2.getZoneForAddress)(addressNode.address);
        if (!zone) {
            throw new Error(`Failed to derive a Qi valid address zone for the index ${addressIndex}`);
        }
        if (!(0, index_js_1.isQiAddress)(addressNode.address)) {
            throw new Error(`Address ${addressNode.address} is not a valid Qi address`);
        }
        const addressInfo = {
            pubKey: addressNode.publicKey,
            address: addressNode.address,
            account,
            index: addressIndex,
            change: isChange,
            zone,
            status: AddressStatus.UNUSED,
            derivationPath,
        };
        const addresses = this._addressesMap.get(derivationPath);
        if (!addresses) {
            this._addressesMap.set(derivationPath, [addressInfo]);
        }
        else {
            addresses.push(addressInfo);
        }
        return addressInfo;
    }
    /**
     * Gets the addresses for a given account.
     *
     * @param {number} account - The account number.
     * @returns {QiAddressInfo[]} The addresses for the account.
     */
    getAddressesForAccount(account) {
        const addresses = this._addressesMap.values();
        return Array.from(addresses)
            .flat()
            .filter((info) => info.account === account);
    }
}
exports.QiHDWallet = QiHDWallet;
