// @goat-sdk/core/dist/chunk-OZ37CS44.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/plugins/send-eth.ts
var _viem = require('viem');
var _zod = require('zod');
function sendETH() {
  return {
    name: "send_eth",
    supportsSmartWallets: () => true,
    supportsChain: (chain) => chain.type === "evm",
    getTools: async () => {
      return [
        {
          name: "send_eth",
          description: "This {{tool}} sends ETH to an address on an EVM chain.",
          parameters: sendETHParametersSchema,
          method: sendETHMethod
        }
      ];
    }
  };
}
var sendETHParametersSchema = _zod.z.object({
  to: _zod.z.string().describe("The address to send ETH to"),
  amount: _zod.z.string().describe("The amount of ETH to send")
});
async function sendETHMethod(walletClient, parameters) {
  try {
    const amount = _viem.parseEther.call(void 0, parameters.amount);
    const tx = await walletClient.sendTransaction({
      to: parameters.to,
      value: amount
    });
    return tx.hash;
  } catch (error) {
    throw new Error(`Failed to send ETH: ${error}`);
  }
}



exports.sendETH = sendETH;
