// @tiplink/api/dist/escrow/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INDEXER_URL_BASE = exports.PRIO_FEES_LAMPORTS = exports.DEPOSIT_URL_BASE = exports.PDA_SEED = exports.TREASURY_PUBLIC_KEY = exports.ESCROW_PROGRAM_ID = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.ESCROW_PROGRAM_ID = new web3_js_1.PublicKey("8TqqugH88U3fDEWeKHqBSxZKeqoRrXkdpy3ciX5GAruK");
exports.TREASURY_PUBLIC_KEY = new web3_js_1.PublicKey("BGZMcTjyTCbkRszC1CBpFpP9CbVh3Ah2ZhjzCsc9PsAr");
exports.PDA_SEED = "escrow";
const DEFAULT_DEPOSIT_URL_BASE = "https://tiplink-mailer.vercel.app/depositor-url";
exports.DEPOSIT_URL_BASE = process !== undefined && process.env !== undefined
    ? process.env.NEXT_PUBLIC_ESCROW_DEPOSITOR_URL_OVERRIDE ||
        DEFAULT_DEPOSIT_URL_BASE
    : DEFAULT_DEPOSIT_URL_BASE;
exports.PRIO_FEES_LAMPORTS = 10000;
const DEFAULT_INDEXER_URL_BASE = "https://backend.tiplink.io";
exports.INDEXER_URL_BASE = process !== undefined && process.env !== undefined
    ? process.env.NEXT_PUBLIC_INDEXER_URL_OVERRIDE || DEFAULT_INDEXER_URL_BASE
    : DEFAULT_INDEXER_URL_BASE;
