// interchain/dist/codegen/cosmos/distribution/v1beta1/query.rpc.Query.js
import { BinaryReader } from "../../../binary";
import { createProtobufRpcClient } from "@cosmjs/stargate";
import { QueryParamsRequest, QueryParamsResponse, QueryValidatorDistributionInfoRequest, QueryValidatorDistributionInfoResponse, QueryValidatorOutstandingRewardsRequest, QueryValidatorOutstandingRewardsResponse, QueryValidatorCommissionRequest, QueryValidatorCommissionResponse, QueryValidatorSlashesRequest, QueryValidatorSlashesResponse, QueryDelegationRewardsRequest, QueryDelegationRewardsResponse, QueryDelegationTotalRewardsRequest, QueryDelegationTotalRewardsResponse, QueryDelegatorValidatorsRequest, QueryDelegatorValidatorsResponse, QueryDelegatorWithdrawAddressRequest, QueryDelegatorWithdrawAddressResponse, QueryCommunityPoolRequest, QueryCommunityPoolResponse } from "./query";
export class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* Params queries params of the distribution module. */
    params = async (request = {}) => {
        const data = QueryParamsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "Params", data);
        return promise.then(data => QueryParamsResponse.decode(new BinaryReader(data)));
    };
    /* ValidatorDistributionInfo queries validator commission and self-delegation rewards for validator */
    validatorDistributionInfo = async (request) => {
        const data = QueryValidatorDistributionInfoRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "ValidatorDistributionInfo", data);
        return promise.then(data => QueryValidatorDistributionInfoResponse.decode(new BinaryReader(data)));
    };
    /* ValidatorOutstandingRewards queries rewards of a validator address. */
    validatorOutstandingRewards = async (request) => {
        const data = QueryValidatorOutstandingRewardsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "ValidatorOutstandingRewards", data);
        return promise.then(data => QueryValidatorOutstandingRewardsResponse.decode(new BinaryReader(data)));
    };
    /* ValidatorCommission queries accumulated commission for a validator. */
    validatorCommission = async (request) => {
        const data = QueryValidatorCommissionRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "ValidatorCommission", data);
        return promise.then(data => QueryValidatorCommissionResponse.decode(new BinaryReader(data)));
    };
    /* ValidatorSlashes queries slash events of a validator. */
    validatorSlashes = async (request) => {
        const data = QueryValidatorSlashesRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "ValidatorSlashes", data);
        return promise.then(data => QueryValidatorSlashesResponse.decode(new BinaryReader(data)));
    };
    /* DelegationRewards queries the total rewards accrued by a delegation. */
    delegationRewards = async (request) => {
        const data = QueryDelegationRewardsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "DelegationRewards", data);
        return promise.then(data => QueryDelegationRewardsResponse.decode(new BinaryReader(data)));
    };
    /* DelegationTotalRewards queries the total rewards accrued by each
     validator. */
    delegationTotalRewards = async (request) => {
        const data = QueryDelegationTotalRewardsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "DelegationTotalRewards", data);
        return promise.then(data => QueryDelegationTotalRewardsResponse.decode(new BinaryReader(data)));
    };
    /* DelegatorValidators queries the validators of a delegator. */
    delegatorValidators = async (request) => {
        const data = QueryDelegatorValidatorsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "DelegatorValidators", data);
        return promise.then(data => QueryDelegatorValidatorsResponse.decode(new BinaryReader(data)));
    };
    /* DelegatorWithdrawAddress queries withdraw address of a delegator. */
    delegatorWithdrawAddress = async (request) => {
        const data = QueryDelegatorWithdrawAddressRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "DelegatorWithdrawAddress", data);
        return promise.then(data => QueryDelegatorWithdrawAddressResponse.decode(new BinaryReader(data)));
    };
    /* CommunityPool queries the community pool coins. */
    communityPool = async (request = {}) => {
        const data = QueryCommunityPoolRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.distribution.v1beta1.Query", "CommunityPool", data);
        return promise.then(data => QueryCommunityPoolResponse.decode(new BinaryReader(data)));
    };
}
export const createRpcQueryExtension = (base) => {
    const rpc = createProtobufRpcClient(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        params(request) {
            return queryService.params(request);
        },
        validatorDistributionInfo(request) {
            return queryService.validatorDistributionInfo(request);
        },
        validatorOutstandingRewards(request) {
            return queryService.validatorOutstandingRewards(request);
        },
        validatorCommission(request) {
            return queryService.validatorCommission(request);
        },
        validatorSlashes(request) {
            return queryService.validatorSlashes(request);
        },
        delegationRewards(request) {
            return queryService.delegationRewards(request);
        },
        delegationTotalRewards(request) {
            return queryService.delegationTotalRewards(request);
        },
        delegatorValidators(request) {
            return queryService.delegatorValidators(request);
        },
        delegatorWithdrawAddress(request) {
            return queryService.delegatorWithdrawAddress(request);
        },
        communityPool(request) {
            return queryService.communityPool(request);
        }
    };
};
