// @lifi/sdk/src/_cjs/errors/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = exports.ServerError = exports.BalanceError = exports.UnknownError = exports.TransactionError = exports.ProviderError = exports.RPCError = void 0;
const baseError_js_1 = require("./baseError.js");
const constants_js_1 = require("./constants.js");
class RPCError extends baseError_js_1.BaseError {
    constructor(code, message, cause) {
        super(constants_js_1.ErrorName.RPCError, code, message, cause);
    }
}
exports.RPCError = RPCError;
class ProviderError extends baseError_js_1.BaseError {
    constructor(code, message, cause) {
        super(constants_js_1.ErrorName.ProviderError, code, message, cause);
    }
}
exports.ProviderError = ProviderError;
class TransactionError extends baseError_js_1.BaseError {
    constructor(code, message, cause) {
        super(constants_js_1.ErrorName.TransactionError, code, message, cause);
    }
}
exports.TransactionError = TransactionError;
class UnknownError extends baseError_js_1.BaseError {
    constructor(message, cause) {
        super(constants_js_1.ErrorName.UnknownError, constants_js_1.LiFiErrorCode.InternalError, message, cause);
    }
}
exports.UnknownError = UnknownError;
class BalanceError extends baseError_js_1.BaseError {
    constructor(message, cause) {
        super(constants_js_1.ErrorName.BalanceError, constants_js_1.LiFiErrorCode.BalanceError, message, cause);
    }
}
exports.BalanceError = BalanceError;
class ServerError extends baseError_js_1.BaseError {
    constructor(message) {
        super(constants_js_1.ErrorName.ServerError, constants_js_1.LiFiErrorCode.InternalError, message);
    }
}
exports.ServerError = ServerError;
class ValidationError extends baseError_js_1.BaseError {
    constructor(message) {
        super(constants_js_1.ErrorName.ValidationError, constants_js_1.LiFiErrorCode.ValidationError, message);
    }
}
exports.ValidationError = ValidationError;
