// @lifi/sdk/src/_cjs/errors/SDKError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDKError = void 0;
const version_js_1 = require("../version.js");
class SDKError extends Error {
    constructor(cause, step, process) {
        const errorMessage = `${cause.message ? `[${cause.name}] ${cause.message}` : 'Unknown error occurred'}\nLI.FI SDK version: ${version_js_1.version}`;
        super(errorMessage);
        Object.defineProperty(this, "step", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "process", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: 'SDKError'
        });
        Object.defineProperty(this, "cause", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'SDKError';
        this.step = step;
        this.process = process;
        this.cause = cause;
        this.stack = this.cause.stack;
        this.code = cause.code;
    }
}
exports.SDKError = SDKError;
