// interchain/dist/cjs/codegen/ibc/core/channel/v1/query.rpc.Query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
const binary_1 = require("../../../../binary");
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("./query");
class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* Channel queries an IBC Channel. */
    channel = async (request) => {
        const data = query_1.QueryChannelRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "Channel", data);
        return promise.then(data => query_1.QueryChannelResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* Channels queries all the IBC channels of a chain. */
    channels = async (request = {
        pagination: undefined
    }) => {
        const data = query_1.QueryChannelsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "Channels", data);
        return promise.then(data => query_1.QueryChannelsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConnectionChannels queries all the channels associated with a connection
     end. */
    connectionChannels = async (request) => {
        const data = query_1.QueryConnectionChannelsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "ConnectionChannels", data);
        return promise.then(data => query_1.QueryConnectionChannelsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ChannelClientState queries for the client state for the channel associated
     with the provided channel identifiers. */
    channelClientState = async (request) => {
        const data = query_1.QueryChannelClientStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "ChannelClientState", data);
        return promise.then(data => query_1.QueryChannelClientStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ChannelConsensusState queries for the consensus state for the channel
     associated with the provided channel identifiers. */
    channelConsensusState = async (request) => {
        const data = query_1.QueryChannelConsensusStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "ChannelConsensusState", data);
        return promise.then(data => query_1.QueryChannelConsensusStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* PacketCommitment queries a stored packet commitment hash. */
    packetCommitment = async (request) => {
        const data = query_1.QueryPacketCommitmentRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketCommitment", data);
        return promise.then(data => query_1.QueryPacketCommitmentResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* PacketCommitments returns all the packet commitments hashes associated
     with a channel. */
    packetCommitments = async (request) => {
        const data = query_1.QueryPacketCommitmentsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketCommitments", data);
        return promise.then(data => query_1.QueryPacketCommitmentsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* PacketReceipt queries if a given packet sequence has been received on the
     queried chain */
    packetReceipt = async (request) => {
        const data = query_1.QueryPacketReceiptRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketReceipt", data);
        return promise.then(data => query_1.QueryPacketReceiptResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* PacketAcknowledgement queries a stored packet acknowledgement hash. */
    packetAcknowledgement = async (request) => {
        const data = query_1.QueryPacketAcknowledgementRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketAcknowledgement", data);
        return promise.then(data => query_1.QueryPacketAcknowledgementResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* PacketAcknowledgements returns all the packet acknowledgements associated
     with a channel. */
    packetAcknowledgements = async (request) => {
        const data = query_1.QueryPacketAcknowledgementsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "PacketAcknowledgements", data);
        return promise.then(data => query_1.QueryPacketAcknowledgementsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* UnreceivedPackets returns all the unreceived IBC packets associated with a
     channel and sequences. */
    unreceivedPackets = async (request) => {
        const data = query_1.QueryUnreceivedPacketsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "UnreceivedPackets", data);
        return promise.then(data => query_1.QueryUnreceivedPacketsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* UnreceivedAcks returns all the unreceived IBC acknowledgements associated
     with a channel and sequences. */
    unreceivedAcks = async (request) => {
        const data = query_1.QueryUnreceivedAcksRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "UnreceivedAcks", data);
        return promise.then(data => query_1.QueryUnreceivedAcksResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* NextSequenceReceive returns the next receive sequence for a given channel. */
    nextSequenceReceive = async (request) => {
        const data = query_1.QueryNextSequenceReceiveRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "NextSequenceReceive", data);
        return promise.then(data => query_1.QueryNextSequenceReceiveResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* NextSequenceSend returns the next send sequence for a given channel. */
    nextSequenceSend = async (request) => {
        const data = query_1.QueryNextSequenceSendRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.channel.v1.Query", "NextSequenceSend", data);
        return promise.then(data => query_1.QueryNextSequenceSendResponse.decode(new binary_1.BinaryReader(data)));
    };
}
exports.QueryClientImpl = QueryClientImpl;
const createRpcQueryExtension = (base) => {
    const rpc = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        channel(request) {
            return queryService.channel(request);
        },
        channels(request) {
            return queryService.channels(request);
        },
        connectionChannels(request) {
            return queryService.connectionChannels(request);
        },
        channelClientState(request) {
            return queryService.channelClientState(request);
        },
        channelConsensusState(request) {
            return queryService.channelConsensusState(request);
        },
        packetCommitment(request) {
            return queryService.packetCommitment(request);
        },
        packetCommitments(request) {
            return queryService.packetCommitments(request);
        },
        packetReceipt(request) {
            return queryService.packetReceipt(request);
        },
        packetAcknowledgement(request) {
            return queryService.packetAcknowledgement(request);
        },
        packetAcknowledgements(request) {
            return queryService.packetAcknowledgements(request);
        },
        unreceivedPackets(request) {
            return queryService.unreceivedPackets(request);
        },
        unreceivedAcks(request) {
            return queryService.unreceivedAcks(request);
        },
        nextSequenceReceive(request) {
            return queryService.nextSequenceReceive(request);
        },
        nextSequenceSend(request) {
            return queryService.nextSequenceSend(request);
        }
    };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
