// @akashnetwork/akash-api/dist/generated/akash/deployment/v1beta1/params.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = void 0;
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const coin_1 = require("../../../cosmos/base/v1beta1/coin");
const typeRegistry_1 = require("../../../typeRegistry");
function createBaseParams() {
    return {
        $type: 'akash.deployment.v1beta1.Params',
        deploymentMinDeposit: undefined,
    };
}
exports.Params = {
    $type: 'akash.deployment.v1beta1.Params',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.deploymentMinDeposit !== undefined) {
            coin_1.Coin.encode(message.deploymentMinDeposit, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.deploymentMinDeposit = coin_1.Coin.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.Params.$type,
            deploymentMinDeposit: isSet(object.deploymentMinDeposit)
                ? coin_1.Coin.fromJSON(object.deploymentMinDeposit)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.deploymentMinDeposit !== undefined) {
            obj.deploymentMinDeposit = coin_1.Coin.toJSON(message.deploymentMinDeposit);
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseParams();
        message.deploymentMinDeposit =
            object.deploymentMinDeposit !== undefined &&
                object.deploymentMinDeposit !== null
                ? coin_1.Coin.fromPartial(object.deploymentMinDeposit)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.Params.$type, exports.Params);
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
