// quais/lib/commonjs/constants/index.js
"use strict";
/**
 * Some common constants useful for Ethereum.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneData = exports.Zone = exports.toZone = exports.Ledger = exports.ShardData = exports.Shard = exports.toShard = exports.MessagePrefix = exports.quaisymbol = exports.MaxInt256 = exports.MinInt256 = exports.MaxUint256 = exports.WeiPerEther = exports.N = exports.ZeroHash = exports.ZeroAddress = void 0;
var addresses_js_1 = require("./addresses.js");
Object.defineProperty(exports, "ZeroAddress", { enumerable: true, get: function () { return addresses_js_1.ZeroAddress; } });
var hashes_js_1 = require("./hashes.js");
Object.defineProperty(exports, "ZeroHash", { enumerable: true, get: function () { return hashes_js_1.ZeroHash; } });
var numbers_js_1 = require("./numbers.js");
Object.defineProperty(exports, "N", { enumerable: true, get: function () { return numbers_js_1.N; } });
Object.defineProperty(exports, "WeiPerEther", { enumerable: true, get: function () { return numbers_js_1.WeiPerEther; } });
Object.defineProperty(exports, "MaxUint256", { enumerable: true, get: function () { return numbers_js_1.MaxUint256; } });
Object.defineProperty(exports, "MinInt256", { enumerable: true, get: function () { return numbers_js_1.MinInt256; } });
Object.defineProperty(exports, "MaxInt256", { enumerable: true, get: function () { return numbers_js_1.MaxInt256; } });
var strings_js_1 = require("./strings.js");
Object.defineProperty(exports, "quaisymbol", { enumerable: true, get: function () { return strings_js_1.quaisymbol; } });
Object.defineProperty(exports, "MessagePrefix", { enumerable: true, get: function () { return strings_js_1.MessagePrefix; } });
var shards_js_1 = require("./shards.js");
Object.defineProperty(exports, "toShard", { enumerable: true, get: function () { return shards_js_1.toShard; } });
Object.defineProperty(exports, "Shard", { enumerable: true, get: function () { return shards_js_1.Shard; } });
Object.defineProperty(exports, "ShardData", { enumerable: true, get: function () { return shards_js_1.ShardData; } });
var zones_js_1 = require("./zones.js");
Object.defineProperty(exports, "Ledger", { enumerable: true, get: function () { return zones_js_1.Ledger; } });
Object.defineProperty(exports, "toZone", { enumerable: true, get: function () { return zones_js_1.toZone; } });
Object.defineProperty(exports, "Zone", { enumerable: true, get: function () { return zones_js_1.Zone; } });
Object.defineProperty(exports, "ZoneData", { enumerable: true, get: function () { return zones_js_1.ZoneData; } });
