// interchain/dist/cjs/codegen/ibc/core/client/v1/query.rpc.Query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
const binary_1 = require("../../../../binary");
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("./query");
class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* ClientState queries an IBC light client. */
    clientState = async (request) => {
        const data = query_1.QueryClientStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ClientState", data);
        return promise.then(data => query_1.QueryClientStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ClientStates queries all the IBC light clients of a chain. */
    clientStates = async (request = {
        pagination: undefined
    }) => {
        const data = query_1.QueryClientStatesRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ClientStates", data);
        return promise.then(data => query_1.QueryClientStatesResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConsensusState queries a consensus state associated with a client state at
     a given height. */
    consensusState = async (request) => {
        const data = query_1.QueryConsensusStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ConsensusState", data);
        return promise.then(data => query_1.QueryConsensusStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConsensusStates queries all the consensus state associated with a given
     client. */
    consensusStates = async (request) => {
        const data = query_1.QueryConsensusStatesRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ConsensusStates", data);
        return promise.then(data => query_1.QueryConsensusStatesResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ConsensusStateHeights queries the height of every consensus states associated with a given client. */
    consensusStateHeights = async (request) => {
        const data = query_1.QueryConsensusStateHeightsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ConsensusStateHeights", data);
        return promise.then(data => query_1.QueryConsensusStateHeightsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* Status queries the status of an IBC client. */
    clientStatus = async (request) => {
        const data = query_1.QueryClientStatusRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ClientStatus", data);
        return promise.then(data => query_1.QueryClientStatusResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ClientParams queries all parameters of the ibc client submodule. */
    clientParams = async (request = {}) => {
        const data = query_1.QueryClientParamsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "ClientParams", data);
        return promise.then(data => query_1.QueryClientParamsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* UpgradedClientState queries an Upgraded IBC light client. */
    upgradedClientState = async (request = {}) => {
        const data = query_1.QueryUpgradedClientStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "UpgradedClientState", data);
        return promise.then(data => query_1.QueryUpgradedClientStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* UpgradedConsensusState queries an Upgraded IBC consensus state. */
    upgradedConsensusState = async (request = {}) => {
        const data = query_1.QueryUpgradedConsensusStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.client.v1.Query", "UpgradedConsensusState", data);
        return promise.then(data => query_1.QueryUpgradedConsensusStateResponse.decode(new binary_1.BinaryReader(data)));
    };
}
exports.QueryClientImpl = QueryClientImpl;
const createRpcQueryExtension = (base) => {
    const rpc = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        clientState(request) {
            return queryService.clientState(request);
        },
        clientStates(request) {
            return queryService.clientStates(request);
        },
        consensusState(request) {
            return queryService.consensusState(request);
        },
        consensusStates(request) {
            return queryService.consensusStates(request);
        },
        consensusStateHeights(request) {
            return queryService.consensusStateHeights(request);
        },
        clientStatus(request) {
            return queryService.clientStatus(request);
        },
        clientParams(request) {
            return queryService.clientParams(request);
        },
        upgradedClientState(request) {
            return queryService.upgradedClientState(request);
        },
        upgradedConsensusState(request) {
            return queryService.upgradedConsensusState(request);
        }
    };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
