// quais/lib/commonjs/contract/wrappers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractEventPayload = exports.ContractUnknownEventPayload = exports.ContractTransactionResponse = exports.ContractTransactionReceipt = exports.UndecodedEventLog = exports.EventLog = void 0;
// import from provider.ts instead of index.ts to prevent circular dep
// from quaiscanProvider
const provider_js_1 = require("../providers/provider.js");
const index_js_1 = require("../utils/index.js");
/**
 * An **EventLog** contains additional properties parsed from the {@link Log | **Log**}.
 *
 * @category Contract
 */
class EventLog extends provider_js_1.Log {
    /**
     * The Contract Interface.
     */
    interface;
    /**
     * The matching event.
     */
    fragment;
    /**
     * The parsed arguments passed to the event by `emit`.
     */
    args;
    /**
     * @ignore
     */
    constructor(log, iface, fragment) {
        super(log, log.provider);
        const args = iface.decodeEventLog(fragment, log.data, log.topics);
        (0, index_js_1.defineProperties)(this, { args, fragment, interface: iface });
    }
    /**
     * The name of the event.
     */
    get eventName() {
        return this.fragment.name;
    }
    /**
     * The signature of the event.
     */
    get eventSignature() {
        return this.fragment.format();
    }
}
exports.EventLog = EventLog;
/**
 * An **EventLog** contains additional properties parsed from the {@link Log | **Log**}.
 *
 * @category Contract
 */
class UndecodedEventLog extends provider_js_1.Log {
    /**
     * The error encounted when trying to decode the log.
     */
    error;
    /**
     * @ignore
     */
    constructor(log, error) {
        super(log, log.provider);
        (0, index_js_1.defineProperties)(this, { error });
    }
}
exports.UndecodedEventLog = UndecodedEventLog;
/**
 * A **ContractTransactionReceipt** includes the parsed logs from a {@link TransactionReceipt | **TransactionReceipt**}.
 *
 * @category Contract
 */
class ContractTransactionReceipt extends provider_js_1.TransactionReceipt {
    #iface;
    /**
     * @ignore
     */
    constructor(iface, provider, tx) {
        super(tx, provider);
        this.#iface = iface;
    }
    /**
     * The parsed logs for any {@link Log | **Log**} which has a matching event in the Contract ABI.
     */
    get logs() {
        return super.logs.map((log) => {
            const fragment = log.topics.length ? this.#iface.getEvent(log.topics[0]) : null;
            if (fragment) {
                try {
                    return new EventLog(log, this.#iface, fragment);
                }
                catch (error) {
                    return new UndecodedEventLog(log, error);
                }
            }
            return log;
        });
    }
}
exports.ContractTransactionReceipt = ContractTransactionReceipt;
/**
 * A **ContractTransactionResponse** will return a {@link ContractTransactionReceipt | **ContractTransactionReceipt**}
 * when waited on.
 *
 * @category Contract
 */
class ContractTransactionResponse extends provider_js_1.QuaiTransactionResponse {
    #iface;
    /**
     * @ignore
     */
    constructor(iface, provider, tx) {
        super(tx, provider);
        this.#iface = iface;
    }
    /**
     * Resolves once this transaction has been mined and has `confirms` blocks including it (default: `1`) with an
     * optional `timeout`.
     *
     * This can resolve to `null` only if `confirms` is `0` and the transaction has not been mined, otherwise this will
     * wait until enough confirmations have completed.
     *
     * @param {number} confirms - The number of confirmations to wait for.
     * @returns {Promise<ContractTransactionReceipt | null>} The transaction receipt, or `null` if `confirms` is `0`.
     */
    async wait(confirms) {
        const receipt = await super.wait(confirms);
        if (receipt == null) {
            return null;
        }
        return new ContractTransactionReceipt(this.#iface, this.provider, receipt);
    }
}
exports.ContractTransactionResponse = ContractTransactionResponse;
/**
 * A **ContractUnknownEventPayload** is included as the last parameter to Contract Events when the event does not match
 * any events in the ABI.
 *
 * @category Contract
 */
class ContractUnknownEventPayload extends index_js_1.EventPayload {
    /**
     * The log with no matching events.
     */
    log;
    /**
     * @ignore
     */
    constructor(contract, listener, filter, log) {
        super(contract, listener, filter);
        (0, index_js_1.defineProperties)(this, { log });
    }
    /**
     * Resolves to the block the event occured in.
     *
     * @param {Shard} shard - The shard to get the block from.
     * @returns {Promise<Block>} A promise resolving to the block the event occured in.
     */
    async getBlock(shard) {
        return await this.log.getBlock(shard);
    }
    /**
     * Resolves to the transaction the event occured in.
     *
     * @returns {Promise<TransactionResponse>} A promise resolving to the transaction the event occured in.
     */
    async getTransaction() {
        return await this.log.getTransaction();
    }
    /**
     * Resolves to the transaction receipt the event occured in.
     *
     * @returns {Promise<TransactionReceipt>} A promise resolving to the transaction receipt the event occured in.
     */
    async getTransactionReceipt() {
        return await this.log.getTransactionReceipt();
    }
}
exports.ContractUnknownEventPayload = ContractUnknownEventPayload;
/**
 * A **ContractEventPayload** is included as the last parameter to Contract Events when the event is known.
 *
 * @category Contract
 */
class ContractEventPayload extends ContractUnknownEventPayload {
    /**
     * @ignore
     */
    constructor(contract, listener, filter, fragment, _log) {
        super(contract, listener, filter, new EventLog(_log, contract.interface, fragment));
        const args = contract.interface.decodeEventLog(fragment, this.log.data, this.log.topics);
        (0, index_js_1.defineProperties)(this, { args, fragment });
    }
    /**
     * The event name.
     */
    get eventName() {
        return this.fragment.name;
    }
    /**
     * The event signature.
     */
    get eventSignature() {
        return this.fragment.format();
    }
}
exports.ContractEventPayload = ContractEventPayload;
