// @3land/listings-sdk/src/types/instructions/closeRegisterTraits.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CloseRegisterTraitsAccounts {
  packAccount: PublicKey
  packTraitsRegistry: PublicKey
  payer: PublicKey
  systemProgram: PublicKey
}

export function closeRegisterTraits(
  accounts: CloseRegisterTraitsAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.packAccount, isSigner: false, isWritable: true },
    { pubkey: accounts.packTraitsRegistry, isSigner: false, isWritable: true },
    { pubkey: accounts.payer, isSigner: true, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([54, 231, 126, 219, 139, 144, 251, 143])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
