// interchain/dist/cjs/codegen/cosmos/bank/v1beta1/tx.registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComposer = exports.load = exports.registry = void 0;
const tx_1 = require("./tx");
exports.registry = [["/cosmos.bank.v1beta1.MsgSend", tx_1.MsgSend], ["/cosmos.bank.v1beta1.MsgMultiSend", tx_1.MsgMultiSend], ["/cosmos.bank.v1beta1.MsgBurn", tx_1.MsgBurn], ["/cosmos.bank.v1beta1.MsgUpdateParams", tx_1.MsgUpdateParams], ["/cosmos.bank.v1beta1.MsgSetSendEnabled", tx_1.MsgSetSendEnabled]];
const load = (protoRegistry) => {
    exports.registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
exports.load = load;
exports.MessageComposer = {
    encoded: {
        send(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgSend",
                value: tx_1.MsgSend.encode(value).finish()
            };
        },
        multiSend(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgMultiSend",
                value: tx_1.MsgMultiSend.encode(value).finish()
            };
        },
        burn(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgBurn",
                value: tx_1.MsgBurn.encode(value).finish()
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgUpdateParams",
                value: tx_1.MsgUpdateParams.encode(value).finish()
            };
        },
        setSendEnabled(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgSetSendEnabled",
                value: tx_1.MsgSetSendEnabled.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        send(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgSend",
                value
            };
        },
        multiSend(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgMultiSend",
                value
            };
        },
        burn(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgBurn",
                value
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgUpdateParams",
                value
            };
        },
        setSendEnabled(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgSetSendEnabled",
                value
            };
        }
    },
    fromPartial: {
        send(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgSend",
                value: tx_1.MsgSend.fromPartial(value)
            };
        },
        multiSend(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgMultiSend",
                value: tx_1.MsgMultiSend.fromPartial(value)
            };
        },
        burn(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgBurn",
                value: tx_1.MsgBurn.fromPartial(value)
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgUpdateParams",
                value: tx_1.MsgUpdateParams.fromPartial(value)
            };
        },
        setSendEnabled(value) {
            return {
                typeUrl: "/cosmos.bank.v1beta1.MsgSetSendEnabled",
                value: tx_1.MsgSetSendEnabled.fromPartial(value)
            };
        }
    }
};
