// interchain/dist/codegen/cosmos/mint/v1beta1/query.rpc.Query.js
import { BinaryReader } from "../../../binary";
import { createProtobufRpcClient } from "@cosmjs/stargate";
import { QueryParamsRequest, QueryParamsResponse, QueryInflationRequest, QueryInflationResponse, QueryAnnualProvisionsRequest, QueryAnnualProvisionsResponse } from "./query";
export class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* Params returns the total set of minting parameters. */
    params = async (request = {}) => {
        const data = QueryParamsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.mint.v1beta1.Query", "Params", data);
        return promise.then(data => QueryParamsResponse.decode(new BinaryReader(data)));
    };
    /* Inflation returns the current minting inflation value. */
    inflation = async (request = {}) => {
        const data = QueryInflationRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.mint.v1beta1.Query", "Inflation", data);
        return promise.then(data => QueryInflationResponse.decode(new BinaryReader(data)));
    };
    /* AnnualProvisions current minting annual provisions value. */
    annualProvisions = async (request = {}) => {
        const data = QueryAnnualProvisionsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.mint.v1beta1.Query", "AnnualProvisions", data);
        return promise.then(data => QueryAnnualProvisionsResponse.decode(new BinaryReader(data)));
    };
}
export const createRpcQueryExtension = (base) => {
    const rpc = createProtobufRpcClient(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        params(request) {
            return queryService.params(request);
        },
        inflation(request) {
            return queryService.inflation(request);
        },
        annualProvisions(request) {
            return queryService.annualProvisions(request);
        }
    };
};
