// chain-registry/testnet/named.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectivetestnet = exports.imversedtestnet = exports.impacthubtestnet = exports.hypersigntestnet = exports.humanstestnet = exports.hedgetestnet = exports.gitopiatestnet = exports.galacticatestnet = exports.finschiatestnet = exports.fiammatestnet = exports.fetchhubtestnet = exports.evmostestnet = exports.epixtestnet = exports.empowertestnet = exports.empetestnet = exports.elystestnet = exports.dymensiontestnet1 = exports.dydxtestnet = exports.doravotatestnet = exports.dhealthtestnet = exports.desmostestnet = exports.deardogetestnet = exports.cudostestnet = exports.cosstestnet = exports.cosmwasmtestnet = exports.cosmosicsprovidertestnet = exports.cosmoshubtestnet = exports.coreumtestnet = exports.coolcattestnet = exports.composabletestnet = exports.chimbatestnet = exports.cheqdtestnet = exports.chain4energytestnet = exports.celestiatestnet3 = exports.cascadiatestnet = exports.blockxtestnet = exports.bitcannadevnet = exports.bfhevmtestnet = exports.babylontestnet = exports.axelartestnet = exports.autheotestnet = exports.auratestnet = exports.atomonetestnet = exports.artelatestnet = exports.arkeonetworktestnet = exports.archwaytestnet = exports.alloratestnet = exports.akashtestnet = exports.airchainsvaranasitestnet = exports.airchainstestnet = void 0;
exports.selfchaindevnet = exports.seitestnet2 = exports.seitestnet = exports.sedatestnet = exports.secretnetworktestnet = exports.sagatestnet = exports.qwoyntestnet = exports.quicksilvertestnet = exports.quasartestnet = exports.pryzmtestnet = exports.pockettestnet = exports.planqtestnet = exports.persistencetestnet2 = exports.persistencetestnet = exports.permtestnet = exports.pelltestnet = exports.osmosistestnet = exports.okp4testnet = exports.nyxtestnet = exports.nomictestnet = exports.nolustestnet = exports.noistestnet = exports.nobletestnet = exports.nilliontestnet = exports.nibirutestnet3 = exports.nibirutestnet2 = exports.nibirutestnet = exports.neutrontestnet = exports.neuratestnet = exports.migalootestnet = exports.metanovaversetestnet = exports.marstestnet = exports.mantrachaintestnet2 = exports.mantrachaintestnet = exports.manifesttestnet = exports.lumeratestnet = exports.lumenxtestnet = exports.lombardledgertestnet = exports.likecointestnet = exports.lavatestnet = exports.kyvetestnet = exports.kujiratestnet = exports.kimanetworktestnet = exports.kichaintestnet = exports.junotestnet = exports.jackaltestnet2 = exports.jackaltestnet = exports.iopntestnet = exports.intentotestnet = exports.int3facetestnet = void 0;
exports.zigchaintestnet = exports.zetachaintestnet = exports.zenrocktestnet = exports.xrplevmtestnet = exports.xplatestnet = exports.xiontestnet2 = exports.wavehashtestnet = exports.wardenprotocoltestnet = exports.upticktestnet = exports.uniontestnet = exports.unificationtestnet = exports.ulastestnet = exports.tucanatestnet = exports.titantestnet = exports.titannettestnet = exports.terra2testnet = exports.terpnettestnet2 = exports.terpnettestnet = exports.synternettestnet = exports.symphonytestnet3 = exports.symphonytestnet = exports.swisstroniktestnet = exports.sunrisetestnet = exports.stridetestnet = exports.statesettestnet = exports.stargazetestnet = exports.sourcetestnet = exports.soarchaintestnet = exports.sixtestnet = exports.shentutestnet = exports.sgetestnet4 = exports.sgetestnet = void 0;
exports.airchainstestnet = __importStar(require("./airchainstestnet"));
exports.airchainsvaranasitestnet = __importStar(require("./airchainsvaranasitestnet"));
exports.akashtestnet = __importStar(require("./akashtestnet"));
exports.alloratestnet = __importStar(require("./alloratestnet"));
exports.archwaytestnet = __importStar(require("./archwaytestnet"));
exports.arkeonetworktestnet = __importStar(require("./arkeonetworktestnet"));
exports.artelatestnet = __importStar(require("./artelatestnet"));
exports.atomonetestnet = __importStar(require("./atomonetestnet"));
exports.auratestnet = __importStar(require("./auratestnet"));
exports.autheotestnet = __importStar(require("./autheotestnet"));
exports.axelartestnet = __importStar(require("./axelartestnet"));
exports.babylontestnet = __importStar(require("./babylontestnet"));
exports.bfhevmtestnet = __importStar(require("./bfhevmtestnet"));
exports.bitcannadevnet = __importStar(require("./bitcannadevnet"));
exports.blockxtestnet = __importStar(require("./blockxtestnet"));
exports.cascadiatestnet = __importStar(require("./cascadiatestnet"));
exports.celestiatestnet3 = __importStar(require("./celestiatestnet3"));
exports.chain4energytestnet = __importStar(require("./chain4energytestnet"));
exports.cheqdtestnet = __importStar(require("./cheqdtestnet"));
exports.chimbatestnet = __importStar(require("./chimbatestnet"));
exports.composabletestnet = __importStar(require("./composabletestnet"));
exports.coolcattestnet = __importStar(require("./coolcattestnet"));
exports.coreumtestnet = __importStar(require("./coreumtestnet"));
exports.cosmoshubtestnet = __importStar(require("./cosmoshubtestnet"));
exports.cosmosicsprovidertestnet = __importStar(require("./cosmosicsprovidertestnet"));
exports.cosmwasmtestnet = __importStar(require("./cosmwasmtestnet"));
exports.cosstestnet = __importStar(require("./cosstestnet"));
exports.cudostestnet = __importStar(require("./cudostestnet"));
exports.deardogetestnet = __importStar(require("./deardogetestnet"));
exports.desmostestnet = __importStar(require("./desmostestnet"));
exports.dhealthtestnet = __importStar(require("./dhealthtestnet"));
exports.doravotatestnet = __importStar(require("./doravotatestnet"));
exports.dydxtestnet = __importStar(require("./dydxtestnet"));
exports.dymensiontestnet1 = __importStar(require("./dymensiontestnet1"));
exports.elystestnet = __importStar(require("./elystestnet"));
exports.empetestnet = __importStar(require("./empetestnet"));
exports.empowertestnet = __importStar(require("./empowertestnet"));
exports.epixtestnet = __importStar(require("./epixtestnet"));
exports.evmostestnet = __importStar(require("./evmostestnet"));
exports.fetchhubtestnet = __importStar(require("./fetchhubtestnet"));
exports.fiammatestnet = __importStar(require("./fiammatestnet"));
exports.finschiatestnet = __importStar(require("./finschiatestnet"));
exports.galacticatestnet = __importStar(require("./galacticatestnet"));
exports.gitopiatestnet = __importStar(require("./gitopiatestnet"));
exports.hedgetestnet = __importStar(require("./hedgetestnet"));
exports.humanstestnet = __importStar(require("./humanstestnet"));
exports.hypersigntestnet = __importStar(require("./hypersigntestnet"));
exports.impacthubtestnet = __importStar(require("./impacthubtestnet"));
exports.imversedtestnet = __importStar(require("./imversedtestnet"));
exports.injectivetestnet = __importStar(require("./injectivetestnet"));
exports.int3facetestnet = __importStar(require("./int3facetestnet"));
exports.intentotestnet = __importStar(require("./intentotestnet"));
exports.iopntestnet = __importStar(require("./iopntestnet"));
exports.jackaltestnet = __importStar(require("./jackaltestnet"));
exports.jackaltestnet2 = __importStar(require("./jackaltestnet2"));
exports.junotestnet = __importStar(require("./junotestnet"));
exports.kichaintestnet = __importStar(require("./kichaintestnet"));
exports.kimanetworktestnet = __importStar(require("./kimanetworktestnet"));
exports.kujiratestnet = __importStar(require("./kujiratestnet"));
exports.kyvetestnet = __importStar(require("./kyvetestnet"));
exports.lavatestnet = __importStar(require("./lavatestnet"));
exports.likecointestnet = __importStar(require("./likecointestnet"));
exports.lombardledgertestnet = __importStar(require("./lombardledgertestnet"));
exports.lumenxtestnet = __importStar(require("./lumenxtestnet"));
exports.lumeratestnet = __importStar(require("./lumeratestnet"));
exports.manifesttestnet = __importStar(require("./manifesttestnet"));
exports.mantrachaintestnet = __importStar(require("./mantrachaintestnet"));
exports.mantrachaintestnet2 = __importStar(require("./mantrachaintestnet2"));
exports.marstestnet = __importStar(require("./marstestnet"));
exports.metanovaversetestnet = __importStar(require("./metanovaversetestnet"));
exports.migalootestnet = __importStar(require("./migalootestnet"));
exports.neuratestnet = __importStar(require("./neuratestnet"));
exports.neutrontestnet = __importStar(require("./neutrontestnet"));
exports.nibirutestnet = __importStar(require("./nibirutestnet"));
exports.nibirutestnet2 = __importStar(require("./nibirutestnet2"));
exports.nibirutestnet3 = __importStar(require("./nibirutestnet3"));
exports.nilliontestnet = __importStar(require("./nilliontestnet"));
exports.nobletestnet = __importStar(require("./nobletestnet"));
exports.noistestnet = __importStar(require("./noistestnet"));
exports.nolustestnet = __importStar(require("./nolustestnet"));
exports.nomictestnet = __importStar(require("./nomictestnet"));
exports.nyxtestnet = __importStar(require("./nyxtestnet"));
exports.okp4testnet = __importStar(require("./okp4testnet"));
exports.osmosistestnet = __importStar(require("./osmosistestnet"));
exports.pelltestnet = __importStar(require("./pelltestnet"));
exports.permtestnet = __importStar(require("./permtestnet"));
exports.persistencetestnet = __importStar(require("./persistencetestnet"));
exports.persistencetestnet2 = __importStar(require("./persistencetestnet2"));
exports.planqtestnet = __importStar(require("./planqtestnet"));
exports.pockettestnet = __importStar(require("./pockettestnet"));
exports.pryzmtestnet = __importStar(require("./pryzmtestnet"));
exports.quasartestnet = __importStar(require("./quasartestnet"));
exports.quicksilvertestnet = __importStar(require("./quicksilvertestnet"));
exports.qwoyntestnet = __importStar(require("./qwoyntestnet"));
exports.sagatestnet = __importStar(require("./sagatestnet"));
exports.secretnetworktestnet = __importStar(require("./secretnetworktestnet"));
exports.sedatestnet = __importStar(require("./sedatestnet"));
exports.seitestnet = __importStar(require("./seitestnet"));
exports.seitestnet2 = __importStar(require("./seitestnet2"));
exports.selfchaindevnet = __importStar(require("./selfchaindevnet"));
exports.sgetestnet = __importStar(require("./sgetestnet"));
exports.sgetestnet4 = __importStar(require("./sgetestnet4"));
exports.shentutestnet = __importStar(require("./shentutestnet"));
exports.sixtestnet = __importStar(require("./sixtestnet"));
exports.soarchaintestnet = __importStar(require("./soarchaintestnet"));
exports.sourcetestnet = __importStar(require("./sourcetestnet"));
exports.stargazetestnet = __importStar(require("./stargazetestnet"));
exports.statesettestnet = __importStar(require("./statesettestnet"));
exports.stridetestnet = __importStar(require("./stridetestnet"));
exports.sunrisetestnet = __importStar(require("./sunrisetestnet"));
exports.swisstroniktestnet = __importStar(require("./swisstroniktestnet"));
exports.symphonytestnet = __importStar(require("./symphonytestnet"));
exports.symphonytestnet3 = __importStar(require("./symphonytestnet3"));
exports.synternettestnet = __importStar(require("./synternettestnet"));
exports.terpnettestnet = __importStar(require("./terpnettestnet"));
exports.terpnettestnet2 = __importStar(require("./terpnettestnet2"));
exports.terra2testnet = __importStar(require("./terra2testnet"));
exports.titannettestnet = __importStar(require("./titannettestnet"));
exports.titantestnet = __importStar(require("./titantestnet"));
exports.tucanatestnet = __importStar(require("./tucanatestnet"));
exports.ulastestnet = __importStar(require("./ulastestnet"));
exports.unificationtestnet = __importStar(require("./unificationtestnet"));
exports.uniontestnet = __importStar(require("./uniontestnet"));
exports.upticktestnet = __importStar(require("./upticktestnet"));
exports.wardenprotocoltestnet = __importStar(require("./wardenprotocoltestnet"));
exports.wavehashtestnet = __importStar(require("./wavehashtestnet"));
exports.xiontestnet2 = __importStar(require("./xiontestnet2"));
exports.xplatestnet = __importStar(require("./xplatestnet"));
exports.xrplevmtestnet = __importStar(require("./xrplevmtestnet"));
exports.zenrocktestnet = __importStar(require("./zenrocktestnet"));
exports.zetachaintestnet = __importStar(require("./zetachaintestnet"));
exports.zigchaintestnet = __importStar(require("./zigchaintestnet"));
