// solana-agent-kit/dist/actions/okx-dex/chain_data_action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const okx_dex_1 = require("../../tools/okx-dex");
const okxDexChainDataAction = {
    name: "OKX_DEX_CHAIN_DATA",
    similes: ["get chain data", "okx dex chain data"],
    description: "Get chain data from OKX DEX.",
    examples: [
        [
            {
                input: {},
                output: {
                    status: "success",
                    summary: {
                        chains: [
                            {
                                symbol: "SOL",
                                name: "Solana",
                                address: "So11111111111111111111111111111111111111112",
                            },
                        ],
                    },
                },
                explanation: "Getting chain data from OKX DEX",
            },
        ],
    ],
    schema: zod_1.z.object({}),
    handler: async (agent, input) => {
        const chains = await (0, okx_dex_1.getChainData)(agent);
        return {
            status: "success",
            summary: {
                chains: chains,
            },
        };
    },
};
exports.default = okxDexChainDataAction;
