// solana-agent-kit/dist/actions/okx-dex/liquidity_action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const okx_dex_1 = require("../../tools/okx-dex");
const okxDexLiquidityAction = {
    name: "OKX_DEX_LIQUIDITY",
    similes: ["get liquidity", "okx dex liquidity"],
    description: "Get liquidity sources supported by OKX DEX.",
    examples: [
        [
            {
                input: {},
                output: {
                    status: "success",
                    summary: {
                        liquidity: [
                            {
                                symbol: "SOL",
                                name: "Solana",
                                address: "So11111111111111111111111111111111111111112",
                            },
                        ],
                    },
                },
                explanation: "Getting liquidity sources supported by OKX DEX",
            },
        ],
    ],
    schema: zod_1.z.object({}),
    handler: async (agent, input) => {
        const liquidity = await (0, okx_dex_1.getLiquidity)(agent, "solana");
        return {
            status: "success",
            summary: {
                liquidity: liquidity,
            },
        };
    },
};
exports.default = okxDexLiquidityAction;
