// solana-agent-kit/dist/actions/okx-dex/tokens_action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const okx_dex_1 = require("../../tools/okx-dex");
const okxDexTokensAction = {
    name: "OKX_DEX_TOKENS",
    similes: ["list tokens", "okx dex tokens"],
    description: "List all tokens supported by OKX DEX.",
    examples: [
        [
            {
                input: {},
                output: {
                    status: "success",
                    summary: {
                        tokens: [
                            {
                                symbol: "SOL",
                                name: "Solana",
                                address: "So11111111111111111111111111111111111111112",
                            },
                            {
                                symbol: "USDC",
                                name: "USD Coin",
                                address: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
                            },
                        ],
                    },
                },
                explanation: "Listing all tokens supported by OKX DEX",
            },
        ],
    ],
    schema: zod_1.z.object({}),
    handler: async (agent, input) => {
        const tokens = await (0, okx_dex_1.getTokens)(agent);
        return {
            status: "success",
            summary: {
                tokens: tokens,
            },
        };
    },
};
exports.default = okxDexTokensAction;
