// @okxweb3/coin-sui/dist/types/dynamic_fields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicFieldPage = exports.DynamicFieldInfo = exports.DynamicFieldName = exports.DynamicFieldType = void 0;
const superstruct_1 = require("superstruct");
const common_1 = require("./common");
exports.DynamicFieldType = (0, superstruct_1.union)([
    (0, superstruct_1.literal)('DynamicField'),
    (0, superstruct_1.literal)('DynamicObject'),
]);
exports.DynamicFieldName = (0, superstruct_1.object)({
    type: (0, superstruct_1.string)(),
    value: (0, superstruct_1.any)(),
});
exports.DynamicFieldInfo = (0, superstruct_1.object)({
    name: exports.DynamicFieldName,
    bcsName: (0, superstruct_1.string)(),
    type: exports.DynamicFieldType,
    objectType: (0, superstruct_1.string)(),
    objectId: common_1.ObjectId,
    version: (0, superstruct_1.number)(),
    digest: (0, superstruct_1.string)(),
});
exports.DynamicFieldPage = (0, superstruct_1.object)({
    data: (0, superstruct_1.array)(exports.DynamicFieldInfo),
    nextCursor: (0, superstruct_1.union)([common_1.ObjectId, (0, superstruct_1.literal)(null)]),
    hasNextPage: (0, superstruct_1.boolean)(),
});
