// @okxweb3/coin-sui/dist/types/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEventPackage = exports.getEventSender = exports.SubscriptionEvent = exports.SubscriptionId = exports.PaginatedEvents = exports.SuiEvent = exports.EventId = void 0;
const superstruct_1 = require("superstruct");
const common_1 = require("./common");
exports.EventId = (0, superstruct_1.object)({
    txDigest: common_1.TransactionDigest,
    eventSeq: common_1.SequenceNumber,
});
exports.SuiEvent = (0, superstruct_1.object)({
    id: exports.EventId,
    packageId: common_1.ObjectId,
    transactionModule: (0, superstruct_1.string)(),
    sender: common_1.SuiAddress,
    type: (0, superstruct_1.string)(),
    parsedJson: (0, superstruct_1.optional)((0, superstruct_1.record)((0, superstruct_1.string)(), (0, superstruct_1.any)())),
    bcs: (0, superstruct_1.optional)((0, superstruct_1.string)()),
    timestampMs: (0, superstruct_1.optional)((0, superstruct_1.integer)()),
});
exports.PaginatedEvents = (0, superstruct_1.object)({
    data: (0, superstruct_1.array)(exports.SuiEvent),
    nextCursor: (0, superstruct_1.union)([exports.EventId, (0, superstruct_1.literal)(null)]),
    hasNextPage: (0, superstruct_1.boolean)(),
});
exports.SubscriptionId = (0, superstruct_1.number)();
exports.SubscriptionEvent = (0, superstruct_1.object)({
    subscription: exports.SubscriptionId,
    result: exports.SuiEvent,
});
function getEventSender(event) {
    return event.sender;
}
exports.getEventSender = getEventSender;
function getEventPackage(event) {
    return event.packageId;
}
exports.getEventPackage = getEventPackage;
