// @okxweb3/coin-sui/dist/types/transactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExecutionStatusType = exports.getProgrammableTransaction = exports.getTransactionKindName = exports.getTransactionKind = exports.getConsensusCommitPrologueTransaction = exports.getChangeEpochTransaction = exports.getTransactionGasBudget = exports.getTransactionGasPrice = exports.getTransactionGasObject = exports.getGasData = exports.getTransactionSender = exports.getTransactionSignature = exports.getTransactionDigest = exports.getTransaction = exports.DryRunTransactionBlockResponse = exports.PaginatedTransactionResponse = exports.SuiTransactionBlockResponseOptions = exports.SuiTransactionBlockResponse = exports.BalanceChange = exports.SuiObjectChange = exports.SuiObjectChangeCreated = exports.SuiObjectChangeWrapped = exports.SuiObjectChangeDeleted = exports.SuiObjectChangeMutated = exports.SuiObjectChangeTransferred = exports.SuiObjectChangePublished = exports.SuiTransactionBlock = exports.AuthorityName = exports.DevInspectResults = exports.TransactionEvents = exports.TransactionEffects = exports.TransactionEffectsModifiedAtVersions = exports.OwnedObjectRef = exports.ExecutionStatus = exports.ExecutionStatusType = exports.GasCostSummary = exports.AuthorityQuorumSignInfo = exports.GenericAuthoritySignature = exports.AuthoritySignature = exports.SuiTransactionBlockData = exports.SuiTransactionBlockKind = exports.ProgrammableTransaction = exports.SuiCallArg = exports.SuiTransaction = exports.MoveCallSuiTransaction = exports.SuiArgument = exports.Genesis = exports.SuiConsensusCommitPrologue = exports.SuiChangeEpoch = exports.EpochId = void 0;
exports.getPublishedObjectChanges = exports.getObjectChanges = exports.getNewlyCreatedCoinRefsAfterSplit = exports.getTimestampFromTransactionResponse = exports.getCreatedObjects = exports.getEvents = exports.getTransactionEffects = exports.getTotalGasUsedUpperBound = exports.getTotalGasUsed = exports.getExecutionStatusGasSummary = exports.getExecutionStatusError = exports.getExecutionStatus = void 0;
const superstruct_1 = require("superstruct");
const common_1 = require("./common");
const events_1 = require("./events");
const objects_1 = require("./objects");
exports.EpochId = (0, superstruct_1.string)();
exports.SuiChangeEpoch = (0, superstruct_1.object)({
    epoch: exports.EpochId,
    storage_charge: (0, superstruct_1.number)(),
    computation_charge: (0, superstruct_1.number)(),
    storage_rebate: (0, superstruct_1.number)(),
    epoch_start_timestamp_ms: (0, superstruct_1.optional)((0, superstruct_1.number)()),
});
exports.SuiConsensusCommitPrologue = (0, superstruct_1.object)({
    epoch: (0, superstruct_1.number)(),
    round: (0, superstruct_1.number)(),
    commit_timestamp_ms: (0, superstruct_1.number)(),
});
exports.Genesis = (0, superstruct_1.object)({
    objects: (0, superstruct_1.array)(common_1.ObjectId),
});
exports.SuiArgument = (0, superstruct_1.union)([
    (0, superstruct_1.literal)('GasCoin'),
    (0, superstruct_1.object)({ Input: (0, superstruct_1.number)() }),
    (0, superstruct_1.object)({ Result: (0, superstruct_1.number)() }),
    (0, superstruct_1.object)({ NestedResult: (0, superstruct_1.tuple)([(0, superstruct_1.number)(), (0, superstruct_1.number)()]) }),
]);
exports.MoveCallSuiTransaction = (0, superstruct_1.object)({
    arguments: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.SuiArgument)),
    type_arguments: (0, superstruct_1.optional)((0, superstruct_1.array)((0, superstruct_1.string)())),
    package: common_1.ObjectId,
    module: (0, superstruct_1.string)(),
    function: (0, superstruct_1.string)(),
});
exports.SuiTransaction = (0, superstruct_1.union)([
    (0, superstruct_1.object)({ MoveCall: exports.MoveCallSuiTransaction }),
    (0, superstruct_1.object)({ TransferObjects: (0, superstruct_1.tuple)([(0, superstruct_1.array)(exports.SuiArgument), exports.SuiArgument]) }),
    (0, superstruct_1.object)({ SplitCoins: (0, superstruct_1.tuple)([exports.SuiArgument, (0, superstruct_1.array)(exports.SuiArgument)]) }),
    (0, superstruct_1.object)({ MergeCoins: (0, superstruct_1.tuple)([exports.SuiArgument, (0, superstruct_1.array)(exports.SuiArgument)]) }),
    (0, superstruct_1.object)({ Publish: objects_1.SuiMovePackage }),
    (0, superstruct_1.object)({ MakeMoveVec: (0, superstruct_1.tuple)([(0, superstruct_1.nullable)((0, superstruct_1.string)()), (0, superstruct_1.array)(exports.SuiArgument)]) }),
]);
exports.SuiCallArg = (0, superstruct_1.union)([
    (0, superstruct_1.object)({
        type: (0, superstruct_1.literal)('pure'),
        valueType: (0, superstruct_1.optional)((0, superstruct_1.string)()),
        value: common_1.SuiJsonValue,
    }),
    (0, superstruct_1.object)({
        type: (0, superstruct_1.literal)('object'),
        objectType: (0, superstruct_1.literal)('immOrOwnedObject'),
        objectId: common_1.ObjectId,
        version: common_1.SequenceNumber,
        digest: objects_1.ObjectDigest,
    }),
    (0, superstruct_1.object)({
        type: (0, superstruct_1.literal)('object'),
        objectType: (0, superstruct_1.literal)('sharedObject'),
        objectId: common_1.ObjectId,
        initialSharedVersion: common_1.SequenceNumber,
        mutable: (0, superstruct_1.boolean)(),
    }),
]);
exports.ProgrammableTransaction = (0, superstruct_1.object)({
    transactions: (0, superstruct_1.array)(exports.SuiTransaction),
    inputs: (0, superstruct_1.array)(exports.SuiCallArg),
});
exports.SuiTransactionBlockKind = (0, superstruct_1.union)([
    (0, superstruct_1.assign)(exports.SuiChangeEpoch, (0, superstruct_1.object)({ kind: (0, superstruct_1.literal)('ChangeEpoch') })),
    (0, superstruct_1.assign)(exports.SuiConsensusCommitPrologue, (0, superstruct_1.object)({
        kind: (0, superstruct_1.literal)('ConsensusCommitPrologue'),
    })),
    (0, superstruct_1.assign)(exports.Genesis, (0, superstruct_1.object)({ kind: (0, superstruct_1.literal)('Genesis') })),
    (0, superstruct_1.assign)(exports.ProgrammableTransaction, (0, superstruct_1.object)({ kind: (0, superstruct_1.literal)('ProgrammableTransaction') })),
]);
exports.SuiTransactionBlockData = (0, superstruct_1.object)({
    messageVersion: (0, superstruct_1.literal)('v1'),
    transaction: exports.SuiTransactionBlockKind,
    sender: common_1.SuiAddress,
    gasData: objects_1.SuiGasData,
});
exports.AuthoritySignature = (0, superstruct_1.string)();
exports.GenericAuthoritySignature = (0, superstruct_1.union)([
    exports.AuthoritySignature,
    (0, superstruct_1.array)(exports.AuthoritySignature),
]);
exports.AuthorityQuorumSignInfo = (0, superstruct_1.object)({
    epoch: exports.EpochId,
    signature: exports.GenericAuthoritySignature,
    signers_map: (0, superstruct_1.array)((0, superstruct_1.number)()),
});
exports.GasCostSummary = (0, superstruct_1.object)({
    computationCost: (0, superstruct_1.string)(),
    storageCost: (0, superstruct_1.string)(),
    storageRebate: (0, superstruct_1.string)(),
    nonRefundableStorageFee: (0, superstruct_1.string)(),
});
exports.ExecutionStatusType = (0, superstruct_1.union)([
    (0, superstruct_1.literal)('success'),
    (0, superstruct_1.literal)('failure'),
]);
exports.ExecutionStatus = (0, superstruct_1.object)({
    status: exports.ExecutionStatusType,
    error: (0, superstruct_1.optional)((0, superstruct_1.string)()),
});
exports.OwnedObjectRef = (0, superstruct_1.object)({
    owner: common_1.ObjectOwner,
    reference: objects_1.SuiObjectRef,
});
exports.TransactionEffectsModifiedAtVersions = (0, superstruct_1.object)({
    objectId: common_1.ObjectId,
    sequenceNumber: common_1.SequenceNumber,
});
exports.TransactionEffects = (0, superstruct_1.object)({
    messageVersion: (0, superstruct_1.literal)('v1'),
    status: exports.ExecutionStatus,
    executedEpoch: exports.EpochId,
    modifiedAtVersions: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.TransactionEffectsModifiedAtVersions)),
    gasUsed: exports.GasCostSummary,
    sharedObjects: (0, superstruct_1.optional)((0, superstruct_1.array)(objects_1.SuiObjectRef)),
    transactionDigest: common_1.TransactionDigest,
    created: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.OwnedObjectRef)),
    mutated: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.OwnedObjectRef)),
    unwrapped: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.OwnedObjectRef)),
    deleted: (0, superstruct_1.optional)((0, superstruct_1.array)(objects_1.SuiObjectRef)),
    unwrapped_then_deleted: (0, superstruct_1.optional)((0, superstruct_1.array)(objects_1.SuiObjectRef)),
    wrapped: (0, superstruct_1.optional)((0, superstruct_1.array)(objects_1.SuiObjectRef)),
    gasObject: exports.OwnedObjectRef,
    eventsDigest: (0, superstruct_1.optional)(common_1.TransactionEventDigest),
    dependencies: (0, superstruct_1.optional)((0, superstruct_1.array)(common_1.TransactionDigest)),
});
exports.TransactionEvents = (0, superstruct_1.array)(events_1.SuiEvent);
const ReturnValueType = (0, superstruct_1.tuple)([(0, superstruct_1.array)((0, superstruct_1.number)()), (0, superstruct_1.string)()]);
const MutableReferenceOutputType = (0, superstruct_1.tuple)([
    exports.SuiArgument,
    (0, superstruct_1.array)((0, superstruct_1.number)()),
    (0, superstruct_1.string)(),
]);
const ExecutionResultType = (0, superstruct_1.object)({
    mutableReferenceOutputs: (0, superstruct_1.optional)((0, superstruct_1.array)(MutableReferenceOutputType)),
    returnValues: (0, superstruct_1.optional)((0, superstruct_1.array)(ReturnValueType)),
});
exports.DevInspectResults = (0, superstruct_1.object)({
    effects: exports.TransactionEffects,
    events: exports.TransactionEvents,
    results: (0, superstruct_1.optional)((0, superstruct_1.array)(ExecutionResultType)),
    error: (0, superstruct_1.optional)((0, superstruct_1.string)()),
});
exports.AuthorityName = (0, superstruct_1.string)();
exports.SuiTransactionBlock = (0, superstruct_1.object)({
    data: exports.SuiTransactionBlockData,
    txSignatures: (0, superstruct_1.array)((0, superstruct_1.string)()),
});
exports.SuiObjectChangePublished = (0, superstruct_1.object)({
    type: (0, superstruct_1.literal)('published'),
    packageId: common_1.ObjectId,
    version: common_1.SequenceNumber,
    digest: objects_1.ObjectDigest,
    modules: (0, superstruct_1.array)((0, superstruct_1.string)()),
});
exports.SuiObjectChangeTransferred = (0, superstruct_1.object)({
    type: (0, superstruct_1.literal)('transferred'),
    sender: common_1.SuiAddress,
    recipient: common_1.ObjectOwner,
    objectType: (0, superstruct_1.string)(),
    objectId: common_1.ObjectId,
    version: common_1.SequenceNumber,
    digest: objects_1.ObjectDigest,
});
exports.SuiObjectChangeMutated = (0, superstruct_1.object)({
    type: (0, superstruct_1.literal)('mutated'),
    sender: common_1.SuiAddress,
    owner: common_1.ObjectOwner,
    objectType: (0, superstruct_1.string)(),
    objectId: common_1.ObjectId,
    version: common_1.SequenceNumber,
    previousVersion: common_1.SequenceNumber,
    digest: objects_1.ObjectDigest,
});
exports.SuiObjectChangeDeleted = (0, superstruct_1.object)({
    type: (0, superstruct_1.literal)('deleted'),
    sender: common_1.SuiAddress,
    objectType: (0, superstruct_1.string)(),
    objectId: common_1.ObjectId,
    version: common_1.SequenceNumber,
});
exports.SuiObjectChangeWrapped = (0, superstruct_1.object)({
    type: (0, superstruct_1.literal)('wrapped'),
    sender: common_1.SuiAddress,
    objectType: (0, superstruct_1.string)(),
    objectId: common_1.ObjectId,
    version: common_1.SequenceNumber,
});
exports.SuiObjectChangeCreated = (0, superstruct_1.object)({
    type: (0, superstruct_1.literal)('created'),
    sender: common_1.SuiAddress,
    owner: common_1.ObjectOwner,
    objectType: (0, superstruct_1.string)(),
    objectId: common_1.ObjectId,
    version: common_1.SequenceNumber,
    digest: objects_1.ObjectDigest,
});
exports.SuiObjectChange = (0, superstruct_1.union)([
    exports.SuiObjectChangePublished,
    exports.SuiObjectChangeTransferred,
    exports.SuiObjectChangeMutated,
    exports.SuiObjectChangeDeleted,
    exports.SuiObjectChangeWrapped,
    exports.SuiObjectChangeCreated,
]);
exports.BalanceChange = (0, superstruct_1.object)({
    owner: common_1.ObjectOwner,
    coinType: (0, superstruct_1.string)(),
    amount: (0, superstruct_1.string)(),
});
exports.SuiTransactionBlockResponse = (0, superstruct_1.object)({
    digest: common_1.TransactionDigest,
    transaction: (0, superstruct_1.optional)(exports.SuiTransactionBlock),
    effects: (0, superstruct_1.optional)(exports.TransactionEffects),
    events: (0, superstruct_1.optional)(exports.TransactionEvents),
    timestampMs: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    checkpoint: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    confirmedLocalExecution: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    objectChanges: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.SuiObjectChange)),
    balanceChanges: (0, superstruct_1.optional)((0, superstruct_1.array)(exports.BalanceChange)),
    errors: (0, superstruct_1.optional)((0, superstruct_1.array)((0, superstruct_1.string)())),
});
exports.SuiTransactionBlockResponseOptions = (0, superstruct_1.object)({
    showInput: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    showEffects: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    showEvents: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    showObjectChanges: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    showBalanceChanges: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
});
exports.PaginatedTransactionResponse = (0, superstruct_1.object)({
    data: (0, superstruct_1.array)(exports.SuiTransactionBlockResponse),
    nextCursor: (0, superstruct_1.union)([common_1.TransactionDigest, (0, superstruct_1.literal)(null)]),
    hasNextPage: (0, superstruct_1.boolean)(),
});
exports.DryRunTransactionBlockResponse = (0, superstruct_1.object)({
    effects: exports.TransactionEffects,
    events: exports.TransactionEvents,
    objectChanges: (0, superstruct_1.array)(exports.SuiObjectChange),
    balanceChanges: (0, superstruct_1.array)(exports.BalanceChange),
});
function getTransaction(tx) {
    return tx.transaction;
}
exports.getTransaction = getTransaction;
function getTransactionDigest(tx) {
    return tx.digest;
}
exports.getTransactionDigest = getTransactionDigest;
function getTransactionSignature(tx) {
    return tx.transaction?.txSignatures;
}
exports.getTransactionSignature = getTransactionSignature;
function getTransactionSender(tx) {
    return tx.transaction?.data.sender;
}
exports.getTransactionSender = getTransactionSender;
function getGasData(tx) {
    return tx.transaction?.data.gasData;
}
exports.getGasData = getGasData;
function getTransactionGasObject(tx) {
    return getGasData(tx)?.payment;
}
exports.getTransactionGasObject = getTransactionGasObject;
function getTransactionGasPrice(tx) {
    return getGasData(tx)?.price;
}
exports.getTransactionGasPrice = getTransactionGasPrice;
function getTransactionGasBudget(tx) {
    return getGasData(tx)?.budget;
}
exports.getTransactionGasBudget = getTransactionGasBudget;
function getChangeEpochTransaction(data) {
    return data.kind === 'ChangeEpoch' ? data : undefined;
}
exports.getChangeEpochTransaction = getChangeEpochTransaction;
function getConsensusCommitPrologueTransaction(data) {
    return data.kind === 'ConsensusCommitPrologue' ? data : undefined;
}
exports.getConsensusCommitPrologueTransaction = getConsensusCommitPrologueTransaction;
function getTransactionKind(data) {
    return data.transaction?.data.transaction;
}
exports.getTransactionKind = getTransactionKind;
function getTransactionKindName(data) {
    return data.kind;
}
exports.getTransactionKindName = getTransactionKindName;
function getProgrammableTransaction(data) {
    return data.kind === 'ProgrammableTransaction' ? data : undefined;
}
exports.getProgrammableTransaction = getProgrammableTransaction;
function getExecutionStatusType(data) {
    return getExecutionStatus(data)?.status;
}
exports.getExecutionStatusType = getExecutionStatusType;
function getExecutionStatus(data) {
    return getTransactionEffects(data)?.status;
}
exports.getExecutionStatus = getExecutionStatus;
function getExecutionStatusError(data) {
    return getExecutionStatus(data)?.error;
}
exports.getExecutionStatusError = getExecutionStatusError;
function getExecutionStatusGasSummary(data) {
    if ((0, superstruct_1.is)(data, exports.TransactionEffects)) {
        return data.gasUsed;
    }
    return getTransactionEffects(data)?.gasUsed;
}
exports.getExecutionStatusGasSummary = getExecutionStatusGasSummary;
function getTotalGasUsed(data) {
    const gasSummary = getExecutionStatusGasSummary(data);
    return gasSummary
        ? BigInt(gasSummary.computationCost) +
            BigInt(gasSummary.storageCost) -
            BigInt(gasSummary.storageRebate)
        : undefined;
}
exports.getTotalGasUsed = getTotalGasUsed;
function getTotalGasUsedUpperBound(data) {
    const gasSummary = getExecutionStatusGasSummary(data);
    return gasSummary
        ? BigInt(gasSummary.computationCost) + BigInt(gasSummary.storageCost)
        : undefined;
}
exports.getTotalGasUsedUpperBound = getTotalGasUsedUpperBound;
function getTransactionEffects(data) {
    return data.effects;
}
exports.getTransactionEffects = getTransactionEffects;
function getEvents(data) {
    return data.events;
}
exports.getEvents = getEvents;
function getCreatedObjects(data) {
    return getTransactionEffects(data)?.created;
}
exports.getCreatedObjects = getCreatedObjects;
function getTimestampFromTransactionResponse(data) {
    return data.timestampMs ?? undefined;
}
exports.getTimestampFromTransactionResponse = getTimestampFromTransactionResponse;
function getNewlyCreatedCoinRefsAfterSplit(data) {
    return getTransactionEffects(data)?.created?.map((c) => c.reference);
}
exports.getNewlyCreatedCoinRefsAfterSplit = getNewlyCreatedCoinRefsAfterSplit;
function getObjectChanges(data) {
    return data.objectChanges;
}
exports.getObjectChanges = getObjectChanges;
function getPublishedObjectChanges(data) {
    return (data.objectChanges?.filter((a) => (0, superstruct_1.is)(a, exports.SuiObjectChangePublished)) ?? []);
}
exports.getPublishedObjectChanges = getPublishedObjectChanges;
