// interchain/dist/cjs/codegen/cosmos/accounts/v1/tx.registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComposer = exports.load = exports.registry = void 0;
const tx_1 = require("./tx");
exports.registry = [["/cosmos.accounts.v1.MsgInit", tx_1.MsgInit], ["/cosmos.accounts.v1.MsgExecute", tx_1.MsgExecute]];
const load = (protoRegistry) => {
    exports.registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
exports.load = load;
exports.MessageComposer = {
    encoded: {
        init(value) {
            return {
                typeUrl: "/cosmos.accounts.v1.MsgInit",
                value: tx_1.MsgInit.encode(value).finish()
            };
        },
        execute(value) {
            return {
                typeUrl: "/cosmos.accounts.v1.MsgExecute",
                value: tx_1.MsgExecute.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        init(value) {
            return {
                typeUrl: "/cosmos.accounts.v1.MsgInit",
                value
            };
        },
        execute(value) {
            return {
                typeUrl: "/cosmos.accounts.v1.MsgExecute",
                value
            };
        }
    },
    fromPartial: {
        init(value) {
            return {
                typeUrl: "/cosmos.accounts.v1.MsgInit",
                value: tx_1.MsgInit.fromPartial(value)
            };
        },
        execute(value) {
            return {
                typeUrl: "/cosmos.accounts.v1.MsgExecute",
                value: tx_1.MsgExecute.fromPartial(value)
            };
        }
    }
};
