// chain-registry/esm/mainnet/mantrachain/ibc.js
const info = [
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'axelar',
            client_id: '07-tendermint-237',
            connection_id: 'connection-227'
        },
        chain_2: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-6',
            connection_id: 'connection-6'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-170',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-6',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'cosmoshub',
            client_id: '07-tendermint-1331',
            connection_id: 'connection-1062'
        },
        chain_2: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-3',
            connection_id: 'connection-3'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-1252',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-3',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'injective',
            client_id: '07-tendermint-275',
            connection_id: 'connection-277'
        },
        chain_2: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-2',
            connection_id: 'connection-2'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-363',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-2',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-7',
            connection_id: 'connection-7'
        },
        chain_2: {
            chain_name: 'neutron',
            client_id: '07-tendermint-162',
            connection_id: 'connection-119'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-7',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-6964',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-1',
            connection_id: 'connection-1'
        },
        chain_2: {
            chain_name: 'noble',
            client_id: '07-tendermint-117',
            connection_id: 'connection-111'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-1',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-101',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-0',
            connection_id: 'connection-0'
        },
        chain_2: {
            chain_name: 'osmosis',
            client_id: '07-tendermint-3297',
            connection_id: 'connection-2766'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-0',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-85077',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'mantrachain',
            client_id: '07-tendermint-4',
            connection_id: 'connection-4'
        },
        chain_2: {
            chain_name: 'stargaze',
            client_id: '07-tendermint-372',
            connection_id: 'connection-332'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-4',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-406',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    }
];
export default info;
