// solana-agent-kit/dist/actions/wormhole/cctpTransfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cctpTransferAction = void 0;
const zod_1 = require("zod");
exports.cctpTransferAction = {
    name: "CCTP_TRANSFER",
    description: `Transfer a token from solana  to another supported chain using CCTP
  
  
    `,
    similes: [
        "transfer token",
        "transfer usdc",
        "transfer usdt",
        "transfer usdc to ethereum",
        "transfer usdc to base on testnet",
        "transfer usdc to base sepolia",
        "transfer usdc to base mainnet",
        "transfer usdc to avalanche",
        "transfer usdc to avalanche testnet",
        "transfer usdc to avalanche mainnet",
        "transfer usdc to polygon",
        "transfer usdc to polygon testnet",
        "transfer usdc to polygon mainnet",
        "transfer usdc to sui",
    ],
    examples: [
        [
            {
                input: {
                    destinationChain: "Base Sepolia",
                    transferAmount: "1",
                    network: "Testnet",
                },
                output: {
                    status: "success",
                    message: "USDC transferred successfully",
                },
                explanation: "Transfer USDC from Solana to Base Sepolia",
            },
        ],
    ],
    schema: zod_1.z.object({
        destinationChain: zod_1.z
            .string()
            .describe("The destination chain to transfer the token to"),
        transferAmount: zod_1.z.string().describe("The amount of tokens to transfer"),
        network: zod_1.z
            .string()
            .optional()
            .describe("The network to use for the transfer"),
    }),
    handler: async (agent, input) => {
        const response = await agent.cctpTransfer(input);
        return response;
    },
};
exports.default = exports.cctpTransferAction;
