// solana-agent-kit/dist/actions/wormhole/tokenTransfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenTransferAction = void 0;
const zod_1 = require("zod");
exports.tokenTransferAction = {
    name: "token_transfer",
    description: "Transfer a token from Solana to another chain",
    similes: [
        "transfer token",
        "transfer 0.1 sol to BaseSepolia",
        "transfer 0.1 sol to BaseMainnet",
        "transfer 0.1 sol to Avalanche Mainnet",
    ],
    examples: [
        [
            {
                input: {
                    destinationChain: "BaseSepolia",
                    transferAmount: "0.1",
                    network: "Testnet",
                },
                output: {
                    status: "success",
                    message: "Token transferred successfully",
                },
                explanation: "Transfer 0.1 sol to BaseSepolia",
            },
        ],
    ],
    schema: zod_1.z.object({
        destinationChain: zod_1.z
            .string()
            .describe("The destination chain to transfer the token to"),
        network: zod_1.z
            .string()
            .optional()
            .describe("The network to use for the transfer"),
        transferAmount: zod_1.z.string().describe("The amount of tokens to transfer"),
        tokenAddress: zod_1.z
            .string()
            .optional()
            .describe("The address of the token to transfer, in case of solana, it is empty"),
    }),
    handler: async (agent, input) => {
        const result = await agent.tokenTransfer(input);
        return result;
    },
};
exports.default = exports.tokenTransferAction;
