// chain-registry/esm/testnet/mantrachaintestnet2/ibc.js
const info = [
    {
        $schema: '../../ibc_data.schema.json',
        chain_1: {
            chain_name: 'cosmosicsprovidertestnet',
            client_id: '07-tendermint-159',
            connection_id: 'connection-107'
        },
        chain_2: {
            chain_name: 'mantrachaintestnet2',
            client_id: '07-tendermint-3',
            connection_id: 'connection-3'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-241',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-3',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../../ibc_data.schema.json',
        chain_1: {
            chain_name: 'mantrachaintestnet2',
            client_id: '07-tendermint-2',
            connection_id: 'connection-2'
        },
        chain_2: {
            chain_name: 'nobletestnet',
            client_id: '07-tendermint-341',
            connection_id: 'connection-290'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-2',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-237',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../../ibc_data.schema.json',
        chain_1: {
            chain_name: 'mantrachaintestnet2',
            client_id: '07-tendermint-0',
            connection_id: 'connection-0'
        },
        chain_2: {
            chain_name: 'osmosistestnet',
            client_id: '07-tendermint-4086',
            connection_id: 'connection-3531'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-0',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-9126',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    }
];
export default info;
