// @goat-sdk/plugin-erc20/dist/chunk-J6575BSM.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkDTI6JYV3js = require('./chunk-DTI6JYV3.js');

// src/methods.ts
var _viem = require('viem');
async function balanceOf(walletClient, token, parameters) {
  try {
    const resolvedWalletAddress = await walletClient.resolveAddress(
      parameters.wallet
    );
    const rawBalance = await walletClient.read({
      address: token.contractAddress,
      abi: _chunkDTI6JYV3js.ERC20_ABI,
      functionName: "balanceOf",
      args: [resolvedWalletAddress]
    });
    return _viem.formatUnits.call(void 0, rawBalance.value, token.decimals);
  } catch (error) {
    throw Error(`Failed to fetch balance: ${error}`);
  }
}
async function transfer(walletClient, token, parameters) {
  try {
    const amountInBaseUnits = _viem.parseUnits.call(void 0, parameters.amount, token.decimals);
    const resolvedRecipientAddress = await walletClient.resolveAddress(
      parameters.to
    );
    const hash = await walletClient.sendTransaction({
      to: token.contractAddress,
      abi: _chunkDTI6JYV3js.ERC20_ABI,
      functionName: "transfer",
      args: [resolvedRecipientAddress, amountInBaseUnits]
    });
    return hash.hash;
  } catch (error) {
    throw Error(`Failed to transfer: ${error}`);
  }
}




exports.balanceOf = balanceOf; exports.transfer = transfer;
