// solana-agent-kit/dist/langchain/debridge/create_bridge_order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateDebridgeOrderTool = void 0;
const tools_1 = require("langchain/tools");
class CreateDebridgeOrderTool extends tools_1.Tool {
    constructor(solanaKit) {
        super();
        this.solanaKit = solanaKit;
        this.name = "create_bridge_order";
        this.description = `This tool creates a new bridge order for cross-chain token transfers.

  Inputs (input is a JSON string):
  Required fields:
  srcChainId: string, eg "1" for Ethereum or "7565164" for Solana
  srcChainTokenIn: string, eg "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48" for EVM or base58 for Solana
  srcChainTokenInAmount: string, eg "1000000" for 1 USDC
  dstChainId: string, eg "56" for BSC
  dstChainTokenOut: string, eg "0x8ac76a51cc950d9822d68b83fe1ad97b32cd580d"
  dstChainTokenOutRecipient: string, eg "0x742d35Cc6634C0532925a3b844Bc454e4438f44e"
  senderAddress: string, eg "Abws588GagNKeMViBPE2e1WjQ2jViDyw81ZRq8oMSx75" for Solana sender

  Optional fields:
  dstChainTokenOutAmount: string, eg "auto"
  slippage: number, eg 0.5 for 0.5%
  additionalTakerRewardBps: number
  srcIntermediaryTokenAddress: string
  dstIntermediaryTokenAddress: string
  dstIntermediaryTokenSpenderAddress: string
  intermediaryTokenUSDPrice: number
  srcAllowedCancelBeneficiary: string
  referralCode: number
  affiliateFeePercent: number
  srcChainOrderAuthorityAddress: string
  srcChainRefundAddress: string
  dstChainOrderAuthorityAddress: string
  prependOperatingExpenses: boolean
  deBridgeApp: string`;
    }
    async _call(input) {
        try {
            const orderInput = JSON.parse(input);
            const result = await this.solanaKit.createDebridgeOrder(orderInput);
            return JSON.stringify({
                status: "success",
                message: "Successfully created bridge order",
                data: result,
            });
        }
        catch (error) {
            return JSON.stringify({
                status: "error",
                message: error.message,
                code: error.code || "UNKNOWN_ERROR",
            });
        }
    }
}
exports.CreateDebridgeOrderTool = CreateDebridgeOrderTool;
