// chain-registry/noncosmos/named.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tontestnet = exports.solanatestnet = exports.polygontestnet = exports.polkadottestnet = exports.moonbeamtestnet = exports.litecointestnet = exports.fantomtestnet = exports.ethereumtestnet = exports.dogecointestnet = exports.bitcointestnet = exports.bitcoincashtestnet = exports.binancesmartchaintestnet = exports.avalanchetestnet = exports.cosmoshubicstestnet = exports.namada = exports.hall = exports.berachain = exports.zilliqa = exports.xrpl = exports.tinkernet = exports.sui = exports.stellar = exports.statemine = exports.polygon = exports.polkadot = exports.optimism = exports.neo = exports.movement = exports.moonbeam = exports.mantle = exports.litecoin = exports.kusama = exports.internetcomputer = exports.forex = exports.filecoin = exports.fantom = exports.dogecoin = exports.comex = exports.bitcoincash = exports.bitcoin = exports.binancesmartchain = exports.base = exports.avalanche = exports.avail = exports.arbitrum = exports.aptos = exports._0l = exports.penumbra = exports.composablepolkadot = exports.picasso = void 0;
exports.ojotestnet = void 0;
exports.picasso = __importStar(require("./picasso"));
exports.composablepolkadot = __importStar(require("./composablepolkadot"));
exports.penumbra = __importStar(require("./penumbra"));
exports._0l = __importStar(require("./0l"));
exports.aptos = __importStar(require("./aptos"));
exports.arbitrum = __importStar(require("./arbitrum"));
exports.avail = __importStar(require("./avail"));
exports.avalanche = __importStar(require("./avalanche"));
exports.base = __importStar(require("./base"));
exports.binancesmartchain = __importStar(require("./binancesmartchain"));
exports.bitcoin = __importStar(require("./bitcoin"));
exports.bitcoincash = __importStar(require("./bitcoincash"));
exports.comex = __importStar(require("./comex"));
exports.dogecoin = __importStar(require("./dogecoin"));
exports.fantom = __importStar(require("./fantom"));
exports.filecoin = __importStar(require("./filecoin"));
exports.forex = __importStar(require("./forex"));
exports.internetcomputer = __importStar(require("./internetcomputer"));
exports.kusama = __importStar(require("./kusama"));
exports.litecoin = __importStar(require("./litecoin"));
exports.mantle = __importStar(require("./mantle"));
exports.moonbeam = __importStar(require("./moonbeam"));
exports.movement = __importStar(require("./movement"));
exports.neo = __importStar(require("./neo"));
exports.optimism = __importStar(require("./optimism"));
exports.polkadot = __importStar(require("./polkadot"));
exports.polygon = __importStar(require("./polygon"));
exports.statemine = __importStar(require("./statemine"));
exports.stellar = __importStar(require("./stellar"));
exports.sui = __importStar(require("./sui"));
exports.tinkernet = __importStar(require("./tinkernet"));
exports.xrpl = __importStar(require("./xrpl"));
exports.zilliqa = __importStar(require("./zilliqa"));
exports.berachain = __importStar(require("./berachain"));
exports.hall = __importStar(require("./hall"));
exports.namada = __importStar(require("./namada"));
exports.cosmoshubicstestnet = __importStar(require("./cosmoshubicstestnet"));
exports.avalanchetestnet = __importStar(require("./avalanchetestnet"));
exports.binancesmartchaintestnet = __importStar(require("./binancesmartchaintestnet"));
exports.bitcoincashtestnet = __importStar(require("./bitcoincashtestnet"));
exports.bitcointestnet = __importStar(require("./bitcointestnet"));
exports.dogecointestnet = __importStar(require("./dogecointestnet"));
exports.ethereumtestnet = __importStar(require("./ethereumtestnet"));
exports.fantomtestnet = __importStar(require("./fantomtestnet"));
exports.litecointestnet = __importStar(require("./litecointestnet"));
exports.moonbeamtestnet = __importStar(require("./moonbeamtestnet"));
exports.polkadottestnet = __importStar(require("./polkadottestnet"));
exports.polygontestnet = __importStar(require("./polygontestnet"));
exports.solanatestnet = __importStar(require("./solanatestnet"));
exports.tontestnet = __importStar(require("./tontestnet"));
exports.ojotestnet = __importStar(require("./ojotestnet"));
