// @switchboard-xyz/common/lib/protos.js
import protobuf from "protobufjs/minimal.js";
protobuf.util.toJSONOptions = {
    longs: String,
    enums: String,
    bytes: String,
    json: true,
    // oneofs: true,
};
import { oracle_job } from "./protos/index.js";
export var OracleJob = oracle_job.OracleJob;
export var Task = OracleJob.Task;
export var HttpTask = OracleJob.HttpTask;
export var JsonParseTask = OracleJob.JsonParseTask;
export var MedianTask = OracleJob.MedianTask;
export var MeanTask = OracleJob.MeanTask;
export var WebsocketTask = OracleJob.WebsocketTask;
export var DivideTask = OracleJob.DivideTask;
export var MultiplyTask = OracleJob.MultiplyTask;
export var LpTokenPriceTask = OracleJob.LpTokenPriceTask;
export var LpExchangeRateTask = OracleJob.LpExchangeRateTask;
export var ConditionalTask = OracleJob.ConditionalTask;
export var ValueTask = OracleJob.ValueTask;
export var MaxTask = OracleJob.MaxTask;
export var RegexExtractTask = OracleJob.RegexExtractTask;
export var XStepPriceTask = OracleJob.XStepPriceTask;
export var AddTask = OracleJob.AddTask;
export var SubtractTask = OracleJob.SubtractTask;
export var TwapTask = OracleJob.TwapTask;
export var SerumSwapTask = OracleJob.SerumSwapTask;
export var PowTask = OracleJob.PowTask;
export var LendingRateTask = OracleJob.LendingRateTask;
export var MangoPerpMarketTask = OracleJob.MangoPerpMarketTask;
export var JupiterSwapTask = OracleJob.JupiterSwapTask;
export var PerpMarketTask = OracleJob.PerpMarketTask;
export var OracleTask = OracleJob.OracleTask;
export var AnchorFetchTask = OracleJob.AnchorFetchTask;
export var SplStakePoolTask = OracleJob.SplStakePoolTask;
export var SplTokenParseTask = OracleJob.SplTokenParseTask;
export var UniswapExchangeRateTask = OracleJob.UniswapExchangeRateTask;
export var SushiswapExchangeRateTask = OracleJob.SushiswapExchangeRateTask;
export var PancakeswapExchangeRateTask = OracleJob.PancakeswapExchangeRateTask;
export var CacheTask = OracleJob.CacheTask;
export var SysclockOffsetTask = OracleJob.SysclockOffsetTask;
export var MarinadeStateTask = OracleJob.MarinadeStateTask;
export var SolanaAccountDataFetchTask = OracleJob.SolanaAccountDataFetchTask;
export var BufferLayoutParseTask = OracleJob.BufferLayoutParseTask;
export var CronParseTask = OracleJob.CronParseTask;
export var MinTask = OracleJob.MinTask;
export var HistoryFunctionTask = OracleJob.HistoryFunctionTask;
export var VwapTask = OracleJob.VwapTask;
export var EwmaTask = OracleJob.EwmaTask;
export var ComparisonTask = OracleJob.ComparisonTask;
export var RoundTask = OracleJob.RoundTask;
export var BoundTask = OracleJob.BoundTask;
export var SecretsTask = OracleJob.SecretsTask;
export var UnixTimeTask = OracleJob.UnixTimeTask;
