// @lens-protocol/client/gated/dist/lens-protocol-client-gated.esm.js
import { aE as _createClass, aF as _classCallCheck, aG as _asyncToGenerator, aH as _regeneratorRuntime, aI as _inherits, aJ as _createSuper, L as LensClient$1 } from '../../dist/environments-7e55ba3d.esm.js';
export { aw as Actions, A as Authentication, at as Bookmarks, a as ChangeProfileManagerActionType, b as ClaimProfileWithHandleErrorReasonType, c as CollectOpenActionModuleType, e as CommentRankingFilterType, f as ComparisonOperatorConditionType, g as CreateProfileWithHandleErrorReasonType, C as CredentialsExpiredError, h as CustomFiltersType, D as DecryptFailReasonType, E as Environment, a8 as Explore, i as ExploreProfilesOrderByType, k as ExplorePublicationType, j as ExplorePublicationsOrderByType, a9 as Feed, F as FeedEventItemType, l as FollowModuleType, m as FrameVerifySignatureResult, ac as Frames, ad as Handle, H as HiddenCommentsType, ae as Invites, n as LensProfileManagerRelayErrorReasonType, o as LensTransactionFailureType, q as LensTransactionStatusType, r as LimitType, M as ManagedProfileVisibility, t as MarketplaceMetadataAttributeDisplayType, u as ModuleType, af as Modules, aj as Momoka, v as MomokaValidatorError, w as NftCollectionOwnersOrder, x as NftContractType, ak as Nfts, N as NotAuthenticatedError, av as NotInterested, y as NotificationType, al as Notifications, O as OpenActionCategoryType, z as OpenActionModuleType, P as PopularNftCollectionsOrder, am as Profile, B as ProfileActionHistoryType, G as ProfileInterestTypes, I as ProfileReportingFraudSubreason, J as ProfileReportingReason, K as ProfileReportingSpamSubreason, Q as ProfilesOrderBy, an as Publication, R as PublicationContentWarningType, S as PublicationMetadataLicenseType, T as PublicationMetadataMainFocusType, U as PublicationMetadataTransactionType, V as PublicationReactionType, W as PublicationReportingFraudSubreason, X as PublicationReportingIllegalSubreason, Y as PublicationReportingReason, Z as PublicationReportingSensitiveSubreason, _ as PublicationReportingSpamSubreason, $ as PublicationType, au as Reactions, a0 as ReferenceModuleType, a1 as RefreshPublicationMetadataResultType, a2 as RelayErrorReasonType, a3 as RelayRoleKey, ax as Revenue, ay as Search, a4 as SearchPublicationType, a5 as SupportedFiatType, a6 as TagSortCriteriaType, aA as Transaction, az as TransactionPollingError, a7 as TriStateValue, aD as Wallet, d as development, ap as isCommentPublication, aC as isCreateMomokaPublicationResult, aa as isFollowPaidAction, aq as isMirrorPublication, as as isOpenActionModuleWithReferralFee, ab as isOpenActionPaidAction, ao as isPostPublication, ar as isQuotePublication, aB as isRelaySuccess, ah as isUnknownFollowModuleSettings, ag as isUnknownOpenActionModuleSettings, ai as isUnknownReferenceModuleSettings, p as production, s as staging } from '../../dist/environments-7e55ba3d.esm.js';
import { GatedClient } from '@lens-protocol/gated-content';
export { isEncryptedPublicationMetadata } from '@lens-protocol/gated-content';
import { webCryptoProvider } from '@lens-protocol/gated-content/web';
import { InMemoryStorageProvider } from '@lens-protocol/storage';
import { accessCondition, profileOwnershipCondition } from '@lens-protocol/metadata';
import { success } from '@lens-protocol/shared-kernel';
export { ClientError as GraphQLClientError } from 'graphql-request';
export { decodeData, encodeData, isValidHandle } from '@lens-protocol/blockchain-bindings';
import 'graphql';
import 'jwt-decode';
import '@lens-protocol/gated-content/environments';

/**
 * Gated module.
 *
 * @group LensClient Modules
 */
var Gated = /*#__PURE__*/function () {
  function Gated(client, authentication) {
    _classCallCheck(this, Gated);
    this.client = client;
    this.authentication = authentication;
  }

  /**
   * Encrypts Publication Metadata with the given access condition.
   *
   * ⚠️ Requires authenticated LensClient.
   *
   * @example
   * Typical usage:
   * 1) create metadata,
   * 2) encrypt it,
   * 3) upload it to a public location (e.g. IPFS, Arweave),
   * 4) use the resulting contentURI to create a Lens publication.
   * ```ts
   * // create publication metadata via '@lens-protocol/metadata' helpers
   * const metadata = article({ content: '...' });
   *
   * // encrypt the metadata specifying the access condition
   * const result = await client.gated.encryptPublicationMetadata(
   *   metadata,
   *   eoaOwnershipCondition({
   *     address: signer.address,
   *   }),
   * );
   * const encrypted = result.unwrap();
   *
   * // upload the encrypted metadata your storage of choice
   * const contentURI = await uploadToIPFS(encrypted);
   *
   * // use the contentURI to create a publication
   * ```
   *
   * @example
   * Multiple criteria can be combined using the `orCondition` and `andCondition` helpers.
   * ```ts
   * const result = await client.gated.encryptPublicationMetadata(
   *   metadata,
   *   orCondition([
   *     profileOwnershipCondition({
   *       profileId: profile.id,
   *     }),
   *     eoaOwnershipCondition({
   *       address: signer.address,
   *     }),
   *   ])
   * );
   * ```
   *
   * Supported criteria:
   * - `collectCondition` - the collection of a given publication is required
   * - `eoaOwnershipCondition` - the ownership of a given EOA is required
   * - `erc20OwnershipCondition` - the ownership of a given ERC20 amount is required
   * - `erc721OwnershipCondition` - the ownership of a given ERC721 token is required
   * - `erc1155OwnershipCondition` - the ownership of a given ERC1155 token is required
   * - `profileOwnershipCondition` - the ownership of a given profile is required
   * - `followCondition` - following a given profile is required
   * - `andCondition` - up to 5 conditions can be combined using the AND operator (except `orCondition` and `andCondition`)
   * - `orCondition` - up to 5 conditions can be combined using the OR operator (except `orCondition` and `andCondition`)
   *
   * @see [Metadata helpers](https://lens-protocol.github.io/metadata/) for more information on how to create publication metadata and access conditions.
   */
  _createClass(Gated, [{
    key: "encryptPublicationMetadata",
    value: (function () {
      var _encryptPublicationMetadata = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(metadata, condition) {
        var _this = this;
        return _regeneratorRuntime().wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              return _context2.abrupt("return", this.authentication.requireAuthentication( /*#__PURE__*/function () {
                var _ref = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(profileId) {
                  var root;
                  return _regeneratorRuntime().wrap(function _callee$(_context) {
                    while (1) switch (_context.prev = _context.next) {
                      case 0:
                        root = accessCondition([profileOwnershipCondition({
                          profileId: profileId
                        }), condition]);
                        return _context.abrupt("return", _this.client.encryptPublicationMetadata(metadata, root));
                      case 2:
                      case "end":
                        return _context.stop();
                    }
                  }, _callee);
                }));
                return function (_x3) {
                  return _ref.apply(this, arguments);
                };
              }()));
            case 1:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function encryptPublicationMetadata(_x, _x2) {
        return _encryptPublicationMetadata.apply(this, arguments);
      }
      return encryptPublicationMetadata;
    }()
    /**
     * Decrypts a Publication Metadata fragment returned by the Lens API.
     *
     * ⚠️ Requires authenticated LensClient.
     *
     * The fragment MUST be encrypted. Use the {@link isEncryptedPublicationMetadata} type guard to check if the fragment is encrypted.
     *
     * @example
     * Typical usage:
     * 1) fetch a publication via a `LensClient` method,
     * 2) determine if the publication metadata is encrypted,
     * 3) decrypt the metadata fragment,
     * ```ts
     * const post = await client.publication.fetch({ forId: '...' });
     *
     * if (isEncryptedPublicationMetadata(post.metadata)) {
     *   const result = await client.gated.decryptPublicationMetadataFragment(post.metadata);
     *
     *   if (result.isSuccess()) {
     *     console.log(result.value);
     *   }
     * }
     * ```
     */
    )
  }, {
    key: "decryptPublicationMetadataFragment",
    value: (function () {
      var _decryptPublicationMetadataFragment = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee4(encryptedMetadata) {
        var _this2 = this;
        return _regeneratorRuntime().wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              return _context4.abrupt("return", this.authentication.requireAtLeastWalletAuthentication( /*#__PURE__*/function () {
                var _ref2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee3(profileId) {
                  var result;
                  return _regeneratorRuntime().wrap(function _callee3$(_context3) {
                    while (1) switch (_context3.prev = _context3.next) {
                      case 0:
                        _context3.next = 2;
                        return _this2.client.decryptPublicationMetadataFragment(encryptedMetadata, {
                          profileId: profileId !== null && profileId !== void 0 ? profileId : undefined
                        });
                      case 2:
                        result = _context3.sent;
                        if (!result.isSuccess()) {
                          _context3.next = 5;
                          break;
                        }
                        return _context3.abrupt("return", success(result.value));
                      case 5:
                        return _context3.abrupt("return", result);
                      case 6:
                      case "end":
                        return _context3.stop();
                    }
                  }, _callee3);
                }));
                return function (_x5) {
                  return _ref2.apply(this, arguments);
                };
              }()));
            case 1:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this);
      }));
      function decryptPublicationMetadataFragment(_x4) {
        return _decryptPublicationMetadataFragment.apply(this, arguments);
      }
      return decryptPublicationMetadataFragment;
    }())
  }]);
  return Gated;
}();

/**
 * The configuration for the LensClient
 *
 * @group LensClient
 */

/**
 * Lens Protocol Client with token-gated content support.
 *
 * It provides access to all the core {@link Core.LensClient} modules and the {@link Gated} module.
 *
 * @example
 * NodeJS example:
 * ```ts
 * import { Wallet } from 'ethers';
 * import { production } from '@lens-protocol/client';
 * import { LensClient } from '@lens-protocol/client/gated';
 *
 * const signer = new Wallet(process.env.PRIVATE_KEY);
 *
 * const client = new LensClient({
 *   environment: production,
 *
 *   authentication: {
 *     domain: process.env.DOMAIN,
 *     uri: process.env.URI,
 *   },
 *
 *   signer,
 * });
 * ```
 *
 * @example
 * Browser example with `ethers` v6 and browser wallet (e.g. MetaMask):
 * ```ts
 * import { BrowserProvider } from 'ethers';
 * import { production } from '@lens-protocol/client';
 * import { LensClient } from '@lens-protocol/client/gated';
 *
 * const provider = new BrowserProvider(window.ethereum);
 *
 * const client = new LensClient({
 *   environment: production,
 *
 *   authentication: {
 *     domain: window.location.hostname,
 *     uri: window.location.href,
 *   },
 *
 *   signer: await provider.getSigner(),
 * });
 * ```
 *
 * @group LensClient
 */
var LensClient = /*#__PURE__*/function (_core$LensClient) {
  _inherits(LensClient, _core$LensClient);
  var _super = _createSuper(LensClient);
  function LensClient(config) {
    var _config$storage, _config$encryption;
    var _this;
    _classCallCheck(this, LensClient);
    _this = _super.call(this, config);
    _this._gated = new Gated(new GatedClient({
      environment: config.environment.gated,
      authentication: config.authentication,
      signer: config.signer,
      storageProvider: (_config$storage = config.storage) !== null && _config$storage !== void 0 ? _config$storage : new InMemoryStorageProvider(),
      encryptionProvider: (_config$encryption = config.encryption) !== null && _config$encryption !== void 0 ? _config$encryption : webCryptoProvider()
    }), _this._authentication);
    return _this;
  }

  /**
   * The Gated module
   */
  _createClass(LensClient, [{
    key: "gated",
    get: function get() {
      return this._gated;
    }
  }]);
  return LensClient;
}(LensClient$1);

export { Gated, LensClient };
