// @okxweb3/coin-ethereum/dist/sdk/ethereumjs-util/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BIGINT_2EXP256 = exports.BIGINT_2EXP224 = exports.BIGINT_2EXP160 = exports.BIGINT_2EXP96 = exports.BIGINT_224 = exports.BIGINT_160 = exports.BIGINT_100 = exports.BIGINT_96 = exports.BIGINT_256 = exports.BIGINT_255 = exports.BIGINT_128 = exports.BIGINT_64 = exports.BIGINT_32 = exports.BIGINT_31 = exports.BIGINT_28 = exports.BIGINT_27 = exports.BIGINT_8 = exports.BIGINT_7 = exports.BIGINT_3 = exports.BIGINT_2 = exports.BIGINT_1 = exports.BIGINT_0 = exports.BIGINT_NEG1 = exports.MAX_UINT64_BI = exports.MAX_UINT64 = exports.MAX_INTEGER_BI = exports.MAX_INTEGER = void 0;
const Buffer = require('buffer').Buffer;
const crypto_lib_1 = require("@okxweb3/crypto-lib");
exports.MAX_INTEGER = new crypto_lib_1.BN('ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff', 16);
exports.MAX_INTEGER_BI = BigInt('0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff');
exports.MAX_UINT64 = new crypto_lib_1.BN('ffffffffffffffff', 16);
exports.MAX_UINT64_BI = BigInt('0xffffffffffffffff');
exports.BIGINT_NEG1 = BigInt(-1);
exports.BIGINT_0 = BigInt(0);
exports.BIGINT_1 = BigInt(1);
exports.BIGINT_2 = BigInt(2);
exports.BIGINT_3 = BigInt(3);
exports.BIGINT_7 = BigInt(7);
exports.BIGINT_8 = BigInt(8);
exports.BIGINT_27 = BigInt(27);
exports.BIGINT_28 = BigInt(28);
exports.BIGINT_31 = BigInt(31);
exports.BIGINT_32 = BigInt(32);
exports.BIGINT_64 = BigInt(64);
exports.BIGINT_128 = BigInt(128);
exports.BIGINT_255 = BigInt(255);
exports.BIGINT_256 = BigInt(256);
exports.BIGINT_96 = BigInt(96);
exports.BIGINT_100 = BigInt(100);
exports.BIGINT_160 = BigInt(160);
exports.BIGINT_224 = BigInt(224);
exports.BIGINT_2EXP96 = BigInt(79228162514264337593543950336);
exports.BIGINT_2EXP160 = BigInt(1461501637330902918203684832716283019655932542976);
exports.BIGINT_2EXP224 = BigInt(26959946667150639794667015087019630673637144422540572481103610249216);
exports.BIGINT_2EXP256 = exports.BIGINT_2 ** exports.BIGINT_256;
