// @lifi/sdk/src/_esm/core/BaseStepExecutor.js
import { StatusManager } from './StatusManager.js';
// Please be careful when changing the defaults as it may break the behavior (e.g., background execution)
const defaultInteractionSettings = {
    allowInteraction: true,
    allowUpdates: true,
    allowExecution: true,
};
export class BaseStepExecutor {
    constructor(options) {
        Object.defineProperty(this, "executionOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusManager", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "allowUserInteraction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "allowExecution", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: true
        });
        Object.defineProperty(this, "setInteraction", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (settings) => {
                const interactionSettings = {
                    ...defaultInteractionSettings,
                    ...settings,
                };
                this.allowUserInteraction = interactionSettings.allowInteraction;
                this.statusManager.allowUpdates(interactionSettings.allowUpdates);
                this.allowExecution = interactionSettings.allowExecution;
            }
        });
        this.statusManager = new StatusManager(options.routeId);
        this.executionOptions = options.executionOptions;
    }
}
