// solana-agent-kit/dist/actions/orca/createOrcaCLMM.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const web3_js_1 = require("@solana/web3.js");
const tools_1 = require("../../tools");
const decimal_js_1 = __importDefault(require("decimal.js"));
const createOrcaCLMMAction = {
    name: "CREATE_ORCA_CLMM_ACTION",
    description: "Create a Concentrated Liquidity Market Maker (CLMM) pool on Orca, the most efficient and capital-optimized CLMM on Solana. This function initializes a CLMM pool but does not add liquidity. You can add liquidity later using a centered position or a single-sided position.",
    similes: [
        "create orca clmm",
        "create orca concentrated pool",
        "create orca clmm pool",
        "create orca concentrated liquidity",
    ],
    examples: [
        [
            {
                input: {
                    mintDeploy: "JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN",
                    mintPair: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
                    initialPrice: 1.1,
                    feeTier: 1,
                },
                output: {
                    status: "success",
                    message: "CLMM pool created successfully. Note: No liquidity was added.",
                },
                explanation: "Create a CLMM pool with USDC and JUP",
            },
        ],
    ],
    schema: zod_1.z.object({
        mintDeploy: zod_1.z
            .string()
            .describe("The mint address of the token you want to deploy"),
        mintPair: zod_1.z
            .string()
            .describe("The mint address of the token you want to pair the deployed mint with"),
        initialPrice: zod_1.z
            .number()
            .positive()
            .describe("Initial price of mintDeploy in terms of mintPair"),
        feeTier: zod_1.z
            .number()
            .positive()
            .min(1)
            .describe("The fee tier in bps. Options: 1, 2, 4, 5, 16, 30, 65, 100, 200"),
    }),
    handler: async (agent, input) => {
        try {
            const [mintDeploy, mintPair, initialPrice, feeTier] = [
                new web3_js_1.PublicKey(input.mintDeploy),
                new web3_js_1.PublicKey(input.mintPair),
                new decimal_js_1.default(input.initialPrice),
                input.feeTier,
            ];
            const signature = await (0, tools_1.orcaCreateCLMM)(agent, mintDeploy, mintPair, initialPrice, feeTier);
            return {
                status: "success",
                message: "CLMM pool created successfully. Note: No liquidity was added.",
                signature,
            };
        }
        catch (e) {
            return {
                status: "error",
                // @ts-expect-error - TS doesn't know that `e` has a `message` property
                message: `Failed to create Orca CLMM pool: ${e.message}`,
            };
        }
    },
};
exports.default = createOrcaCLMMAction;
