// solana-agent-kit/dist/langchain/okx-dex/get_chain_data_tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOKXDexGetChainDataTool = createOKXDexGetChainDataTool;
const get_chain_data_1 = require("../../tools/okx-dex/get_chain_data");
const tools_1 = require("@langchain/core/tools");
const zod_1 = require("zod");
/**
 * Creates a LangChain tool to get OKX DEX chain data
 * @param agent SolanaAgentKit instance
 * @returns DynamicStructuredTool for getting chain data
 */
function createOKXDexGetChainDataTool(agent) {
    return new tools_1.DynamicStructuredTool({
        name: "get_okx_dex_chain_data",
        description: "Gets chain data from OKX DEX",
        schema: zod_1.z.object({}),
        func: async () => {
            try {
                const result = await (0, get_chain_data_1.getChainData)(agent);
                return JSON.stringify(result);
            }
            catch (error) {
                return JSON.stringify({
                    status: "error",
                    message: error.message || "Failed to get chain data",
                });
            }
        },
    });
}
