// solana-agent-kit/dist/langchain/okx-dex/get_liquidity_tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOKXDexGetLiquidityTool = createOKXDexGetLiquidityTool;
const get_liquidity_1 = require("../../tools/okx-dex/get_liquidity");
const tools_1 = require("@langchain/core/tools");
const zod_1 = require("zod");
/**
 * Creates a LangChain tool to get OKX DEX liquidity
 * @param agent SolanaAgentKit instance
 * @returns DynamicStructuredTool for getting liquidity
 */
function createOKXDexGetLiquidityTool(agent) {
    return new tools_1.DynamicStructuredTool({
        name: "get_okx_dex_liquidity",
        description: "Gets liquidity information from OKX DEX",
        schema: zod_1.z.object({}),
        func: async () => {
            try {
                const result = await (0, get_liquidity_1.getLiquidity)(agent, "501"); // Solana chain ID
                return JSON.stringify(result);
            }
            catch (error) {
                return JSON.stringify({
                    status: "error",
                    message: error.message || "Failed to get liquidity information",
                });
            }
        },
    });
}
