// solana-agent-kit/dist/langchain/okx-dex/get_tokens_tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOKXDexGetTokensTool = createOKXDexGetTokensTool;
const get_tokens_1 = require("../../tools/okx-dex/get_tokens");
const tools_1 = require("@langchain/core/tools");
const zod_1 = require("zod");
/**
 * Creates a LangChain tool to get OKX DEX tokens
 * @param agent SolanaAgentKit instance
 * @returns DynamicStructuredTool for getting tokens
 */
function createOKXDexGetTokensTool(agent) {
    return new tools_1.DynamicStructuredTool({
        name: "get_okx_dex_tokens",
        description: "Gets a list of all tokens available on OKX DEX",
        schema: zod_1.z.object({}),
        func: async () => {
            try {
                const result = await (0, get_tokens_1.getTokens)(agent);
                return JSON.stringify(result);
            }
            catch (error) {
                return JSON.stringify({
                    status: "error",
                    message: error.message || "Failed to get tokens",
                });
            }
        },
    });
}
