// interchain/dist/cjs/codegen/cosmwasm/wasm/v1/query.rpc.Query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
const binary_1 = require("../../../binary");
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("./query");
class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* ContractInfo gets the contract meta data */
    contractInfo = async (request) => {
        const data = query_1.QueryContractInfoRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "ContractInfo", data);
        return promise.then(data => query_1.QueryContractInfoResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ContractHistory gets the contract code history */
    contractHistory = async (request) => {
        const data = query_1.QueryContractHistoryRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "ContractHistory", data);
        return promise.then(data => query_1.QueryContractHistoryResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ContractsByCode lists all smart contracts for a code id */
    contractsByCode = async (request) => {
        const data = query_1.QueryContractsByCodeRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "ContractsByCode", data);
        return promise.then(data => query_1.QueryContractsByCodeResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* AllContractState gets all raw store data for a single contract */
    allContractState = async (request) => {
        const data = query_1.QueryAllContractStateRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "AllContractState", data);
        return promise.then(data => query_1.QueryAllContractStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* RawContractState gets single key from the raw store data of a contract */
    rawContractState = async (request) => {
        const data = query_1.QueryRawContractStateRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "RawContractState", data);
        return promise.then(data => query_1.QueryRawContractStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* SmartContractState get smart query result from the contract */
    smartContractState = async (request) => {
        const data = query_1.QuerySmartContractStateRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "SmartContractState", data);
        return promise.then(data => query_1.QuerySmartContractStateResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* Code gets the binary code and metadata for a singe wasm code */
    code = async (request) => {
        const data = query_1.QueryCodeRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "Code", data);
        return promise.then(data => query_1.QueryCodeResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* Codes gets the metadata for all stored wasm codes */
    codes = async (request = {
        pagination: undefined
    }) => {
        const data = query_1.QueryCodesRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "Codes", data);
        return promise.then(data => query_1.QueryCodesResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* PinnedCodes gets the pinned code ids */
    pinnedCodes = async (request = {
        pagination: undefined
    }) => {
        const data = query_1.QueryPinnedCodesRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "PinnedCodes", data);
        return promise.then(data => query_1.QueryPinnedCodesResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* Params gets the module params */
    params = async (request = {}) => {
        const data = query_1.QueryParamsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "Params", data);
        return promise.then(data => query_1.QueryParamsResponse.decode(new binary_1.BinaryReader(data)));
    };
    /* ContractsByCreator gets the contracts by creator */
    contractsByCreator = async (request) => {
        const data = query_1.QueryContractsByCreatorRequest.encode(request).finish();
        const promise = this.rpc.request("cosmwasm.wasm.v1.Query", "ContractsByCreator", data);
        return promise.then(data => query_1.QueryContractsByCreatorResponse.decode(new binary_1.BinaryReader(data)));
    };
}
exports.QueryClientImpl = QueryClientImpl;
const createRpcQueryExtension = (base) => {
    const rpc = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        contractInfo(request) {
            return queryService.contractInfo(request);
        },
        contractHistory(request) {
            return queryService.contractHistory(request);
        },
        contractsByCode(request) {
            return queryService.contractsByCode(request);
        },
        allContractState(request) {
            return queryService.allContractState(request);
        },
        rawContractState(request) {
            return queryService.rawContractState(request);
        },
        smartContractState(request) {
            return queryService.smartContractState(request);
        },
        code(request) {
            return queryService.code(request);
        },
        codes(request) {
            return queryService.codes(request);
        },
        pinnedCodes(request) {
            return queryService.pinnedCodes(request);
        },
        params(request) {
            return queryService.params(request);
        },
        contractsByCreator(request) {
            return queryService.contractsByCreator(request);
        }
    };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
