// interchain/dist/codegen/cosmos/group/v1/query.rpc.Query.js
import { BinaryReader } from "../../../binary";
import { createProtobufRpcClient } from "@cosmjs/stargate";
import { QueryGroupInfoRequest, QueryGroupInfoResponse, QueryGroupPolicyInfoRequest, QueryGroupPolicyInfoResponse, QueryGroupMembersRequest, QueryGroupMembersResponse, QueryGroupsByAdminRequest, QueryGroupsByAdminResponse, QueryGroupPoliciesByGroupRequest, QueryGroupPoliciesByGroupResponse, QueryGroupPoliciesByAdminRequest, QueryGroupPoliciesByAdminResponse, QueryProposalRequest, QueryProposalResponse, QueryProposalsByGroupPolicyRequest, QueryProposalsByGroupPolicyResponse, QueryVoteByProposalVoterRequest, QueryVoteByProposalVoterResponse, QueryVotesByProposalRequest, QueryVotesByProposalResponse, QueryVotesByVoterRequest, QueryVotesByVoterResponse, QueryGroupsByMemberRequest, QueryGroupsByMemberResponse, QueryTallyResultRequest, QueryTallyResultResponse, QueryGroupsRequest, QueryGroupsResponse } from "./query";
export class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* GroupInfo queries group info based on group id. */
    groupInfo = async (request) => {
        const data = QueryGroupInfoRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupInfo", data);
        return promise.then(data => QueryGroupInfoResponse.decode(new BinaryReader(data)));
    };
    /* GroupPolicyInfo queries group policy info based on account address of group policy. */
    groupPolicyInfo = async (request) => {
        const data = QueryGroupPolicyInfoRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupPolicyInfo", data);
        return promise.then(data => QueryGroupPolicyInfoResponse.decode(new BinaryReader(data)));
    };
    /* GroupMembers queries members of a group by group id. */
    groupMembers = async (request) => {
        const data = QueryGroupMembersRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupMembers", data);
        return promise.then(data => QueryGroupMembersResponse.decode(new BinaryReader(data)));
    };
    /* GroupsByAdmin queries groups by admin address. */
    groupsByAdmin = async (request) => {
        const data = QueryGroupsByAdminRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupsByAdmin", data);
        return promise.then(data => QueryGroupsByAdminResponse.decode(new BinaryReader(data)));
    };
    /* GroupPoliciesByGroup queries group policies by group id. */
    groupPoliciesByGroup = async (request) => {
        const data = QueryGroupPoliciesByGroupRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupPoliciesByGroup", data);
        return promise.then(data => QueryGroupPoliciesByGroupResponse.decode(new BinaryReader(data)));
    };
    /* GroupPoliciesByAdmin queries group policies by admin address. */
    groupPoliciesByAdmin = async (request) => {
        const data = QueryGroupPoliciesByAdminRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupPoliciesByAdmin", data);
        return promise.then(data => QueryGroupPoliciesByAdminResponse.decode(new BinaryReader(data)));
    };
    /* Proposal queries a proposal based on proposal id. */
    proposal = async (request) => {
        const data = QueryProposalRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "Proposal", data);
        return promise.then(data => QueryProposalResponse.decode(new BinaryReader(data)));
    };
    /* ProposalsByGroupPolicy queries proposals based on account address of group policy. */
    proposalsByGroupPolicy = async (request) => {
        const data = QueryProposalsByGroupPolicyRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "ProposalsByGroupPolicy", data);
        return promise.then(data => QueryProposalsByGroupPolicyResponse.decode(new BinaryReader(data)));
    };
    /* VoteByProposalVoter queries a vote by proposal id and voter. */
    voteByProposalVoter = async (request) => {
        const data = QueryVoteByProposalVoterRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "VoteByProposalVoter", data);
        return promise.then(data => QueryVoteByProposalVoterResponse.decode(new BinaryReader(data)));
    };
    /* VotesByProposal queries a vote by proposal id. */
    votesByProposal = async (request) => {
        const data = QueryVotesByProposalRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "VotesByProposal", data);
        return promise.then(data => QueryVotesByProposalResponse.decode(new BinaryReader(data)));
    };
    /* VotesByVoter queries a vote by voter. */
    votesByVoter = async (request) => {
        const data = QueryVotesByVoterRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "VotesByVoter", data);
        return promise.then(data => QueryVotesByVoterResponse.decode(new BinaryReader(data)));
    };
    /* GroupsByMember queries groups by member address. */
    groupsByMember = async (request) => {
        const data = QueryGroupsByMemberRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "GroupsByMember", data);
        return promise.then(data => QueryGroupsByMemberResponse.decode(new BinaryReader(data)));
    };
    /* TallyResult returns the tally result of a proposal. If the proposal is
     still in voting period, then this query computes the current tally state,
     which might not be final. On the other hand, if the proposal is final,
     then it simply returns the `final_tally_result` state stored in the
     proposal itself. */
    tallyResult = async (request) => {
        const data = QueryTallyResultRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "TallyResult", data);
        return promise.then(data => QueryTallyResultResponse.decode(new BinaryReader(data)));
    };
    /* Groups queries all groups in state.
    
     Since: cosmos-sdk 0.47.1 */
    groups = async (request = {
        pagination: undefined
    }) => {
        const data = QueryGroupsRequest.encode(request).finish();
        const promise = this.rpc.request("cosmos.group.v1.Query", "Groups", data);
        return promise.then(data => QueryGroupsResponse.decode(new BinaryReader(data)));
    };
}
export const createRpcQueryExtension = (base) => {
    const rpc = createProtobufRpcClient(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        groupInfo(request) {
            return queryService.groupInfo(request);
        },
        groupPolicyInfo(request) {
            return queryService.groupPolicyInfo(request);
        },
        groupMembers(request) {
            return queryService.groupMembers(request);
        },
        groupsByAdmin(request) {
            return queryService.groupsByAdmin(request);
        },
        groupPoliciesByGroup(request) {
            return queryService.groupPoliciesByGroup(request);
        },
        groupPoliciesByAdmin(request) {
            return queryService.groupPoliciesByAdmin(request);
        },
        proposal(request) {
            return queryService.proposal(request);
        },
        proposalsByGroupPolicy(request) {
            return queryService.proposalsByGroupPolicy(request);
        },
        voteByProposalVoter(request) {
            return queryService.voteByProposalVoter(request);
        },
        votesByProposal(request) {
            return queryService.votesByProposal(request);
        },
        votesByVoter(request) {
            return queryService.votesByVoter(request);
        },
        groupsByMember(request) {
            return queryService.groupsByMember(request);
        },
        tallyResult(request) {
            return queryService.tallyResult(request);
        },
        groups(request) {
            return queryService.groups(request);
        }
    };
};
