// @cks-systems/manifest-sdk/dist/esm/src/wrapper/instructions/BatchUpdateQuoteGlobal.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { wrapperBatchUpdateParamsBeet, } from '../types/WrapperBatchUpdateParams';
/**
 * @category Instructions
 * @category BatchUpdateQuoteGlobal
 * @category generated
 */
export const BatchUpdateQuoteGlobalStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['params', wrapperBatchUpdateParamsBeet],
], 'BatchUpdateQuoteGlobalInstructionArgs');
export const batchUpdateQuoteGlobalInstructionDiscriminator = 6;
/**
 * Creates a _BatchUpdateQuoteGlobal_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category BatchUpdateQuoteGlobal
 * @category generated
 */
export function createBatchUpdateQuoteGlobalInstruction(accounts, args, programId = new web3.PublicKey('wMNFSTkir3HgyZTsB7uqu3i7FA73grFCptPXgrZjksL')) {
    const [data] = BatchUpdateQuoteGlobalStruct.serialize({
        instructionDiscriminator: batchUpdateQuoteGlobalInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.wrapperState,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.manifestProgram,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.owner,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.market,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteMint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteGlobal,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteGlobalVault,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteMarketVault,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.quoteTokenProgram,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
