// @cks-systems/manifest-sdk/dist/esm/src/wrapper/instructions/BatchUpdateBaseGlobal.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { wrapperBatchUpdateParamsBeet, } from '../types/WrapperBatchUpdateParams';
/**
 * @category Instructions
 * @category BatchUpdateBaseGlobal
 * @category generated
 */
export const BatchUpdateBaseGlobalStruct = new beet.FixableBeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['params', wrapperBatchUpdateParamsBeet],
], 'BatchUpdateBaseGlobalInstructionArgs');
export const batchUpdateBaseGlobalInstructionDiscriminator = 5;
/**
 * Creates a _BatchUpdateBaseGlobal_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category BatchUpdateBaseGlobal
 * @category generated
 */
export function createBatchUpdateBaseGlobalInstruction(accounts, args, programId = new web3.PublicKey('wMNFSTkir3HgyZTsB7uqu3i7FA73grFCptPXgrZjksL')) {
    const [data] = BatchUpdateBaseGlobalStruct.serialize({
        instructionDiscriminator: batchUpdateBaseGlobalInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.wrapperState,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.manifestProgram,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.owner,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.market,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.baseMint,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.baseGlobal,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.baseGlobalVault,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.baseMarketVault,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.baseTokenProgram,
            isWritable: false,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
