// interchain/dist/codegen/ibc/core/connection/v1/query.rpc.Query.js
import { BinaryReader } from "../../../../binary";
import { createProtobufRpcClient } from "@cosmjs/stargate";
import { QueryConnectionRequest, QueryConnectionResponse, QueryConnectionsRequest, QueryConnectionsResponse, QueryClientConnectionsRequest, QueryClientConnectionsResponse, QueryConnectionClientStateRequest, QueryConnectionClientStateResponse, QueryConnectionConsensusStateRequest, QueryConnectionConsensusStateResponse, QueryConnectionParamsRequest, QueryConnectionParamsResponse } from "./query";
export class QueryClientImpl {
    rpc;
    constructor(rpc) {
        this.rpc = rpc;
    }
    /* Connection queries an IBC connection end. */
    connection = async (request) => {
        const data = QueryConnectionRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "Connection", data);
        return promise.then(data => QueryConnectionResponse.decode(new BinaryReader(data)));
    };
    /* Connections queries all the IBC connections of a chain. */
    connections = async (request = {
        pagination: undefined
    }) => {
        const data = QueryConnectionsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "Connections", data);
        return promise.then(data => QueryConnectionsResponse.decode(new BinaryReader(data)));
    };
    /* ClientConnections queries the connection paths associated with a client
     state. */
    clientConnections = async (request) => {
        const data = QueryClientConnectionsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ClientConnections", data);
        return promise.then(data => QueryClientConnectionsResponse.decode(new BinaryReader(data)));
    };
    /* ConnectionClientState queries the client state associated with the
     connection. */
    connectionClientState = async (request) => {
        const data = QueryConnectionClientStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ConnectionClientState", data);
        return promise.then(data => QueryConnectionClientStateResponse.decode(new BinaryReader(data)));
    };
    /* ConnectionConsensusState queries the consensus state associated with the
     connection. */
    connectionConsensusState = async (request) => {
        const data = QueryConnectionConsensusStateRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ConnectionConsensusState", data);
        return promise.then(data => QueryConnectionConsensusStateResponse.decode(new BinaryReader(data)));
    };
    /* ConnectionParams queries all parameters of the ibc connection submodule. */
    connectionParams = async (request = {}) => {
        const data = QueryConnectionParamsRequest.encode(request).finish();
        const promise = this.rpc.request("ibc.core.connection.v1.Query", "ConnectionParams", data);
        return promise.then(data => QueryConnectionParamsResponse.decode(new BinaryReader(data)));
    };
}
export const createRpcQueryExtension = (base) => {
    const rpc = createProtobufRpcClient(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        connection(request) {
            return queryService.connection(request);
        },
        connections(request) {
            return queryService.connections(request);
        },
        clientConnections(request) {
            return queryService.clientConnections(request);
        },
        connectionClientState(request) {
            return queryService.connectionClientState(request);
        },
        connectionConsensusState(request) {
            return queryService.connectionConsensusState(request);
        },
        connectionParams(request) {
            return queryService.connectionParams(request);
        }
    };
};
