// chain-registry/mainnet/sidechain/ibc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const info = [
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'axelar',
            client_id: '07-tendermint-242',
            connection_id: 'connection-232'
        },
        chain_2: {
            chain_name: 'sidechain',
            client_id: '07-tendermint-20',
            connection_id: 'connection-7'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-174',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-6',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'cosmoshub',
            client_id: '07-tendermint-1383',
            connection_id: 'connection-1106'
        },
        chain_2: {
            chain_name: 'sidechain',
            client_id: '07-tendermint-25',
            connection_id: 'connection-11'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-1352',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-10',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'injective',
            client_id: '07-tendermint-298',
            connection_id: 'connection-301'
        },
        chain_2: {
            chain_name: 'sidechain',
            client_id: '07-tendermint-24',
            connection_id: 'connection-10'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-431',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-9',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'neutron',
            client_id: '07-tendermint-160',
            connection_id: 'connection-117'
        },
        chain_2: {
            chain_name: 'sidechain',
            client_id: '07-tendermint-23',
            connection_id: 'connection-9'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-6852',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-8',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'noble',
            client_id: '07-tendermint-141',
            connection_id: 'connection-136'
        },
        chain_2: {
            chain_name: 'sidechain',
            client_id: '07-tendermint-7',
            connection_id: 'connection-2'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-123',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-1',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    },
    {
        $schema: '../ibc_data.schema.json',
        chain_1: {
            chain_name: 'osmosis',
            client_id: '07-tendermint-3425',
            connection_id: 'connection-10585'
        },
        chain_2: {
            chain_name: 'sidechain',
            client_id: '07-tendermint-5',
            connection_id: 'connection-1'
        },
        channels: [{
                chain_1: {
                    channel_id: 'channel-98081',
                    port_id: 'transfer'
                },
                chain_2: {
                    channel_id: 'channel-0',
                    port_id: 'transfer'
                },
                ordering: 'unordered',
                version: 'ics20-1',
                tags: {
                    status: 'live',
                    preferred: true
                }
            }]
    }
];
exports.default = info;
