// solana-agent-kit/dist/actions/solutiofi/spreadToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const zod_1 = require("zod");
const spreadTokenAction = {
    name: "SOLUTIOFI_SPREAD_TOKEN",
    description: "Split a token into multiple tokens using SolutioFi",
    similes: [
        "spread token",
        "split token",
        "swap token",
        "spread tokens with solutiofi",
        "swap one token for multiple",
        "buy multiple tokens at once",
        "break token into parts",
        "split token into multiple assets",
        "convert one token into many",
        "swap token into different tokens",
        "allocate token to multiple assets",
        "spread token balance across multiple tokens",
        "swap and distribute token",
        "convert token into smaller portions",
        "exchange one token for multiple",
        "redistribute token into other tokens",
        "convert and spread token balance",
        "swap one asset for multiple tokens",
    ],
    examples: [
        [
            {
                input: {
                    inputAsset: { mint: "inputMint", amount: "1000" },
                    targetTokens: [
                        { mint: "targetMint1", percentage: 60 },
                        { mint: "targetMint2", percentage: 40 },
                    ],
                    priorityFee: "fast",
                },
                output: {
                    status: "success",
                    message: "Successfully spread token",
                    result: { transaction: "tx_signature" },
                },
                explanation: "Split one token into multiple target tokens",
            },
        ],
    ],
    schema: zod_1.z.object({
        inputAsset: zod_1.z
            .object({
            mint: zod_1.z.string(),
            amount: zod_1.z.string(),
        })
            .describe("Input asset to spread"),
        targetTokens: zod_1.z
            .array(zod_1.z.object({
            mint: zod_1.z.string(),
            percentage: zod_1.z.number(),
        }))
            .describe("Target tokens with allocation percentages"),
        priorityFee: zod_1.z
            .enum(["fast", "turbo", "ultra"])
            .describe("Transaction priority level"),
    }),
    handler: async (agent, input) => {
        try {
            const { inputAsset, targetTokens, priorityFee } = input;
            const result = await agent.spreadToken(inputAsset, targetTokens, priorityFee);
            return {
                status: "success",
                result,
                message: "Successfully spread token",
            };
        }
        catch (error) {
            return {
                status: "error",
                message: `Failed to spread token: ${error.message}`,
                code: error.code || "SPREAD_TOKEN_ERROR",
            };
        }
    },
};
exports.default = spreadTokenAction;
