// @akashnetwork/akash-api/dist/generated/akash/market/v1beta3/query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryClientImpl = exports.QueryServiceName = exports.QueryLeaseResponse = exports.QueryLeaseRequest = exports.QueryLeasesResponse = exports.QueryLeasesRequest = exports.QueryBidResponse = exports.QueryBidRequest = exports.QueryBidsResponse = exports.QueryBidsRequest = exports.QueryOrderResponse = exports.QueryOrderRequest = exports.QueryOrdersResponse = exports.QueryOrdersRequest = void 0;
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const pagination_1 = require("../../../cosmos/base/query/v1beta1/pagination");
const typeRegistry_1 = require("../../../typeRegistry");
const types_1 = require("../../escrow/v1beta3/types");
const bid_1 = require("./bid");
const lease_1 = require("./lease");
const order_1 = require("./order");
function createBaseQueryOrdersRequest() {
    return {
        $type: 'akash.market.v1beta3.QueryOrdersRequest',
        filters: undefined,
        pagination: undefined,
    };
}
exports.QueryOrdersRequest = {
    $type: 'akash.market.v1beta3.QueryOrdersRequest',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.filters !== undefined) {
            order_1.OrderFilters.encode(message.filters, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryOrdersRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.filters = order_1.OrderFilters.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryOrdersRequest.$type,
            filters: isSet(object.filters)
                ? order_1.OrderFilters.fromJSON(object.filters)
                : undefined,
            pagination: isSet(object.pagination)
                ? pagination_1.PageRequest.fromJSON(object.pagination)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.filters !== undefined) {
            obj.filters = order_1.OrderFilters.toJSON(message.filters);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.QueryOrdersRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryOrdersRequest();
        message.filters =
            object.filters !== undefined && object.filters !== null
                ? order_1.OrderFilters.fromPartial(object.filters)
                : undefined;
        message.pagination =
            object.pagination !== undefined && object.pagination !== null
                ? pagination_1.PageRequest.fromPartial(object.pagination)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryOrdersRequest.$type, exports.QueryOrdersRequest);
function createBaseQueryOrdersResponse() {
    return {
        $type: 'akash.market.v1beta3.QueryOrdersResponse',
        orders: [],
        pagination: undefined,
    };
}
exports.QueryOrdersResponse = {
    $type: 'akash.market.v1beta3.QueryOrdersResponse',
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.orders) {
            order_1.Order.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryOrdersResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.orders.push(order_1.Order.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryOrdersResponse.$type,
            orders: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.orders)
                ? object.orders.map((e) => order_1.Order.fromJSON(e))
                : [],
            pagination: isSet(object.pagination)
                ? pagination_1.PageResponse.fromJSON(object.pagination)
                : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.orders) === null || _a === void 0 ? void 0 : _a.length) {
            obj.orders = message.orders.map((e) => order_1.Order.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.QueryOrdersResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryOrdersResponse();
        message.orders = ((_a = object.orders) === null || _a === void 0 ? void 0 : _a.map((e) => order_1.Order.fromPartial(e))) || [];
        message.pagination =
            object.pagination !== undefined && object.pagination !== null
                ? pagination_1.PageResponse.fromPartial(object.pagination)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryOrdersResponse.$type, exports.QueryOrdersResponse);
function createBaseQueryOrderRequest() {
    return { $type: 'akash.market.v1beta3.QueryOrderRequest', id: undefined };
}
exports.QueryOrderRequest = {
    $type: 'akash.market.v1beta3.QueryOrderRequest',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            order_1.OrderID.encode(message.id, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryOrderRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = order_1.OrderID.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryOrderRequest.$type,
            id: isSet(object.id) ? order_1.OrderID.fromJSON(object.id) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = order_1.OrderID.toJSON(message.id);
        }
        return obj;
    },
    create(base) {
        return exports.QueryOrderRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryOrderRequest();
        message.id =
            object.id !== undefined && object.id !== null
                ? order_1.OrderID.fromPartial(object.id)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryOrderRequest.$type, exports.QueryOrderRequest);
function createBaseQueryOrderResponse() {
    return { $type: 'akash.market.v1beta3.QueryOrderResponse', order: undefined };
}
exports.QueryOrderResponse = {
    $type: 'akash.market.v1beta3.QueryOrderResponse',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.order !== undefined) {
            order_1.Order.encode(message.order, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryOrderResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.order = order_1.Order.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryOrderResponse.$type,
            order: isSet(object.order) ? order_1.Order.fromJSON(object.order) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.order !== undefined) {
            obj.order = order_1.Order.toJSON(message.order);
        }
        return obj;
    },
    create(base) {
        return exports.QueryOrderResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryOrderResponse();
        message.order =
            object.order !== undefined && object.order !== null
                ? order_1.Order.fromPartial(object.order)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryOrderResponse.$type, exports.QueryOrderResponse);
function createBaseQueryBidsRequest() {
    return {
        $type: 'akash.market.v1beta3.QueryBidsRequest',
        filters: undefined,
        pagination: undefined,
    };
}
exports.QueryBidsRequest = {
    $type: 'akash.market.v1beta3.QueryBidsRequest',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.filters !== undefined) {
            bid_1.BidFilters.encode(message.filters, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBidsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.filters = bid_1.BidFilters.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryBidsRequest.$type,
            filters: isSet(object.filters)
                ? bid_1.BidFilters.fromJSON(object.filters)
                : undefined,
            pagination: isSet(object.pagination)
                ? pagination_1.PageRequest.fromJSON(object.pagination)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.filters !== undefined) {
            obj.filters = bid_1.BidFilters.toJSON(message.filters);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.QueryBidsRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryBidsRequest();
        message.filters =
            object.filters !== undefined && object.filters !== null
                ? bid_1.BidFilters.fromPartial(object.filters)
                : undefined;
        message.pagination =
            object.pagination !== undefined && object.pagination !== null
                ? pagination_1.PageRequest.fromPartial(object.pagination)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryBidsRequest.$type, exports.QueryBidsRequest);
function createBaseQueryBidsResponse() {
    return {
        $type: 'akash.market.v1beta3.QueryBidsResponse',
        bids: [],
        pagination: undefined,
    };
}
exports.QueryBidsResponse = {
    $type: 'akash.market.v1beta3.QueryBidsResponse',
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.bids) {
            exports.QueryBidResponse.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBidsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.bids.push(exports.QueryBidResponse.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryBidsResponse.$type,
            bids: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.bids)
                ? object.bids.map((e) => exports.QueryBidResponse.fromJSON(e))
                : [],
            pagination: isSet(object.pagination)
                ? pagination_1.PageResponse.fromJSON(object.pagination)
                : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.bids) === null || _a === void 0 ? void 0 : _a.length) {
            obj.bids = message.bids.map((e) => exports.QueryBidResponse.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.QueryBidsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryBidsResponse();
        message.bids =
            ((_a = object.bids) === null || _a === void 0 ? void 0 : _a.map((e) => exports.QueryBidResponse.fromPartial(e))) || [];
        message.pagination =
            object.pagination !== undefined && object.pagination !== null
                ? pagination_1.PageResponse.fromPartial(object.pagination)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryBidsResponse.$type, exports.QueryBidsResponse);
function createBaseQueryBidRequest() {
    return { $type: 'akash.market.v1beta3.QueryBidRequest', id: undefined };
}
exports.QueryBidRequest = {
    $type: 'akash.market.v1beta3.QueryBidRequest',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            bid_1.BidID.encode(message.id, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBidRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = bid_1.BidID.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryBidRequest.$type,
            id: isSet(object.id) ? bid_1.BidID.fromJSON(object.id) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = bid_1.BidID.toJSON(message.id);
        }
        return obj;
    },
    create(base) {
        return exports.QueryBidRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryBidRequest();
        message.id =
            object.id !== undefined && object.id !== null
                ? bid_1.BidID.fromPartial(object.id)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryBidRequest.$type, exports.QueryBidRequest);
function createBaseQueryBidResponse() {
    return {
        $type: 'akash.market.v1beta3.QueryBidResponse',
        bid: undefined,
        escrowAccount: undefined,
    };
}
exports.QueryBidResponse = {
    $type: 'akash.market.v1beta3.QueryBidResponse',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.bid !== undefined) {
            bid_1.Bid.encode(message.bid, writer.uint32(10).fork()).ldelim();
        }
        if (message.escrowAccount !== undefined) {
            types_1.Account.encode(message.escrowAccount, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryBidResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.bid = bid_1.Bid.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.escrowAccount = types_1.Account.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryBidResponse.$type,
            bid: isSet(object.bid) ? bid_1.Bid.fromJSON(object.bid) : undefined,
            escrowAccount: isSet(object.escrowAccount)
                ? types_1.Account.fromJSON(object.escrowAccount)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bid !== undefined) {
            obj.bid = bid_1.Bid.toJSON(message.bid);
        }
        if (message.escrowAccount !== undefined) {
            obj.escrowAccount = types_1.Account.toJSON(message.escrowAccount);
        }
        return obj;
    },
    create(base) {
        return exports.QueryBidResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryBidResponse();
        message.bid =
            object.bid !== undefined && object.bid !== null
                ? bid_1.Bid.fromPartial(object.bid)
                : undefined;
        message.escrowAccount =
            object.escrowAccount !== undefined && object.escrowAccount !== null
                ? types_1.Account.fromPartial(object.escrowAccount)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryBidResponse.$type, exports.QueryBidResponse);
function createBaseQueryLeasesRequest() {
    return {
        $type: 'akash.market.v1beta3.QueryLeasesRequest',
        filters: undefined,
        pagination: undefined,
    };
}
exports.QueryLeasesRequest = {
    $type: 'akash.market.v1beta3.QueryLeasesRequest',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.filters !== undefined) {
            lease_1.LeaseFilters.encode(message.filters, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryLeasesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.filters = lease_1.LeaseFilters.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryLeasesRequest.$type,
            filters: isSet(object.filters)
                ? lease_1.LeaseFilters.fromJSON(object.filters)
                : undefined,
            pagination: isSet(object.pagination)
                ? pagination_1.PageRequest.fromJSON(object.pagination)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.filters !== undefined) {
            obj.filters = lease_1.LeaseFilters.toJSON(message.filters);
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.QueryLeasesRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryLeasesRequest();
        message.filters =
            object.filters !== undefined && object.filters !== null
                ? lease_1.LeaseFilters.fromPartial(object.filters)
                : undefined;
        message.pagination =
            object.pagination !== undefined && object.pagination !== null
                ? pagination_1.PageRequest.fromPartial(object.pagination)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryLeasesRequest.$type, exports.QueryLeasesRequest);
function createBaseQueryLeasesResponse() {
    return {
        $type: 'akash.market.v1beta3.QueryLeasesResponse',
        leases: [],
        pagination: undefined,
    };
}
exports.QueryLeasesResponse = {
    $type: 'akash.market.v1beta3.QueryLeasesResponse',
    encode(message, writer = minimal_1.default.Writer.create()) {
        for (const v of message.leases) {
            exports.QueryLeaseResponse.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryLeasesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.leases.push(exports.QueryLeaseResponse.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryLeasesResponse.$type,
            leases: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.leases)
                ? object.leases.map((e) => exports.QueryLeaseResponse.fromJSON(e))
                : [],
            pagination: isSet(object.pagination)
                ? pagination_1.PageResponse.fromJSON(object.pagination)
                : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.leases) === null || _a === void 0 ? void 0 : _a.length) {
            obj.leases = message.leases.map((e) => exports.QueryLeaseResponse.toJSON(e));
        }
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        return obj;
    },
    create(base) {
        return exports.QueryLeasesResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryLeasesResponse();
        message.leases =
            ((_a = object.leases) === null || _a === void 0 ? void 0 : _a.map((e) => exports.QueryLeaseResponse.fromPartial(e))) || [];
        message.pagination =
            object.pagination !== undefined && object.pagination !== null
                ? pagination_1.PageResponse.fromPartial(object.pagination)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryLeasesResponse.$type, exports.QueryLeasesResponse);
function createBaseQueryLeaseRequest() {
    return { $type: 'akash.market.v1beta3.QueryLeaseRequest', id: undefined };
}
exports.QueryLeaseRequest = {
    $type: 'akash.market.v1beta3.QueryLeaseRequest',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.id !== undefined) {
            lease_1.LeaseID.encode(message.id, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryLeaseRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = lease_1.LeaseID.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryLeaseRequest.$type,
            id: isSet(object.id) ? lease_1.LeaseID.fromJSON(object.id) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = lease_1.LeaseID.toJSON(message.id);
        }
        return obj;
    },
    create(base) {
        return exports.QueryLeaseRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryLeaseRequest();
        message.id =
            object.id !== undefined && object.id !== null
                ? lease_1.LeaseID.fromPartial(object.id)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryLeaseRequest.$type, exports.QueryLeaseRequest);
function createBaseQueryLeaseResponse() {
    return {
        $type: 'akash.market.v1beta3.QueryLeaseResponse',
        lease: undefined,
        escrowPayment: undefined,
    };
}
exports.QueryLeaseResponse = {
    $type: 'akash.market.v1beta3.QueryLeaseResponse',
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.lease !== undefined) {
            lease_1.Lease.encode(message.lease, writer.uint32(10).fork()).ldelim();
        }
        if (message.escrowPayment !== undefined) {
            types_1.FractionalPayment.encode(message.escrowPayment, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryLeaseResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.lease = lease_1.Lease.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.escrowPayment = types_1.FractionalPayment.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            $type: exports.QueryLeaseResponse.$type,
            lease: isSet(object.lease) ? lease_1.Lease.fromJSON(object.lease) : undefined,
            escrowPayment: isSet(object.escrowPayment)
                ? types_1.FractionalPayment.fromJSON(object.escrowPayment)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.lease !== undefined) {
            obj.lease = lease_1.Lease.toJSON(message.lease);
        }
        if (message.escrowPayment !== undefined) {
            obj.escrowPayment = types_1.FractionalPayment.toJSON(message.escrowPayment);
        }
        return obj;
    },
    create(base) {
        return exports.QueryLeaseResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseQueryLeaseResponse();
        message.lease =
            object.lease !== undefined && object.lease !== null
                ? lease_1.Lease.fromPartial(object.lease)
                : undefined;
        message.escrowPayment =
            object.escrowPayment !== undefined && object.escrowPayment !== null
                ? types_1.FractionalPayment.fromPartial(object.escrowPayment)
                : undefined;
        return message;
    },
};
typeRegistry_1.messageTypeRegistry.set(exports.QueryLeaseResponse.$type, exports.QueryLeaseResponse);
exports.QueryServiceName = 'akash.market.v1beta3.Query';
class QueryClientImpl {
    constructor(rpc, opts) {
        this.service = (opts === null || opts === void 0 ? void 0 : opts.service) || exports.QueryServiceName;
        this.rpc = rpc;
        this.Orders = this.Orders.bind(this);
        this.Order = this.Order.bind(this);
        this.Bids = this.Bids.bind(this);
        this.Bid = this.Bid.bind(this);
        this.Leases = this.Leases.bind(this);
        this.Lease = this.Lease.bind(this);
    }
    Orders(request) {
        const data = exports.QueryOrdersRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, 'Orders', data);
        return promise.then((data) => exports.QueryOrdersResponse.decode(minimal_1.default.Reader.create(data)));
    }
    Order(request) {
        const data = exports.QueryOrderRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, 'Order', data);
        return promise.then((data) => exports.QueryOrderResponse.decode(minimal_1.default.Reader.create(data)));
    }
    Bids(request) {
        const data = exports.QueryBidsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, 'Bids', data);
        return promise.then((data) => exports.QueryBidsResponse.decode(minimal_1.default.Reader.create(data)));
    }
    Bid(request) {
        const data = exports.QueryBidRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, 'Bid', data);
        return promise.then((data) => exports.QueryBidResponse.decode(minimal_1.default.Reader.create(data)));
    }
    Leases(request) {
        const data = exports.QueryLeasesRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, 'Leases', data);
        return promise.then((data) => exports.QueryLeasesResponse.decode(minimal_1.default.Reader.create(data)));
    }
    Lease(request) {
        const data = exports.QueryLeaseRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, 'Lease', data);
        return promise.then((data) => exports.QueryLeaseResponse.decode(minimal_1.default.Reader.create(data)));
    }
}
exports.QueryClientImpl = QueryClientImpl;
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
