// solana-agent-kit/dist/tools/okx-dex/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initDexClient = initDexClient;
const okx_dex_sdk_1 = require("@okx-dex/okx-dex-sdk");
const bs58_1 = __importDefault(require("bs58"));
function initDexClient(agent) {
    // Validate the required config parameters are present
    if (!agent.config.OKX_API_KEY ||
        !agent.config.OKX_SECRET_KEY ||
        !agent.config.OKX_API_PASSPHRASE ||
        !agent.config.OKX_PROJECT_ID) {
        throw new Error("Missing required OKX DEX configuration in agent config");
    }
    return new okx_dex_sdk_1.OKXDexClient({
        apiKey: agent.config.OKX_API_KEY,
        secretKey: agent.config.OKX_SECRET_KEY,
        apiPassphrase: agent.config.OKX_API_PASSPHRASE,
        projectId: agent.config.OKX_PROJECT_ID,
        solana: {
            connection: {
                rpcUrl: agent.connection.rpcEndpoint,
                confirmTransactionInitialTimeout: 60000,
            },
            privateKey: bs58_1.default.encode(agent.wallet.secretKey),
            walletAddress: agent.wallet_address.toString(),
        },
    });
}
