// quais/lib/commonjs/transaction/index.js
"use strict";
/**
 * Each state-changing operation on Ethereum requires a transaction.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuaiTransaction = exports.QiTransaction = exports.UTXO = exports.denominations = exports.FewestCoinSelector = exports.AbstractTransaction = exports.accessListify = void 0;
null;
var accesslist_js_1 = require("./accesslist.js");
Object.defineProperty(exports, "accessListify", { enumerable: true, get: function () { return accesslist_js_1.accessListify; } });
var abstract_transaction_js_1 = require("./abstract-transaction.js");
Object.defineProperty(exports, "AbstractTransaction", { enumerable: true, get: function () { return abstract_transaction_js_1.AbstractTransaction; } });
var coinselector_fewest_js_1 = require("./coinselector-fewest.js");
Object.defineProperty(exports, "FewestCoinSelector", { enumerable: true, get: function () { return coinselector_fewest_js_1.FewestCoinSelector; } });
var utxo_js_1 = require("./utxo.js");
Object.defineProperty(exports, "denominations", { enumerable: true, get: function () { return utxo_js_1.denominations; } });
Object.defineProperty(exports, "UTXO", { enumerable: true, get: function () { return utxo_js_1.UTXO; } });
var qi_transaction_js_1 = require("./qi-transaction.js");
Object.defineProperty(exports, "QiTransaction", { enumerable: true, get: function () { return qi_transaction_js_1.QiTransaction; } });
var quai_transaction_js_1 = require("./quai-transaction.js");
Object.defineProperty(exports, "QuaiTransaction", { enumerable: true, get: function () { return quai_transaction_js_1.QuaiTransaction; } });
