// @cks-systems/manifest-sdk/dist/esm/src/manifest/instructions/GlobalClean.js
/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */
import * as beet from '@metaplex-foundation/beet';
import * as web3 from '@solana/web3.js';
import { globalCleanParamsBeet, } from '../types/GlobalCleanParams';
/**
 * @category Instructions
 * @category GlobalClean
 * @category generated
 */
export const GlobalCleanStruct = new beet.BeetArgsStruct([
    ['instructionDiscriminator', beet.u8],
    ['params', globalCleanParamsBeet],
], 'GlobalCleanInstructionArgs');
export const globalCleanInstructionDiscriminator = 12;
/**
 * Creates a _GlobalClean_ instruction.
 *
 * @param accounts that will be accessed while the instruction is processed
 * @param args to provide as instruction data to the program
 *
 * @category Instructions
 * @category GlobalClean
 * @category generated
 */
export function createGlobalCleanInstruction(accounts, args, programId = new web3.PublicKey('MNFSTqtC93rEfYHB6hF82sKdZpUDFWkViLByLd1k1Ms')) {
    const [data] = GlobalCleanStruct.serialize({
        instructionDiscriminator: globalCleanInstructionDiscriminator,
        ...args,
    });
    const keys = [
        {
            pubkey: accounts.payer,
            isWritable: true,
            isSigner: true,
        },
        {
            pubkey: accounts.market,
            isWritable: true,
            isSigner: false,
        },
        {
            pubkey: accounts.systemProgram ?? web3.SystemProgram.programId,
            isWritable: false,
            isSigner: false,
        },
        {
            pubkey: accounts.global,
            isWritable: true,
            isSigner: false,
        },
    ];
    const ix = new web3.TransactionInstruction({
        programId,
        keys,
        data,
    });
    return ix;
}
