// @tensor-hq/tensor-common/dist/cjs/tensor/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TLOCK_PDA_ADDR = exports.TROLL_WITHDRAW_COSIGNER_ADDR = exports.TROLL_COSIGNER_ADDR = exports.TROLL_TREASURY_ADDR = exports.BROKER_ADDRS = exports.GS_BROKER_ADDR = exports.TCOMP_TRAIT_BID_COSIGNER = exports.TCOMP_PDA_ADDR = exports.TLIST_OWNER = exports.TLIST_COSIGNER = exports.TSWAP_OWNER = exports.TSWAP_COSIGNER = exports.TSWAP_PDA_ADDR = exports.CURRENCY_ADDRS = exports.USDC_ADDR = exports.TENSOR_LUT_DEVNET_ADDR = exports.TENSOR_LUT_ADDR = exports.TPAIR_PROGRAM_ID = exports.TGARD_PROGRAM_ID = exports.TDROP_PROGRAM_ID = exports.TLOCK_PROGRAM_ID = exports.TROLL_PROGRAM_ID = exports.SWAPSORIAN_PROGRAM_ID = exports.TSTKE_PROGRAM_ID = exports.TCOMP_PROGRAM_ID = exports.TBID_PROGRAM_ID = exports.TLIST_PROGRAM_ID = exports.TSWAP_PROGRAM_ID = void 0;
const web3_js_1 = require("@solana/web3.js");
// (!) KEEP IN SYNC WITH RESPECTIVE PROTOCOLS
// --------------------------------------- program IDs
exports.TSWAP_PROGRAM_ID = new web3_js_1.PublicKey('TSWAPaqyCSx2KABk68Shruf4rp7CxcNi8hAsbdwmHbN');
exports.TLIST_PROGRAM_ID = new web3_js_1.PublicKey('TL1ST2iRBzuGTqLn1KXnGdSnEow62BzPnGiqyRXhWtW');
exports.TBID_PROGRAM_ID = new web3_js_1.PublicKey('TB1Dqt8JeKQh7RLDzfYDJsq8KS4fS2yt87avRjyRxMv');
exports.TCOMP_PROGRAM_ID = new web3_js_1.PublicKey('TCMPhJdwDryooaGtiocG1u3xcYbRpiJzb283XfCZsDp');
exports.TSTKE_PROGRAM_ID = new web3_js_1.PublicKey('TSTKEiz9sqJRypokAkRhaW29rnDYDSxqWxmdv9brkp2');
exports.SWAPSORIAN_PROGRAM_ID = new web3_js_1.PublicKey('SWPhxKY7ponWjkfYCnvWypX8pEMe9hvQHhKo9wS7vim');
exports.TROLL_PROGRAM_ID = new web3_js_1.PublicKey('TRoLL7U1qTaqv2FFQ4jneZx5SetannKmrYCR778AkQZ');
exports.TLOCK_PROGRAM_ID = new web3_js_1.PublicKey('TLoCKic2wGJm7VhZKumih4Lc35fUhYqVMgA4j389Buk');
exports.TDROP_PROGRAM_ID = new web3_js_1.PublicKey('TDRoPy8i5G8AMzuaGPb98fxDRevS81kfhVeaipyWGbN');
exports.TGARD_PROGRAM_ID = new web3_js_1.PublicKey('TGARDaEzs7px1tEUssCCZ9ewhTW7oCA1MnY5y7rQk9n');
exports.TPAIR_PROGRAM_ID = new web3_js_1.PublicKey('TPA1R3GSAgUcZRcJXz5EU8Z7Y7w9XxoXz5fguY3anvM');
// --------------------------------------- address lookup table
exports.TENSOR_LUT_ADDR = new web3_js_1.PublicKey('4NYENhRXdSq1ek7mvJyzMUvdn2aN3JeAr6huzfL7869j');
// Differs due to a recent slot being used in the derivation path.
exports.TENSOR_LUT_DEVNET_ADDR = new web3_js_1.PublicKey('5z3YFrEgVqoTG8Eq1oCBggWSpLzTrofbsZBuytJTvUy6');
// --------------------------------------- SPL currencies
exports.USDC_ADDR = new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v');
exports.CURRENCY_ADDRS = [exports.USDC_ADDR];
// --------------------------------------- tswap
exports.TSWAP_PDA_ADDR = web3_js_1.PublicKey.findProgramAddressSync([], exports.TSWAP_PROGRAM_ID)[0];
exports.TSWAP_COSIGNER = new web3_js_1.PublicKey('6WQvG9Z6D1NZM76Ljz3WjgR7gGXRBJohHASdQxXyKi8q');
exports.TSWAP_OWNER = new web3_js_1.PublicKey('99cmWwQMqMFzMPx85rvZYKwusGSjZUDsu6mqYV4iisiz');
// --------------------------------------- tlist
exports.TLIST_COSIGNER = new web3_js_1.PublicKey('5aB7nyNJTuQZdKnhZXQHNhT16tBNevCuLRp14btvANxu');
exports.TLIST_OWNER = new web3_js_1.PublicKey('99cmWwQMqMFzMPx85rvZYKwusGSjZUDsu6mqYV4iisiz');
// --------------------------------------- tcomp
exports.TCOMP_PDA_ADDR = web3_js_1.PublicKey.findProgramAddressSync([], exports.TCOMP_PROGRAM_ID)[0];
exports.TCOMP_TRAIT_BID_COSIGNER = new web3_js_1.PublicKey('2C1skPhbfCW4q91WBEnbxuwEz4JBLtBwfmLXL1Wwy4MH');
exports.GS_BROKER_ADDR = new web3_js_1.PublicKey('3g2nyraTXqEKke3sTtZw9JtfjCo8Hzw6qhKe8K2hrYuf');
exports.BROKER_ADDRS = [exports.TCOMP_PDA_ADDR, exports.GS_BROKER_ADDR];
// --------------------------------------- troll
exports.TROLL_TREASURY_ADDR = new web3_js_1.PublicKey('HATUHhpGy5moXuwTZKr1qZREmKANcLy3kRiMQZvUseLE');
exports.TROLL_COSIGNER_ADDR = new web3_js_1.PublicKey('5qGy8rknMjt1S6V2YMGVidtuHpj1BVs6chzpjCDb47sB');
exports.TROLL_WITHDRAW_COSIGNER_ADDR = new web3_js_1.PublicKey('C6v1Mb5K9gV1c7iYjEP5YWfQ2VLh1wjkmZ7bA3cJdKP8');
// --------------------------------------- tlock
exports.TLOCK_PDA_ADDR = web3_js_1.PublicKey.findProgramAddressSync([], exports.TLOCK_PROGRAM_ID)[0];
